{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "import feather\n",
    "import matplotlib.pyplot as plt\n",
    "from itertools import combinations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "if \"/proj/yunligrp/users/minzhi/custom_lib\" not in sys.path:\n",
    "    sys.path.insert(0, \"/proj/yunligrp/users/minzhi/custom_lib\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "from function_process_data_eqtl import *\n",
    "from function_asso import *\n",
    "from function_mesa_cca import *\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_snp(snp_name, snp_dir):\n",
    "    snp_filename = \"SNP_%s_whole.txt\"%snp_name\n",
    "    snp_dir_filename = os.path.join(snp_dir, snp_filename)\n",
    "    snp_df = pd.read_csv(snp_dir_filename, sep=\"\\t\")\n",
    "    snp_df_nona = snp_df.dropna(axis=1)\n",
    "    snp_pure_trans = snp_df_nona.transpose()\n",
    "    return snp_pure_trans\n",
    "\n",
    "def recode_snp(snp_df):\n",
    "    sample_num = snp_df.shape[0]\n",
    "    snp_recoded = snp_df.copy()\n",
    "    for sample_i in range(sample_num):\n",
    "        geno_i = snp_df.iloc[sample_i, 0]\n",
    "        geno_i2list = geno_i.split(\"|\")\n",
    "        geno_i_recode = int(geno_i2list[0]) + int(geno_i2list[1])\n",
    "        snp_recoded.iloc[sample_i, 0] = geno_i_recode\n",
    "#        if sample_i%10000 == 0:\n",
    "#            print(\"%d/%d recoded.\"%(sample_i + 1, sample_num + 1))\n",
    "    return snp_recoded\n",
    "\n",
    "def snp_drop_value(snp_df, drop_value):\n",
    "    character_value_dict = {0:[drop_value, \"neq\"]}\n",
    "    snp_filtered = filter_characters(snp_df, character_value_dict)\n",
    "    return snp_filtered\n",
    "\n",
    "def snp_preprocess(snp_name, snp_dir, drop_value, save_dir):\n",
    "    snp_pure_trans = read_snp(snp_name, snp_dir)\n",
    "    snp_recoded = recode_snp(snp_pure_trans)\n",
    "    snp_filtered = snp_drop_value(snp_recoded, drop_value)\n",
    "    snp_filtered_filename = \"snp_%s_hetero_noheader.tsv\"%snp_name\n",
    "    snp_filtered_dir_filename = os.path.join(save_dir, snp_filtered_filename)\n",
    "    snp_filtered.to_csv(snp_filtered_dir_filename, sep=\"\\t\", header=False, index=True)\n",
    "    \n",
    "def snp_add_head(snp_name, snp_dir, save_dir):\n",
    "    snp_filename = \"snp_%s_hetero_noheader.tsv\"%snp_name\n",
    "    snp_dir_filename = os.path.join(snp_dir, snp_filename)\n",
    "    snp_df_orig = pd.read_csv(snp_dir_filename, sep=\"\\t\", index_col=None, header=None)\n",
    "    snp_df = snp_df_orig.rename(columns={0: \"NWDID\", 1: \"geno\"})\n",
    "    snp_save_filename = \"snp_%s_hetero.tsv\"%snp_name\n",
    "    snp_save_dir_filename = os.path.join(save_dir, snp_save_filename)\n",
    "    snp_df.to_csv(snp_save_dir_filename, sep=\"\\t\", header=True, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def kinship_select_sample(kinship_df_ID, sample_df):\n",
    "    common_col = \"NWDID\"\n",
    "    kinship_sample_df_list = [kinship_df_ID, sample_df]\n",
    "    kinship_df_row_selected = merge_df_list(kinship_sample_df_list, common_col, merge_method='first')\n",
    "    sample_list = kinship_df_row_selected[common_col].values.reshape(-1,).tolist()\n",
    "    kinship_sample_selected = kinship_df_row_selected[[\"NWDID\"] + sample_list]\n",
    "    return kinship_sample_selected"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def cn2binary_df(cn_df):\n",
    "    sample_num = cn_df.shape[0]\n",
    "    binary_array = np.zeros((sample_num, 5))\n",
    "    sample_df = cn_df[[\"NWDID\"]]\n",
    "    for sample_i in range(sample_num):\n",
    "        binary_array[sample_i, int(cn_df.iloc[sample_i, 1])] = 1\n",
    "    binary_df_noid_columns = [\"cn0\", \"cn1\", \"cn2\", \"cn3\", \"cn4\"]\n",
    "    binary_df_noid = pd.DataFrame(data=binary_array, columns=binary_df_noid_columns)\n",
    "    binary_df = pd.concat(objs=[sample_df, binary_df_noid], axis=1)\n",
    "    return binary_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def cn2del(cn_df):\n",
    "    cn2del_dict = {0:2, 1:1, 2:0, 3:0, 4:0}\n",
    "    del_series = cn_df[\"cn\"].map(cn2del_dict)\n",
    "#    display(del_series)\n",
    "    del_df = cn_df.copy()\n",
    "#    display(del_df)\n",
    "    del_df[\"cn\"] = del_series\n",
    "#    display(del_df)\n",
    "    del_df = del_df.rename(columns={\"cn\":\"del\"})\n",
    "#    display(del_df)\n",
    "    return del_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def remove_elements(df_1, df_2, common_col):\n",
    "    list_1 = df_1[common_col].values.tolist()\n",
    "    list_2 = df_2[common_col].values.tolist()\n",
    "    list_1_not_in_list_2 = list(filter(lambda x: x not in list_2, list_1))\n",
    "    list_1_not_in_list_2_df = pd.DataFrame(data=list_1_not_in_list_2, columns=[common_col])\n",
    "    df_1_not_in_df_2 = merge_df_list([df_1, list_1_not_in_list_2_df], common_col, merge_method='merge')\n",
    "    return df_1_not_in_df_2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def check_nan_snp(snp):\n",
    "    sample_orig = snp.shape[1]\n",
    "    snp_df_nona = snp.dropna(axis=1, how='any')\n",
    "    sample_nona = snp_df_nona.shape[1]\n",
    "    return sample_orig, sample_nona"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "def one_condition_conversion(df, condition_header, thres, new_header):\n",
    "    df_converted = df.copy()\n",
    "    df_converted[new_header] = np.where(df[condition_header]>=thres, 0, 1)\n",
    "    df_converted.loc[pd.Index(df[condition_header].isna()), new_header] = np.nan\n",
    "    df_converted[\"age_at_%s\"%new_header] = df[\"age_at_%s\"%condition_header]\n",
    "    return df_converted\n",
    "\n",
    "def two_condition_conversion(df, cd1_header, thres1, thres2, cd2_header, new_header):\n",
    "    df_converted = df.copy()\n",
    "    df_converted.loc[pd.Index(df[cd1_header].isna()),new_header] = np.nan\n",
    "    df_converted.loc[pd.Index(df[cd2_header].isna()),new_header] = np.nan\n",
    "    df_converted.loc[(df[cd1_header] < thres1) & (df[cd2_header] == 1), new_header] = 1\n",
    "    df_converted.loc[(df[cd1_header] < thres2) & (df[cd2_header] == 0), new_header] = 1\n",
    "    df_converted.loc[(df[cd1_header] >= thres1) & (df[cd2_header] == 1), new_header] = 0\n",
    "    df_converted.loc[(df[cd1_header] >= thres2) & (df[cd2_header] == 0), new_header] = 0\n",
    "    df_converted[\"age_at_%s\"%new_header] = df[\"age_at_%s\"%cd1_header]\n",
    "    return df_converted"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "def pheno_prep(cohort, egfr_df, anno_df, pheno_df):\n",
    "    cn_filename = \"%s_cn.tsv\"%cohort\n",
    "    cn_dir = os.path.join(\"..\", \"cohort\", cohort, \"cn\")\n",
    "    cn_dir_filename = os.path.join(cn_dir, cn_filename)\n",
    "    cn_df = pd.read_csv(cn_dir_filename, sep = \"\\t\", header = 0, index_col = None)\n",
    "    anno_df = anno_df[[\"NWDID\", \"unique_subject_key\", \"sex\"]]\n",
    "    \n",
    "    anno_cn_common_col = \"NWDID\"\n",
    "    anno_cn_df_list = [anno_df, cn_df]\n",
    "    common_anno_cn = merge_df_list(anno_cn_df_list, anno_cn_common_col, merge_method='first', how = 'inner')\n",
    "    \n",
    "    pheno_anno_common_col = \"unique_subject_key\"\n",
    "    pheno_anno_list = [pheno_df, common_anno_cn]\n",
    "    common_pheno_anno_cn = merge_df_list(pheno_anno_list, pheno_anno_common_col, merge_method='merge', how = 'inner')\n",
    "    \n",
    "    egfr_pheno_common_col = \"SUBJECT_ID\"\n",
    "    pheno_egfr_list = [common_pheno_anno_cn, egfr_df]\n",
    "    common_pheno_anno_cn_egfr = merge_df_list(pheno_egfr_list, egfr_pheno_common_col, merge_method='merge', how = 'left')\n",
    "    return common_pheno_anno_cn_egfr\n",
    "\n",
    "def pheno_adding_features_recode_sex(pheno_raw_df, save_dir, save_filename):\n",
    "    sex_dict = {\"M\":1, \"F\":0}\n",
    "    pheno_raw_df = pheno_raw_df.replace({\"sex\": sex_dict})\n",
    "    pheno_wckd_df = one_condition_conversion(pheno_raw_df, \"EGFRCKDEPI\", 60, \"CKD\")\n",
    "    pheno_wckd_wmicrocytosis_df = one_condition_conversion(pheno_wckd_df, \"mcv_entvol_rbc_1\", 80, \"microcytosis\")\n",
    "    pheno_wckd_wmicrocytosis_anemia_df = two_condition_conversion(pheno_wckd_wmicrocytosis_df, \n",
    "                                                                  \"hemoglobin_mcnc_bld_1\", 13, 12, \"sex\", \"anemia\")\n",
    "    save_dir_filename = os.path.join(save_dir, save_filename)\n",
    "    pheno_wckd_wmicrocytosis_anemia_df.to_csv(save_dir_filename, sep = \"\\t\", header=True, index=False)\n",
    "    return pheno_wckd_wmicrocytosis_anemia_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "def duplicates_num(df, col):\n",
    "    boolean_series = df[[col]].duplicated(keep='first')\n",
    "    dup_num = boolean_series.sum()\n",
    "    return dup_num"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "def move_cn(cn, cn_selection):\n",
    "    cn_header = list(cn)\n",
    "    cn_header.remove(\"NWDID\")\n",
    "    cn_header.remove(cn_selection)\n",
    "    cn_type_list_moved = [\"NWDID\", cn_selection] + cn_header\n",
    "    cn_moved = cn[cn_type_list_moved]\n",
    "    return cn_moved"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "def cati2dummy_df(df_original, col_list):\n",
    "    full_col = list(df_original)\n",
    "    quant_list = [x for x in full_col if x not in col_list]\n",
    "    dummy_col_list = []\n",
    "    df_full = df_original.copy()\n",
    "    for col in col_list:\n",
    "        dummy_df, dummy_col = cati2dummy_single_col(df_full[col], col)\n",
    "        df_full.drop(axis = 1, labels = [col], inplace = True)\n",
    "        if dummy_df is not None and (dummy_col is not None):\n",
    "            df_full = pd.concat([df_full, dummy_df], axis = 1)\n",
    "            dummy_col_list = dummy_col_list + dummy_col\n",
    "        else:\n",
    "            continue\n",
    "    full_list = quant_list + dummy_col_list\n",
    "    return df_full, dummy_col_list\n",
    "\n",
    "def cati2dummy_single_col(cati_series, col):\n",
    "    cati_array = pd.factorize(cati_series)[0]\n",
    "    cati_num = len(set(cati_array))\n",
    "    if cati_num == 1:\n",
    "        dummy_df = None\n",
    "        dummy_col = None\n",
    "        return dummy_df, dummy_col\n",
    "    dummy_array = (cati_array[:,None] == np.arange(cati_array.max())+1).astype(int)\n",
    "    if cati_num <= 2:\n",
    "        dummy_col = [col]\n",
    "    elif cati_num > 2:\n",
    "        dummy_col_num = dummy_array.shape[1]\n",
    "        dummy_col = []\n",
    "        for dummy_col_i in range(dummy_col_num):\n",
    "            tmp_col = [\"%s_%d\"%(col, dummy_col_i)]\n",
    "            dummy_col = dummy_col + tmp_col\n",
    "    dummy_df = pd.DataFrame(data = dummy_array, columns = dummy_col)\n",
    "    dummy_df[pd.isnull(cati_series)] = np.nan\n",
    "    return dummy_df, dummy_col"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Genotype"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-22-1f1ffc019b07>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      4\u001b[0m     \u001b[0msnp_filename\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m\"SNP_%s_whole.txt\"\u001b[0m\u001b[0;34m%\u001b[0m\u001b[0msnp_name_i\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m     \u001b[0msnp_dir_filename\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mos\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpath\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mjoin\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msnp_dir\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msnp_filename\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 6\u001b[0;31m     \u001b[0msnp_df\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread_csv\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msnp_dir_filename\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msep\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"\\t\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mheader\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mindex_col\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      7\u001b[0m     \u001b[0msample_orig\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msample_nona\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcheck_nan_snp\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msnp_df\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      8\u001b[0m     \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msnp_name_i\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\":\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msample_orig\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msample_nona\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/lib/python3.6/site-packages/pandas/io/parsers.py\u001b[0m in \u001b[0;36mparser_f\u001b[0;34m(filepath_or_buffer, sep, delimiter, header, names, index_col, usecols, squeeze, prefix, mangle_dupe_cols, dtype, engine, converters, true_values, false_values, skipinitialspace, skiprows, skipfooter, nrows, na_values, keep_default_na, na_filter, verbose, skip_blank_lines, parse_dates, infer_datetime_format, keep_date_col, date_parser, dayfirst, iterator, chunksize, compression, thousands, decimal, lineterminator, quotechar, quoting, doublequote, escapechar, comment, encoding, dialect, tupleize_cols, error_bad_lines, warn_bad_lines, delim_whitespace, low_memory, memory_map, float_precision)\u001b[0m\n\u001b[1;32m    700\u001b[0m                     skip_blank_lines=skip_blank_lines)\n\u001b[1;32m    701\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 702\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0m_read\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfilepath_or_buffer\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkwds\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    703\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    704\u001b[0m     \u001b[0mparser_f\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__name__\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/lib/python3.6/site-packages/pandas/io/parsers.py\u001b[0m in \u001b[0;36m_read\u001b[0;34m(filepath_or_buffer, kwds)\u001b[0m\n\u001b[1;32m    433\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    434\u001b[0m     \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 435\u001b[0;31m         \u001b[0mdata\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mparser\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mnrows\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    436\u001b[0m     \u001b[0;32mfinally\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    437\u001b[0m         \u001b[0mparser\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mclose\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/lib/python3.6/site-packages/pandas/io/parsers.py\u001b[0m in \u001b[0;36mread\u001b[0;34m(self, nrows)\u001b[0m\n\u001b[1;32m   1137\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mread\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mnrows\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mNone\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1138\u001b[0m         \u001b[0mnrows\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0m_validate_integer\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'nrows'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mnrows\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1139\u001b[0;31m         \u001b[0mret\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_engine\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mnrows\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1140\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1141\u001b[0m         \u001b[0;31m# May alter columns / col_dict\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/lib/python3.6/site-packages/pandas/io/parsers.py\u001b[0m in \u001b[0;36mread\u001b[0;34m(self, nrows)\u001b[0m\n\u001b[1;32m   1993\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mread\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mnrows\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mNone\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1994\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1995\u001b[0;31m             \u001b[0mdata\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_reader\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mnrows\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1996\u001b[0m         \u001b[0;32mexcept\u001b[0m \u001b[0mStopIteration\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1997\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_first_chunk\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/parsers.pyx\u001b[0m in \u001b[0;36mpandas._libs.parsers.TextReader.read\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/parsers.pyx\u001b[0m in \u001b[0;36mpandas._libs.parsers.TextReader._read_low_memory\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/parsers.pyx\u001b[0m in \u001b[0;36mpandas._libs.parsers.TextReader._read_rows\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/parsers.pyx\u001b[0m in \u001b[0;36mpandas._libs.parsers.TextReader._convert_column_data\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/parsers.pyx\u001b[0m in \u001b[0;36mpandas._libs.parsers.TextReader._convert_tokens\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/parsers.pyx\u001b[0m in \u001b[0;36mpandas._libs.parsers.TextReader._convert_with_dtype\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/lib/python3.6/site-packages/pandas/core/dtypes/common.py\u001b[0m in \u001b[0;36mis_extension_array_dtype\u001b[0;34m(arr_or_dtype)\u001b[0m\n\u001b[1;32m   1747\u001b[0m     \u001b[0mdtype\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mgetattr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0marr_or_dtype\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'dtype'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0marr_or_dtype\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1748\u001b[0m     return (isinstance(dtype, ExtensionDtype) or\n\u001b[0;32m-> 1749\u001b[0;31m             registry.find(dtype) is not None)\n\u001b[0m\u001b[1;32m   1750\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1751\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/lib/python3.6/site-packages/pandas/core/dtypes/dtypes.py\u001b[0m in \u001b[0;36mfind\u001b[0;34m(self, dtype)\u001b[0m\n\u001b[1;32m     79\u001b[0m             \u001b[0mdtype_type\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdtype\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     80\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdtype\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtype\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 81\u001b[0;31m                 \u001b[0mdtype_type\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtype\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdtype\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     82\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0missubclass\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdtype_type\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mExtensionDtype\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     83\u001b[0m                 \u001b[0;32mreturn\u001b[0m \u001b[0mdtype\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "snp_name_list = [\"rs33930165\", \"rs11865131\", \"rs11248850\", \"rs334\"]\n",
    "snp_dir = os.path.join(\"..\", \"raw_data\", \"snp\")\n",
    "for snp_name_i in snp_name_list:\n",
    "    snp_filename = \"SNP_%s_whole.txt\"%snp_name_i\n",
    "    snp_dir_filename = os.path.join(snp_dir, snp_filename)\n",
    "    snp_df = pd.read_csv(snp_dir_filename, sep=\"\\t\", header=0, index_col = None)\n",
    "    sample_orig, sample_nona = check_nan_snp(snp_df)\n",
    "    print(snp_name_i, \":\", sample_orig, sample_nona)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "snp_name_list = [\"rs33930165\", \"rs11865131\", \"rs11248850\", \"rs334\"]\n",
    "snp_dir = os.path.join(\"..\", \"raw_data\", \"snp\")\n",
    "save_dir = os.path.join(\"..\", \"prepro_data\", \"snp\")\n",
    "drop_value = 2\n",
    "for snp_name_i in snp_name_list:\n",
    "    snp_preprocess(snp_name_i, snp_dir, drop_value, save_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "snp_name_list = [\"rs33930165\", \"rs11865131\", \"rs11248850\", \"rs334\"]\n",
    "snp_dir = os.path.join(\"..\", \"prepro_data\", \"snp\")\n",
    "save_dir = os.path.join(\"..\", \"prepro_data\", \"snp\")\n",
    "for snp_name_i in snp_name_list:\n",
    "    snp_add_head(snp_name_i, snp_dir, save_dir)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "toc-hr-collapsed": false
   },
   "source": [
    "# Association between Blood Cell Trait and JHS"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Preprocess Data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### CN"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "cohort = \"JHS\"\n",
    "cn_filename = \"JHS_CN.csv\"\n",
    "cn_dir = os.path.join(\"..\", \"cohort\", cohort, \"cn\")\n",
    "cn_dir_filename = os.path.join(cn_dir, cn_filename)\n",
    "cn = pd.read_csv(cn_dir_filename, sep=\",\")\n",
    "\n",
    "cn_save_filename = \"JHS_CN.tsv\"\n",
    "cn_dir_filename = os.path.join(cn_dir, cn_save_filename)\n",
    "cn.to_csv(cn_dir_filename, sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 10PC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "pc_filename = \"freeze5_genetic_ancestry_pca10.csv\"\n",
    "pc_load_dir = os.path.join(\"..\", \"raw_data\", \"10pc\")\n",
    "pc_dir_filename = os.path.join(pc_load_dir, pc_filename)\n",
    "pc_raw = pd.read_csv(pc_dir_filename, sep=\",\")\n",
    "pc = pc_raw.dropna(axis=1)\n",
    "\n",
    "pc_save_filename = \"freeze5_genetic_ancestry_pca10.tsv\"\n",
    "pc_save_dir = os.path.join(\"..\", \"prepro_data\", \"10pc\")\n",
    "pc_save_dir_filename = os.path.join(pc_save_dir, pc_save_filename)\n",
    "pc.to_csv(pc_save_dir_filename, sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Age and Sex"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "pheno_filename = \"jhs_basic_phenotypes_05072019.txt\"\n",
    "pheno_dir = os.path.join(\"..\", \"raw_data\", \"phenotype\")\n",
    "pheno_dir_filename = os.path.join(pheno_dir, pheno_filename)\n",
    "pheno_df = pd.read_csv(pheno_dir_filename, sep=\"\\t\")\n",
    "age_sex_raw = pheno_df[[\"NWDID\", \"age\", \"male\"]]\n",
    "age_sex = age_sex_raw.dropna(axis=1)\n",
    "age_sex_filename = \"jhs_basic_age_sex.tsv\"\n",
    "age_sex_dir = os.path.join(\"..\", \"prepro_data\", \"phenotype\")\n",
    "age_sex_dir_filename = os.path.join(age_sex_dir, age_sex_filename)\n",
    "age_sex.to_csv(age_sex_dir_filename, sep=\"\\t\", header=True, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Common Samples Among (Age and Sex), 10PC, CN and SNPs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Common (Age and Sex)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "cohort = \"JHS\"\n",
    "filename_list = [\"jhs_basic_age_sex.tsv\", \"freeze5_genetic_ancestry_pca10.tsv\", \"snp_rs334_hetero.tsv\",\n",
    "                 \"snp_rs33930165_hetero.tsv\", \"snp_rs11248850_hetero.tsv\", \"JHS_CN.tsv\"]\n",
    "predata_dir_list = [os.path.join(\"..\", \"prepro_data\", \"phenotype\"), os.path.join(\"..\", \"prepro_data\", \"10pc\"),\n",
    "                    os.path.join(\"..\", \"prepro_data\", \"snp\"), os.path.join(\"..\", \"prepro_data\", \"snp\"),\n",
    "                    os.path.join(\"..\", \"prepro_data\", \"snp\"), os.path.join(\"..\", \"cohort\", cohort, \"cn\")]\n",
    "var_num = len(predata_dir_list)\n",
    "age_sex_10pc_snps_cn_dir_filename_list = [os.path.join(predata_dir_list[var_idx], filename_list[var_idx]) for var_idx in range(var_num)]\n",
    "age_sex_10pc_snps_cn_df = read2df_list(age_sex_10pc_snps_cn_dir_filename_list)\n",
    "common_col = \"NWDID\"\n",
    "common_age_sex = merge_df_list(age_sex_10pc_snps_cn_df, common_col, merge_method='first')\n",
    "common_age_sex_filename = \"common_age-sex.tsv\"\n",
    "save_dir = os.path.join(\"..\", \"cohort\", cohort, \"ready_data\")\n",
    "common_age_sex_dir_filename = os.path.join(save_dir, common_age_sex_filename)\n",
    "common_age_sex.to_csv(common_age_sex_dir_filename, sep=\"\\t\", index=False)\n",
    "sample_df = common_age_sex[[\"NWDID\"]]\n",
    "sample_filename = \"common_sample.tsv\"\n",
    "sample_dir_filename = os.path.join(save_dir, sample_filename)\n",
    "sample_df.to_csv(sample_dir_filename, sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Common 10PC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "age_sex_10pc_df_list = [age_sex_10pc_snps_cn_df[1], common_age_sex]\n",
    "common_col = \"NWDID\"\n",
    "common_10pc = merge_df_list(age_sex_10pc_df_list, common_col, merge_method='first')\n",
    "common_10pc_filename = \"common_10pc.tsv\"\n",
    "common_10pc_dir_filename = os.path.join(save_dir, common_10pc_filename)\n",
    "common_10pc.to_csv(common_10pc_dir_filename, sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Common rs334"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "age_sex_rs334_df_list = [age_sex_10pc_snps_cn_df[2], common_age_sex]\n",
    "common_col = \"NWDID\"\n",
    "common_rs334 = merge_df_list(age_sex_rs334_df_list, common_col, merge_method='first')\n",
    "common_rs334_filename = \"common_rs334.tsv\"\n",
    "common_rs334_dir_filename = os.path.join(save_dir, common_rs334_filename)\n",
    "common_rs334.to_csv(common_rs334_dir_filename, sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Common rs33930165"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "age_sex_rs33930165_df_list = [age_sex_10pc_snps_cn_df[3], common_age_sex]\n",
    "common_col = \"NWDID\"\n",
    "common_rs33930165 = merge_df_list(age_sex_rs33930165_df_list, common_col, merge_method='first')\n",
    "common_rs33930165_filename = \"common_rs33930165.tsv\"\n",
    "common_rs33930165_dir_filename = os.path.join(save_dir, common_rs33930165_filename)\n",
    "common_rs33930165.to_csv(common_rs33930165_dir_filename, sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Common rs11248850"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "age_sex_rs11248850_df_list = [age_sex_10pc_snps_cn_df[4], common_age_sex]\n",
    "common_col = \"NWDID\"\n",
    "common_rs11248850 = merge_df_list(age_sex_rs11248850_df_list, common_col, merge_method='first')\n",
    "common_rs11248850_filename = \"common_rs11248850.tsv\"\n",
    "common_rs11248850_dir_filename = os.path.join(save_dir, common_rs11248850_filename)\n",
    "common_rs11248850.to_csv(common_rs11248850_dir_filename, sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Common CN"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "age_sex_cn_df_list = [age_sex_10pc_snps_cn_df[5], common_age_sex]\n",
    "common_col = \"NWDID\"\n",
    "common_cn = merge_df_list(age_sex_cn_df_list, common_col, merge_method='first')\n",
    "common_cn_filename = \"common_cn.tsv\"\n",
    "common_cn_dir_filename = os.path.join(save_dir, common_cn_filename)\n",
    "common_cn.to_csv(common_cn_dir_filename, sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Apply Common Samples to Kinship Matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "predata_dir = os.path.join(\"..\", \"prepro_data\", \"kinship\")\n",
    "kinship_filename = \"freeze5b_round2_pcrelate_kinship_renovate0.5_minzhi\"\n",
    "kinship_dir_filename = os.path.join(predata_dir, kinship_filename)\n",
    "kinship_df = feather.read_dataframe(\"%s.feather\"%kinship_dir_filename)\n",
    "sample_list = list(kinship_df)\n",
    "sample_df = pd.DataFrame(data=sample_list, columns=[\"NWDID\"])\n",
    "kinship_df_ID = pd.concat(objs=[sample_df, kinship_df], axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "save_dir = os.path.join(\"..\", \"cohort\", cohort, \"ready_data\")\n",
    "sample_filename = \"common_sample.tsv\"\n",
    "sample_dir_filename = os.path.join(save_dir, sample_filename)\n",
    "sample_df = pd.read_csv(sample_dir_filename, sep=\"\\t\")\n",
    "kinship_sample_selected = kinship_select_sample(kinship_df_ID, sample_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "common_kinship_filename = \"common_kinship.tsv\"\n",
    "common_kinship_dir_filename = os.path.join(save_dir, common_kinship_filename)\n",
    "kinship_sample_selected.to_csv(common_kinship_dir_filename, sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Phenotype"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "pheno_list = [\"hemoglobin_mcnc_bld_1\", \"hematocrit_vfr_bld_1\", \"rbc_ncnc_bld_1\",\"mcv_entvol_rbc_1\",\n",
    "              \"mch_entmass_rbc_1\", \"mchc_mcnc_rbc_1\", \"rdw_ratio_rbc_1\", \"lnHBA1C\", \"EGFRCKDEPI\",\n",
    "              \"anemia\", \"microcytosis\", \"CKD\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "rawdata_dir = os.path.join(\"..\",\"raw_data\", \"phenotype\")\n",
    "pheno_filename = \"jhs_basic_phenotypes_05072019.txt\"\n",
    "pheno_dir_filename = os.path.join(rawdata_dir, pheno_filename)\n",
    "pheno_full_df = pd.read_csv(pheno_dir_filename, sep=\"\\t\")\n",
    "pheno_df = pheno_full_df[[\"NWDID\"] + pheno_list]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "lnHBA1C_DIABETES_df = pheno_full_df[[\"lnHBA1C\", \"DIABETES\"]]\n",
    "sample_num = lnHBA1C_DIABETES_df.shape[0]\n",
    "lnHBA1C_DIABETES_df_noDIABETES = lnHBA1C_DIABETES_df.copy()\n",
    "for sample_i in range(sample_num):\n",
    "    if lnHBA1C_DIABETES_df.iloc[sample_i, 1] == 1:\n",
    "        lnHBA1C_DIABETES_df_noDIABETES.iloc[sample_i, 0] = np.nan\n",
    "lnHBA1C_DIABETES_df_noDIABETES.rename(columns={\"lnHBA1C\": \"lnHBA1C_noDIABETES\"}, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "pheno_df_lnHBA1C_noDIABETES = pd.concat(objs=[pheno_df, lnHBA1C_DIABETES_df_noDIABETES[\"lnHBA1C_noDIABETES\"]], axis=1)\n",
    "pheno_header_rearange_list = [\"hemoglobin_mcnc_bld_1\", \"hematocrit_vfr_bld_1\", \"rbc_ncnc_bld_1\",\"mcv_entvol_rbc_1\",\n",
    "                              \"mch_entmass_rbc_1\", \"mchc_mcnc_rbc_1\", \"rdw_ratio_rbc_1\", \"lnHBA1C\", \"lnHBA1C_noDIABETES\",\n",
    "                              \"EGFRCKDEPI\", \"anemia\", \"microcytosis\", \"CKD\"]\n",
    "pheno_df_lnHBA1C_noDIABETES = pheno_df_lnHBA1C_noDIABETES[[\"NWDID\"] + pheno_header_rearange_list]\n",
    "pheno_df_lnHBA1C_noDIABETES_filename = \"pheno_needed.tsv\"\n",
    "pheno_df_lnHBA1C_noDIABETES_dir = os.path.join(\"..\", \"cohort\", cohort, \"ready_data\")\n",
    "pheno_df_lnHBA1C_noDIABETES_dir_filename = os.path.join(pheno_df_lnHBA1C_noDIABETES_dir, pheno_df_lnHBA1C_noDIABETES_filename)\n",
    "pheno_df_lnHBA1C_noDIABETES.to_csv(pheno_df_lnHBA1C_noDIABETES_dir_filename, sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Preparing Data for Each Table"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Table 1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Recoding CN Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "cn_orig_filename = \"common_cn.tsv\"\n",
    "cn_orig_dir = os.path.join(\"..\", \"cohort\", cohort, \"ready_data\")\n",
    "cn_orig_dir_filename = os.path.join(cn_orig_dir, cn_orig_filename)\n",
    "cn_orig_df = pd.read_csv(cn_orig_dir_filename, sep=\"\\t\")\n",
    "cn_binary_df_full = cn2binary_df(cn_orig_df)\n",
    "cn_binary_df = cn_binary_df_full[[\"NWDID\", \"cn0\", \"cn1\", \"cn3\", \"cn4\"]]\n",
    "cn_binary_df_filename = \"common_cn-binary.tsv\"\n",
    "cn_binary_df_dir_filename = os.path.join(cn_orig_dir, cn_binary_df_filename)\n",
    "cn_binary_df.to_csv(cn_binary_df_dir_filename, sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### CN Binary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "phenotype_list = [\"hemoglobin_mcnc_bld_1\", \"hematocrit_vfr_bld_1\", \"rbc_ncnc_bld_1\",\"mcv_entvol_rbc_1\",\n",
    "                  \"mch_entmass_rbc_1\", \"mchc_mcnc_rbc_1\", \"rdw_ratio_rbc_1\", \"lnHBA1C\", \"lnHBA1C_noDIABETES\",\n",
    "                  \"EGFRCKDEPI\", \"anemia\", \"microcytosis\", \"CKD\"]\n",
    "load_dir = os.path.join(\"..\", \"cohort\", cohort, \"ready_data\")\n",
    "\n",
    "pheno_full_filename = \"pheno_needed.tsv\"\n",
    "pheno_full_dir_filename = os.path.join(load_dir, pheno_full_filename)\n",
    "pheno_full_df = pd.read_csv(pheno_full_dir_filename, sep=\"\\t\")\n",
    "\n",
    "var_name = \"cn-binary\"\n",
    "\n",
    "var_filename = \"common_cn-binary.tsv\"\n",
    "var_dir_filename = os.path.join(load_dir, var_filename)\n",
    "var_df = pd.read_csv(var_dir_filename, sep=\"\\t\")\n",
    "\n",
    "table_name = \"table1\"\n",
    "\n",
    "kinship_filename = \"common_kinship.tsv\"\n",
    "kinship_dir_filename = os.path.join(load_dir, kinship_filename)\n",
    "kinship_df_ID = pd.read_csv(kinship_dir_filename, sep=\"\\t\")\n",
    "\n",
    "age_sex_filename = \"common_age-sex.tsv\"\n",
    "age_sex_dir_filename = os.path.join(load_dir, age_sex_filename)\n",
    "age_sex_df = pd.read_csv(age_sex_dir_filename, sep=\"\\t\")\n",
    "\n",
    "pc10_filename = \"common_10pc.tsv\"\n",
    "pc10_dir_filename = os.path.join(load_dir, pc10_filename)\n",
    "pc10_df = pd.read_csv(pc10_dir_filename, sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "for phenotype in phenotype_list:\n",
    "    prepare_matrix_pheno(cohort, phenotype, pheno_full_df, var_name, var_df, table_name, kinship_df_ID, age_sex_df, pc10_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### rs334"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "phenotype_list = [\"hemoglobin_mcnc_bld_1\", \"hematocrit_vfr_bld_1\", \"rbc_ncnc_bld_1\",\"mcv_entvol_rbc_1\",\n",
    "                  \"mch_entmass_rbc_1\", \"mchc_mcnc_rbc_1\", \"rdw_ratio_rbc_1\", \"lnHBA1C\", \"lnHBA1C_noDIABETES\",\n",
    "                  \"EGFRCKDEPI\", \"anemia\", \"microcytosis\", \"CKD\"]\n",
    "load_dir = os.path.join(\"..\", \"cohort\", cohort, \"ready_data\")\n",
    "\n",
    "pheno_full_filename = \"pheno_needed.tsv\"\n",
    "pheno_full_dir_filename = os.path.join(load_dir, pheno_full_filename)\n",
    "pheno_full_df = pd.read_csv(pheno_full_dir_filename, sep=\"\\t\")\n",
    "\n",
    "var_name = \"rs334\"\n",
    "\n",
    "var_filename = \"common_rs334.tsv\"\n",
    "var_dir_filename = os.path.join(load_dir, var_filename)\n",
    "var_df = pd.read_csv(var_dir_filename, sep=\"\\t\")\n",
    "\n",
    "table_name = \"table1\"\n",
    "\n",
    "kinship_filename = \"common_kinship.tsv\"\n",
    "kinship_dir_filename = os.path.join(load_dir, kinship_filename)\n",
    "kinship_df_ID = pd.read_csv(kinship_dir_filename, sep=\"\\t\")\n",
    "\n",
    "age_sex_filename = \"common_age-sex.tsv\"\n",
    "age_sex_dir_filename = os.path.join(load_dir, age_sex_filename)\n",
    "age_sex_df = pd.read_csv(age_sex_dir_filename, sep=\"\\t\")\n",
    "\n",
    "pc10_filename = \"common_10pc.tsv\"\n",
    "pc10_dir_filename = os.path.join(load_dir, pc10_filename)\n",
    "pc10_df = pd.read_csv(pc10_dir_filename, sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "for phenotype in phenotype_list:\n",
    "    prepare_matrix_pheno(cohort, phenotype, pheno_full_df, var_name, var_df, table_name, kinship_df_ID, age_sex_df, pc10_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### rs33930165"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "phenotype_list = [\"hemoglobin_mcnc_bld_1\", \"hematocrit_vfr_bld_1\", \"rbc_ncnc_bld_1\",\"mcv_entvol_rbc_1\",\n",
    "                  \"mch_entmass_rbc_1\", \"mchc_mcnc_rbc_1\", \"rdw_ratio_rbc_1\", \"lnHBA1C\", \"lnHBA1C_noDIABETES\",\n",
    "                  \"EGFRCKDEPI\", \"anemia\", \"microcytosis\", \"CKD\"]\n",
    "load_dir = os.path.join(\"..\", \"cohort\", cohort, \"ready_data\")\n",
    "\n",
    "pheno_full_filename = \"pheno_needed.tsv\"\n",
    "pheno_full_dir_filename = os.path.join(load_dir, pheno_full_filename)\n",
    "pheno_full_df = pd.read_csv(pheno_full_dir_filename, sep=\"\\t\")\n",
    "\n",
    "var_name = \"rs33930165\"\n",
    "\n",
    "var_filename = \"common_rs33930165.tsv\"\n",
    "var_dir_filename = os.path.join(load_dir, var_filename)\n",
    "var_df = pd.read_csv(var_dir_filename, sep=\"\\t\")\n",
    "\n",
    "table_name = \"table1\"\n",
    "\n",
    "kinship_filename = \"common_kinship.tsv\"\n",
    "kinship_dir_filename = os.path.join(load_dir, kinship_filename)\n",
    "kinship_df_ID = pd.read_csv(kinship_dir_filename, sep=\"\\t\")\n",
    "\n",
    "age_sex_filename = \"common_age-sex.tsv\"\n",
    "age_sex_dir_filename = os.path.join(load_dir, age_sex_filename)\n",
    "age_sex_df = pd.read_csv(age_sex_dir_filename, sep=\"\\t\")\n",
    "\n",
    "pc10_filename = \"common_10pc.tsv\"\n",
    "pc10_dir_filename = os.path.join(load_dir, pc10_filename)\n",
    "pc10_df = pd.read_csv(pc10_dir_filename, sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "for phenotype in phenotype_list:\n",
    "    prepare_matrix_pheno(cohort, phenotype, pheno_full_df, var_name, var_df, table_name, kinship_df_ID, age_sex_df, pc10_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Table 2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### CN to DEL"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "load_dir = os.path.join(\"..\", \"cohort\", cohort, \"ready_data\")\n",
    "cn_filename = \"common_cn.tsv\"\n",
    "cn_dir_filename = os.path.join(load_dir, cn_filename)\n",
    "cn = pd.read_csv(cn_dir_filename, sep=\"\\t\")\n",
    "del_cn = cn2del(cn)\n",
    "del_filename = \"common_del.tsv\"\n",
    "del_dir_filename = os.path.join(load_dir, del_filename)\n",
    "del_cn.to_csv(del_dir_filename, sep=\"\\t\", header=True, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### DEL Ã— rs334"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "load_dir = os.path.join(\"..\", \"cohort\", cohort, \"ready_data\")\n",
    "del_filename = \"common_del.tsv\"\n",
    "del_dir_filename = os.path.join(load_dir, del_filename)\n",
    "del_cn = pd.read_csv(del_dir_filename, sep=\"\\t\")\n",
    "\n",
    "rs334_filename = \"common_rs334.tsv\"\n",
    "rs334_dir_filename = os.path.join(load_dir, rs334_filename)\n",
    "rs334 = pd.read_csv(rs334_dir_filename, sep=\"\\t\")\n",
    "\n",
    "del_rs334 = pd.concat(objs=[del_cn, rs334[[\"geno\"]]], axis=1)\n",
    "del_rs334_filename = \"common_del-rs334.tsv\"\n",
    "save_dir = load_dir\n",
    "del_rs334_dir_filename = os.path.join(save_dir, del_rs334_filename)\n",
    "del_rs334.to_csv(del_rs334_dir_filename, sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Classifying rs334 by Number of Copies of Deletion"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [],
   "source": [
    "load_dir = os.path.join(\"..\", \"cohort\", cohort, \"ready_data\")\n",
    "del_rs334_filename = \"common_del-rs334.tsv\"\n",
    "del_rs334_dir_filename = os.path.join(load_dir, del_rs334_filename)\n",
    "del_rs334 = pd.read_csv(del_rs334_dir_filename, sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### No Copies of deletion, a.k.a. CN = 2, 3, 4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [],
   "source": [
    "save_dir = os.path.join(\"..\", \"cohort\", cohort, \"ready_data\")\n",
    "character_value_dict = {\"NWDID\":[\"no_thres\"], \"DEL\":[0, \"eq\"], \"geno\":[\"no_thres\"]}\n",
    "snp_classified_del_del0 = filter_characters(del_rs334, character_value_dict)\n",
    "snp_classified_del0 = snp_classified_del_del0[[\"NWDID\", \"geno\"]]\n",
    "snp_classified_del0_filename = \"common_rs334_del0.tsv\"\n",
    "snp_classified_del0_dir_filename = os.path.join(save_dir, snp_classified_del0_filename)\n",
    "snp_classified_del0.to_csv(snp_classified_del0_dir_filename, sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 1 Copy of deletion, a.k.a. CN = 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [],
   "source": [
    "save_dir = os.path.join(\"..\", \"cohort\", cohort, \"ready_data\")\n",
    "character_value_dict = {\"NWDID\":[\"no_thres\"], \"DEL\":[1, \"eq\"], \"geno\":[\"no_thres\"]}\n",
    "snp_classified_del_del1 = filter_characters(del_rs334, character_value_dict)\n",
    "snp_classified_del1 = snp_classified_del_del1[[\"NWDID\", \"geno\"]]\n",
    "snp_classified_del1_filename = \"common_rs334_del1.tsv\"\n",
    "snp_classified_del1_dir_filename = os.path.join(save_dir, snp_classified_del1_filename)\n",
    "snp_classified_del1.to_csv(snp_classified_del1_dir_filename, sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 2 Copies of deletion, a.k.a. CN = 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "save_dir = os.path.join(\"..\", \"cohort\", cohort, \"ready_data\")\n",
    "character_value_dict = {\"NWDID\":[\"no_thres\"], \"DEL\":[2, \"eq\"], \"geno\":[\"no_thres\"]}\n",
    "snp_classified_del_del2 = filter_characters(del_rs334, character_value_dict)\n",
    "snp_classified_del2 = snp_classified_del_del2[[\"NWDID\", \"geno\"]]\n",
    "snp_classified_del2_filename = \"common_rs334_del2.tsv\"\n",
    "snp_classified_del2_dir_filename = os.path.join(save_dir, snp_classified_del2_filename)\n",
    "snp_classified_del2.to_csv(snp_classified_del2_dir_filename, sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### rs334 when DEL = 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [],
   "source": [
    "phenotype_list = [\"hemoglobin_mcnc_bld_1\", \"hematocrit_vfr_bld_1\", \"rbc_ncnc_bld_1\",\"mcv_entvol_rbc_1\",\n",
    "                  \"mch_entmass_rbc_1\", \"mchc_mcnc_rbc_1\", \"rdw_ratio_rbc_1\", \"lnHBA1C\", \"lnHBA1C_noDIABETES\",\n",
    "                  \"EGFRCKDEPI\", \"anemia\", \"microcytosis\", \"CKD\"]\n",
    "load_dir = os.path.join(\"..\", \"cohort\", cohort, \"ready_data\")\n",
    "\n",
    "pheno_full_filename = \"pheno_needed.tsv\"\n",
    "pheno_full_dir_filename = os.path.join(load_dir, pheno_full_filename)\n",
    "pheno_full_df = pd.read_csv(pheno_full_dir_filename, sep=\"\\t\")\n",
    "\n",
    "var_name = \"rs334_del0\"\n",
    "\n",
    "var_filename = \"common_rs334_del0.tsv\"\n",
    "var_dir_filename = os.path.join(load_dir, var_filename)\n",
    "var_df = pd.read_csv(var_dir_filename, sep=\"\\t\")\n",
    "\n",
    "table_name = \"table2\"\n",
    "\n",
    "kinship_filename = \"common_kinship.tsv\"\n",
    "kinship_dir_filename = os.path.join(load_dir, kinship_filename)\n",
    "kinship_df_ID = pd.read_csv(kinship_dir_filename, sep=\"\\t\")\n",
    "\n",
    "age_sex_filename = \"common_age-sex.tsv\"\n",
    "age_sex_dir_filename = os.path.join(load_dir, age_sex_filename)\n",
    "age_sex_df = pd.read_csv(age_sex_dir_filename, sep=\"\\t\")\n",
    "\n",
    "pc10_filename = \"common_10pc.tsv\"\n",
    "pc10_dir_filename = os.path.join(load_dir, pc10_filename)\n",
    "pc10_df = pd.read_csv(pc10_dir_filename, sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [],
   "source": [
    "for phenotype in phenotype_list:\n",
    "    prepare_matrix_pheno(cohort, phenotype, pheno_full_df, var_name, var_df, table_name, kinship_df_ID, age_sex_df, pc10_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### rs334 when DEL = 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [],
   "source": [
    "phenotype_list = [\"hemoglobin_mcnc_bld_1\", \"hematocrit_vfr_bld_1\", \"rbc_ncnc_bld_1\",\"mcv_entvol_rbc_1\",\n",
    "                  \"mch_entmass_rbc_1\", \"mchc_mcnc_rbc_1\", \"rdw_ratio_rbc_1\", \"lnHBA1C\", \"lnHBA1C_noDIABETES\",\n",
    "                  \"EGFRCKDEPI\", \"anemia\", \"microcytosis\", \"CKD\"]\n",
    "load_dir = os.path.join(\"..\", \"cohort\", cohort, \"ready_data\")\n",
    "\n",
    "pheno_full_filename = \"pheno_needed.tsv\"\n",
    "pheno_full_dir_filename = os.path.join(load_dir, pheno_full_filename)\n",
    "pheno_full_df = pd.read_csv(pheno_full_dir_filename, sep=\"\\t\")\n",
    "\n",
    "var_name = \"rs334_del1\"\n",
    "\n",
    "var_filename = \"common_rs334_del1.tsv\"\n",
    "var_dir_filename = os.path.join(load_dir, var_filename)\n",
    "var_df = pd.read_csv(var_dir_filename, sep=\"\\t\")\n",
    "\n",
    "table_name = \"table2\"\n",
    "\n",
    "kinship_filename = \"common_kinship.tsv\"\n",
    "kinship_dir_filename = os.path.join(load_dir, kinship_filename)\n",
    "kinship_df_ID = pd.read_csv(kinship_dir_filename, sep=\"\\t\")\n",
    "\n",
    "age_sex_filename = \"common_age-sex.tsv\"\n",
    "age_sex_dir_filename = os.path.join(load_dir, age_sex_filename)\n",
    "age_sex_df = pd.read_csv(age_sex_dir_filename, sep=\"\\t\")\n",
    "\n",
    "pc10_filename = \"common_10pc.tsv\"\n",
    "pc10_dir_filename = os.path.join(load_dir, pc10_filename)\n",
    "pc10_df = pd.read_csv(pc10_dir_filename, sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [],
   "source": [
    "for phenotype in phenotype_list:\n",
    "    prepare_matrix_pheno(cohort, phenotype, pheno_full_df, var_name, var_df, table_name, kinship_df_ID, age_sex_df, pc10_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### rs334 when DEL = 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [],
   "source": [
    "phenotype_list = [\"hemoglobin_mcnc_bld_1\", \"hematocrit_vfr_bld_1\", \"rbc_ncnc_bld_1\",\"mcv_entvol_rbc_1\",\n",
    "                  \"mch_entmass_rbc_1\", \"mchc_mcnc_rbc_1\", \"rdw_ratio_rbc_1\", \"lnHBA1C\", \"lnHBA1C_noDIABETES\",\n",
    "                  \"EGFRCKDEPI\", \"anemia\", \"microcytosis\", \"CKD\"]\n",
    "load_dir = os.path.join(\"..\", \"cohort\", cohort, \"ready_data\")\n",
    "\n",
    "pheno_full_filename = \"pheno_needed.tsv\"\n",
    "pheno_full_dir_filename = os.path.join(load_dir, pheno_full_filename)\n",
    "pheno_full_df = pd.read_csv(pheno_full_dir_filename, sep=\"\\t\")\n",
    "\n",
    "var_name = \"rs334_del2\"\n",
    "\n",
    "var_filename = \"common_rs334_del2.tsv\"\n",
    "var_dir_filename = os.path.join(load_dir, var_filename)\n",
    "var_df = pd.read_csv(var_dir_filename, sep=\"\\t\")\n",
    "\n",
    "table_name = \"table2\"\n",
    "\n",
    "kinship_filename = \"common_kinship.tsv\"\n",
    "kinship_dir_filename = os.path.join(load_dir, kinship_filename)\n",
    "kinship_df_ID = pd.read_csv(kinship_dir_filename, sep=\"\\t\")\n",
    "\n",
    "age_sex_filename = \"common_age-sex.tsv\"\n",
    "age_sex_dir_filename = os.path.join(load_dir, age_sex_filename)\n",
    "age_sex_df = pd.read_csv(age_sex_dir_filename, sep=\"\\t\")\n",
    "\n",
    "pc10_filename = \"common_10pc.tsv\"\n",
    "pc10_dir_filename = os.path.join(load_dir, pc10_filename)\n",
    "pc10_df = pd.read_csv(pc10_dir_filename, sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [],
   "source": [
    "for phenotype in phenotype_list:\n",
    "    prepare_matrix_pheno(cohort, phenotype, pheno_full_df, var_name, var_df, table_name, kinship_df_ID, age_sex_df, pc10_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Gen Gen Interaction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [],
   "source": [
    "phenotype_list = [\"hemoglobin_mcnc_bld_1\", \"hematocrit_vfr_bld_1\", \"rbc_ncnc_bld_1\",\"mcv_entvol_rbc_1\",\n",
    "                  \"mch_entmass_rbc_1\", \"mchc_mcnc_rbc_1\", \"rdw_ratio_rbc_1\", \"lnHBA1C\", \"lnHBA1C_noDIABETES\",\n",
    "                  \"EGFRCKDEPI\", \"anemia\", \"microcytosis\", \"CKD\"]\n",
    "load_dir = os.path.join(\"..\", \"cohort\", cohort, \"ready_data\")\n",
    "\n",
    "pheno_full_filename = \"pheno_needed.tsv\"\n",
    "pheno_full_dir_filename = os.path.join(load_dir, pheno_full_filename)\n",
    "pheno_full_df = pd.read_csv(pheno_full_dir_filename, sep=\"\\t\")\n",
    "\n",
    "var_name = \"del_rs334\"\n",
    "\n",
    "var_filename = \"common_del-rs334.tsv\"\n",
    "var_dir_filename = os.path.join(load_dir, var_filename)\n",
    "var_df = pd.read_csv(var_dir_filename, sep=\"\\t\")\n",
    "\n",
    "table_name = \"table2\"\n",
    "\n",
    "kinship_filename = \"common_kinship.tsv\"\n",
    "kinship_dir_filename = os.path.join(load_dir, kinship_filename)\n",
    "kinship_df_ID = pd.read_csv(kinship_dir_filename, sep=\"\\t\")\n",
    "\n",
    "age_sex_filename = \"common_age-sex.tsv\"\n",
    "age_sex_dir_filename = os.path.join(load_dir, age_sex_filename)\n",
    "age_sex_df = pd.read_csv(age_sex_dir_filename, sep=\"\\t\")\n",
    "\n",
    "pc10_filename = \"common_10pc.tsv\"\n",
    "pc10_dir_filename = os.path.join(load_dir, pc10_filename)\n",
    "pc10_df = pd.read_csv(pc10_dir_filename, sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [],
   "source": [
    "for phenotype in phenotype_list:\n",
    "    prepare_matrix_pheno(cohort, phenotype, pheno_full_df, var_name, var_df, table_name, kinship_df_ID, age_sex_df, pc10_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Table 3"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### DEL Ã— rs11248850"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [],
   "source": [
    "load_dir = os.path.join(\"..\", \"cohort\", cohort, \"ready_data\")\n",
    "del_filename = \"common_del.tsv\"\n",
    "del_dir_filename = os.path.join(load_dir, del_filename)\n",
    "del_cn = pd.read_csv(del_dir_filename, sep=\"\\t\")\n",
    "\n",
    "rs11248850_filename = \"common_rs11248850.tsv\"\n",
    "rs11248850_dir_filename = os.path.join(load_dir, rs11248850_filename)\n",
    "rs11248850 = pd.read_csv(rs11248850_dir_filename, sep=\"\\t\")\n",
    "\n",
    "del_rs11248850 = pd.concat(objs=[del_cn, rs11248850[[\"geno\"]]], axis=1)\n",
    "del_rs11248850_filename = \"common_del-rs11248850.tsv\"\n",
    "save_dir = load_dir\n",
    "del_rs11248850_dir_filename = os.path.join(save_dir, del_rs11248850_filename)\n",
    "del_rs11248850.to_csv(del_rs11248850_dir_filename, sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Classifying rs334 by # Copies of Deletion"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [],
   "source": [
    "load_dir = os.path.join(\"..\", \"cohort\", cohort, \"ready_data\")\n",
    "del_rs11248850_filename = \"common_del-rs11248850.tsv\"\n",
    "del_rs11248850_dir_filename = os.path.join(load_dir, del_rs11248850_filename)\n",
    "del_rs11248850 = pd.read_csv(del_rs11248850_dir_filename, sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### DEL = 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [],
   "source": [
    "save_dir = os.path.join(\"..\", \"cohort\", cohort, \"ready_data\")\n",
    "character_value_dict = {\"NWDID\":[\"no_thres\"], \"DEL\":[0, \"eq\"], \"geno\":[\"no_thres\"]}\n",
    "snp_classified_del_del0 = filter_characters(del_rs11248850, character_value_dict)\n",
    "snp_classified_del0 = snp_classified\n",
    "save_dir = os.path.join(\"..\", \"cohort\", cohort, \"ready_data\")\n",
    "save_dir = os.path.join(\"..\", \"cohort\", cohort, \"ready_data\")\n",
    "character_value_dict = {\"NWDID\":[\"no_thres\"], \"DEL\":[0, \"eq\"], \"geno\":[\"no_thres\"]}\n",
    "snp_classified_del_del0 = filter_characters(del_rs11248850, character_value_dict)\n",
    "snp_classified_del0 = snp_classified_del_del0[[\"NWDID\", \"geno\"]]\n",
    "snp_classified_del0_filename = \"common_rs11248850_del0.tsv\"\n",
    "snp_classified_del0_dir_filename = os.path.join(save_dir, snp_classified_del0_filename)\n",
    "snp_classified_del0.to_csv(snp_classified_del0_dir_filename, sep=\"\\t\", index=False)\n",
    "_del_del0[[\"NWDID\", \"geno\"]]\n",
    "snp_classified_del0_filename = \"common_rs11248850_del0.tsv\"\n",
    "snp_classified_del0_dir_filename = os.path.join(save_dir, snp_classified_del0_filename)\n",
    "snp_classified_del0.to_csv(snp_classified_del0_dir_filename, sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### DEL = 1, 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [],
   "source": [
    "save_dir = os.path.join(\"..\", \"cohort\", cohort, \"ready_data\")\n",
    "character_value_dict = {\"NWDID\":[\"no_thres\"], \"DEL\":[1, \"eq\"], \"geno\":[\"no_thres\"]}\n",
    "snp_classified_del_del1 = filter_characters(del_rs11248850, character_value_dict)\n",
    "snp_classified_del1 = snp_classified_del_del1[[\"NWDID\", \"geno\"]]\n",
    "\n",
    "character_value_dict = {\"NWDID\":[\"no_thres\"], \"DEL\":[2, \"eq\"], \"geno\":[\"no_thres\"]}\n",
    "snp_classified_del_del2 = filter_characters(del_rs11248850, character_value_dict)\n",
    "snp_classified_del2 = snp_classified_del_del2[[\"NWDID\", \"geno\"]]\n",
    "\n",
    "snp_classified_del12 = pd.concat(objs=[snp_classified_del1, snp_classified_del2], axis=0)\n",
    "\n",
    "snp_classified_del12_filename = \"common_rs11248850_del12.tsv\"\n",
    "snp_classified_del12_dir_filename = os.path.join(save_dir, snp_classified_del12_filename)\n",
    "snp_classified_del12.to_csv(snp_classified_del12_dir_filename, sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### rs11248850 when DEL = 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [],
   "source": [
    "phenotype_list_table3 = [\"hemoglobin_mcnc_bld_1\", \"hematocrit_vfr_bld_1\", \"rbc_ncnc_bld_1\",\"mcv_entvol_rbc_1\",\n",
    "                  \"mch_entmass_rbc_1\", \"mchc_mcnc_rbc_1\", \"rdw_ratio_rbc_1\"]\n",
    "load_dir = os.path.join(\"..\", \"cohort\", cohort, \"ready_data\")\n",
    "\n",
    "pheno_full_filename = \"pheno_needed.tsv\"\n",
    "pheno_full_dir_filename = os.path.join(load_dir, pheno_full_filename)\n",
    "pheno_full_df = pd.read_csv(pheno_full_dir_filename, sep=\"\\t\")\n",
    "\n",
    "var_name = \"rs11248850_del0\"\n",
    "\n",
    "var_filename = \"common_rs11248850_del0.tsv\"\n",
    "var_dir_filename = os.path.join(load_dir, var_filename)\n",
    "var_df = pd.read_csv(var_dir_filename, sep=\"\\t\")\n",
    "\n",
    "table_name = \"table3\"\n",
    "\n",
    "kinship_filename = \"common_kinship.tsv\"\n",
    "kinship_dir_filename = os.path.join(load_dir, kinship_filename)\n",
    "kinship_df_ID = pd.read_csv(kinship_dir_filename, sep=\"\\t\")\n",
    "\n",
    "age_sex_filename = \"common_age-sex.tsv\"\n",
    "age_sex_dir_filename = os.path.join(load_dir, age_sex_filename)\n",
    "age_sex_df = pd.read_csv(age_sex_dir_filename, sep=\"\\t\")\n",
    "\n",
    "pc10_filename = \"common_10pc.tsv\"\n",
    "pc10_dir_filename = os.path.join(load_dir, pc10_filename)\n",
    "pc10_df = pd.read_csv(pc10_dir_filename, sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [],
   "source": [
    "for phenotype in phenotype_list_table3:\n",
    "    prepare_matrix_pheno(cohort, phenotype, pheno_full_df, var_name, var_df, table_name, kinship_df_ID, age_sex_df, pc10_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### rs11248850 when DEL = 1, 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [],
   "source": [
    "phenotype_list_table3 = [\"hemoglobin_mcnc_bld_1\", \"hematocrit_vfr_bld_1\", \"rbc_ncnc_bld_1\",\"mcv_entvol_rbc_1\",\n",
    "                  \"mch_entmass_rbc_1\", \"mchc_mcnc_rbc_1\", \"rdw_ratio_rbc_1\"]\n",
    "load_dir = os.path.join(\"..\", \"cohort\", cohort, \"ready_data\")\n",
    "\n",
    "pheno_full_filename = \"pheno_needed.tsv\"\n",
    "pheno_full_dir_filename = os.path.join(load_dir, pheno_full_filename)\n",
    "pheno_full_df = pd.read_csv(pheno_full_dir_filename, sep=\"\\t\")\n",
    "\n",
    "var_name = \"rs11248850_del12\"\n",
    "\n",
    "var_filename = \"common_rs11248850_del12.tsv\"\n",
    "var_dir_filename = os.path.join(load_dir, var_filename)\n",
    "var_df = pd.read_csv(var_dir_filename, sep=\"\\t\")\n",
    "\n",
    "table_name = \"table3\"\n",
    "\n",
    "kinship_filename = \"common_kinship.tsv\"\n",
    "kinship_dir_filename = os.path.join(load_dir, kinship_filename)\n",
    "kinship_df_ID = pd.read_csv(kinship_dir_filename, sep=\"\\t\")\n",
    "\n",
    "age_sex_filename = \"common_age-sex.tsv\"\n",
    "age_sex_dir_filename = os.path.join(load_dir, age_sex_filename)\n",
    "age_sex_df = pd.read_csv(age_sex_dir_filename, sep=\"\\t\")\n",
    "\n",
    "pc10_filename = \"common_10pc.tsv\"\n",
    "pc10_dir_filename = os.path.join(load_dir, pc10_filename)\n",
    "pc10_df = pd.read_csv(pc10_dir_filename, sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [],
   "source": [
    "for phenotype in phenotype_list_table3:\n",
    "    prepare_matrix_pheno(cohort, phenotype, pheno_full_df, var_name, var_df, table_name, kinship_df_ID, age_sex_df, pc10_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### rs11248850"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [],
   "source": [
    "phenotype_list_table3 = [\"hemoglobin_mcnc_bld_1\", \"hematocrit_vfr_bld_1\", \"rbc_ncnc_bld_1\",\"mcv_entvol_rbc_1\",\n",
    "                  \"mch_entmass_rbc_1\", \"mchc_mcnc_rbc_1\", \"rdw_ratio_rbc_1\"]\n",
    "load_dir = os.path.join(\"..\", \"cohort\", cohort, \"ready_data\")\n",
    "\n",
    "pheno_full_filename = \"pheno_needed.tsv\"\n",
    "pheno_full_dir_filename = os.path.join(load_dir, pheno_full_filename)\n",
    "pheno_full_df = pd.read_csv(pheno_full_dir_filename, sep=\"\\t\")\n",
    "\n",
    "var_name = \"rs11248850\"\n",
    "\n",
    "var_filename = \"common_rs11248850.tsv\"\n",
    "var_dir_filename = os.path.join(load_dir, var_filename)\n",
    "var_df = pd.read_csv(var_dir_filename, sep=\"\\t\")\n",
    "\n",
    "table_name = \"table3\"\n",
    "\n",
    "kinship_filename = \"common_kinship.tsv\"\n",
    "kinship_dir_filename = os.path.join(load_dir, kinship_filename)\n",
    "kinship_df_ID = pd.read_csv(kinship_dir_filename, sep=\"\\t\")\n",
    "\n",
    "age_sex_filename = \"common_age-sex.tsv\"\n",
    "age_sex_dir_filename = os.path.join(load_dir, age_sex_filename)\n",
    "age_sex_df = pd.read_csv(age_sex_dir_filename, sep=\"\\t\")\n",
    "\n",
    "pc10_filename = \"common_10pc.tsv\"\n",
    "pc10_dir_filename = os.path.join(load_dir, pc10_filename)\n",
    "pc10_df = pd.read_csv(pc10_dir_filename, sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [],
   "source": [
    "for phenotype in phenotype_list_table3:\n",
    "    prepare_matrix_pheno(cohort, phenotype, pheno_full_df, var_name, var_df, table_name, kinship_df_ID, age_sex_df, pc10_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Gen Gen Interaction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [],
   "source": [
    "phenotype_list_table3 = [\"hemoglobin_mcnc_bld_1\", \"hematocrit_vfr_bld_1\", \"rbc_ncnc_bld_1\",\n",
    "                         \"mcv_entvol_rbc_1\", \"mch_entmass_rbc_1\", \"mchc_mcnc_rbc_1\", \"rdw_ratio_rbc_1\"]\n",
    "load_dir = os.path.join(\"..\", \"cohort\", cohort, \"ready_data\")\n",
    "\n",
    "pheno_full_filename = \"pheno_needed.tsv\"\n",
    "pheno_full_dir_filename = os.path.join(load_dir, pheno_full_filename)\n",
    "pheno_full_df = pd.read_csv(pheno_full_dir_filename, sep=\"\\t\")\n",
    "\n",
    "var_name = \"del_rs11248850\"\n",
    "\n",
    "var_filename = \"common_del-rs11248850.tsv\"\n",
    "var_dir_filename = os.path.join(load_dir, var_filename)\n",
    "var_df = pd.read_csv(var_dir_filename, sep=\"\\t\")\n",
    "\n",
    "table_name = \"table3\"\n",
    "\n",
    "kinship_filename = \"common_kinship.tsv\"\n",
    "kinship_dir_filename = os.path.join(load_dir, kinship_filename)\n",
    "kinship_df_ID = pd.read_csv(kinship_dir_filename, sep=\"\\t\")\n",
    "\n",
    "age_sex_filename = \"common_age-sex.tsv\"\n",
    "age_sex_dir_filename = os.path.join(load_dir, age_sex_filename)\n",
    "age_sex_df = pd.read_csv(age_sex_dir_filename, sep=\"\\t\")\n",
    "\n",
    "pc10_filename = \"common_10pc.tsv\"\n",
    "pc10_dir_filename = os.path.join(load_dir, pc10_filename)\n",
    "pc10_df = pd.read_csv(pc10_dir_filename, sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [],
   "source": [
    "for phenotype in phenotype_list_table3:\n",
    "    prepare_matrix_pheno(cohort, phenotype, pheno_full_df, var_name, var_df, table_name, kinship_df_ID, age_sex_df, pc10_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Freeze6a Cohorts"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First, loading kinship matrix for later usage."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "predata_dir = os.path.join(\"..\", \"prepro_data\", \"kinship\")\n",
    "freeze6a_kinship_filename = \"freeze6a_kinship.feather\"\n",
    "freeze6a_kinship_dir_filename = os.path.join(predata_dir, freeze6a_kinship_filename)\n",
    "freeze6a_kinship_df = feather.read_dataframe(freeze6a_kinship_dir_filename)\n",
    "freeze6a_sample_list = list(freeze6a_kinship_df)\n",
    "freeze6a_sample_df = pd.DataFrame(data=freeze6a_sample_list, columns=[\"NWDID\"])\n",
    "freeze6a_kinship_df_ID = pd.concat(objs=[freeze6a_sample_df, freeze6a_kinship_df], axis=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Freeze8 Cohorts"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Loading Kinship Matrix for Later Usage"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 1min 30s, sys: 4min 55s, total: 6min 25s\n",
      "Wall time: 6min 42s\n"
     ]
    }
   ],
   "source": [
    "predata_dir = os.path.join(\"..\", \"prepro_data\", \"kinship\")\n",
    "freeze8_kinship_filename = \"freeze8_kinship.feather\"\n",
    "freeze8_kinship_dir_filename = os.path.join(predata_dir, freeze8_kinship_filename)\n",
    "%time freeze8_kinship_df = pd.read_feather(freeze8_kinship_dir_filename, use_threads = True)\n",
    "freeze8_sample_list = list(freeze8_kinship_df)\n",
    "freeze8_sample_df = pd.DataFrame(data=freeze8_sample_list, columns=[\"NWDID\"])\n",
    "freeze8_kinship_df_ID = pd.concat(objs=[freeze8_sample_df, freeze8_kinship_df], axis=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Apply MESA (from Methylation) Samples to Kinship Matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1890, 21)\n",
      "(1890, 1891)\n"
     ]
    }
   ],
   "source": [
    "save_dir = \"/pine/scr/m/i/minzhi/mesa_multiomics/methylation_normalized\"\n",
    "sample_filename = \"TOPMed_MESA.methylomics.samplesheet_with_feno.mixup_fix.only_BIS.pass_QC.txt\"\n",
    "sample_dir_filename = os.path.join(save_dir, sample_filename)\n",
    "sample_df_raw = pd.read_csv(sample_dir_filename, sep=\"\\t\")\n",
    "print(sample_df_raw.shape)\n",
    "sample_df = sample_df_raw[[\"NWDID\"]]\n",
    "kinship_sample_selected = kinship_select_sample(freeze8_kinship_df_ID, sample_df)\n",
    "print(kinship_sample_selected.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(936, 937)\n"
     ]
    }
   ],
   "source": [
    "sample_df_unique = sample_df.drop_duplicates(subset =\"NWDID\", keep = \"first\", inplace = False)\n",
    "kinship_sample_unique_selected = kinship_select_sample(freeze8_kinship_df_ID, sample_df_unique)\n",
    "print(kinship_sample_unique_selected.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(906, 1)\n"
     ]
    }
   ],
   "source": [
    "unrelated_dir = \"/pine/scr/m/i/minzhi/mesa_multiomics/methylation_normalized\"\n",
    "unrelated_filename = \"TOPMed_MESA.1331_samples_unrelated.txt\"\n",
    "unrelated_dir_filename = os.path.join(unrelated_dir, unrelated_filename)\n",
    "unrelated = pd.read_csv(unrelated_dir_filename, sep = \"\\t\", header = None, index_col = None)\n",
    "unrelated.rename(columns = {0:\"NWDID\"}, inplace = True)\n",
    "overlap_unrelated_methy = sample_df_unique.merge(unrelated, on = \"NWDID\", how = \"inner\")\n",
    "print(overlap_unrelated_methy.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "common_kinship_filename = \"mesa_methy_kinship_unique.tsv\"\n",
    "common_kinship_dir_filename = os.path.join(save_dir, common_kinship_filename)\n",
    "kinship_sample_unique_selected.to_csv(common_kinship_dir_filename, sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "common_kinship_filename = \"mesa_methy_kinship_unique.tsv\"\n",
    "common_kinship_dir_filename = os.path.join(save_dir, common_kinship_filename)\n",
    "kinship_sample_unique_selected_IDidx = pd.read_csv(common_kinship_dir_filename, sep=\"\\t\", index_col=0, header = 0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "def kinship_filtering(kinship_matrix, thres):\n",
    "\toriginal_list = list(kinship_matrix.columns)\n",
    "\tsample_num = len(original_list)\n",
    "\tcorrelated_sample_index = []\n",
    "\tuncorrelation_sample_list = []\n",
    "\tfor sample_i in range(1, sample_num):\n",
    "\t\tfor ortho_sample_i in range(sample_i):\n",
    "\t\t\tif kinship_matrix.iloc[sample_i, ortho_sample_i] > thres:\n",
    "\t\t\t\tcorrelated_sample_index.append(ortho_sample_i)\n",
    "\tfor sample_i in range(sample_num):\n",
    "\t\tif sample_i not in correlated_sample_index:\n",
    "\t\t\tuncorrelation_sample_list.append(original_list[sample_i])\n",
    "\treturn uncorrelation_sample_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "936\n",
      "925\n",
      "917\n",
      "912\n",
      "910\n",
      "905\n",
      "629\n"
     ]
    }
   ],
   "source": [
    "thres_list = [1, 0.2, 0.1, 0.05, 0.03125, 0.025, 0.005]\n",
    "for thres in thres_list:\n",
    "    uncorrelation_sample_list = kinship_filtering(kinship_sample_unique_selected_IDidx, thres)\n",
    "    print(len(uncorrelation_sample_list))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYsAAAD8CAYAAACGsIhGAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDMuMC4yLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvOIA7rQAAEWFJREFUeJzt3H+s3XV9x/Hny1aUORGEQkhbV5x1sxoncodNTBYVA4VtlmWwlGyjGrZmDDOXmUycS8hAMt0S2cjQjQmhmDlgbIZOi11FiHGRHxdBsDDGFZ3cQKBaRBwRh773x/0Uj+W2n3N/0HMsz0dycr/f9/fz/d73h1P66vfHOakqJEnalxeMugFJ0vgzLCRJXYaFJKnLsJAkdRkWkqQuw0KS1DVUWCT5RpK7k9yZZLLVXp5ke5L728/DWj1JLk4yleSuJG8cOM7GNv7+JBsH6se140+1fbPYE5Ukzd9czizeWlVvqKqJtn4ucENVrQZuaOsAJwOr22sT8DGYCRfgPOBNwPHAebsDpo3ZNLDfunnPSJK06BZyGWo9sLktbwZOHahfWTNuBg5NcjRwErC9qnZV1WPAdmBd23ZIVX2pZj4heOXAsSRJY2DpkOMK+I8kBfxDVV0KHFVVDwNU1cNJjmxjlwMPDuw73Wr7qk/PUt+nI444olatWjVk+5Kk22+//VtVtWw++w4bFm+uqodaIGxP8l/7GDvb/YaaR/3ZB042MXO5ile84hVMTk7uu2tJ0jOS/M989x3qMlRVPdR+Pgp8ipl7Do+0S0i0n4+24dPAyoHdVwAPdeorZqnP1selVTVRVRPLls0rHCVJ89ANiyQvSfLS3cvAicBXgS3A7ieaNgLXteUtwJntqai1wOPtctU24MQkh7Ub2ycC29q2J5KsbU9BnTlwLEnSGBjmMtRRwKfa06xLgU9W1WeT3AZck+Qs4JvA6W38VuAUYAp4EngXQFXtSnIBcFsbd35V7WrLZwNXAAcD17eXJGlM5Kf1K8onJibKexaSNLwktw98/GFO/AS3JKnLsJAkdRkWkqQuw0KS1GVYSJK6DAtJUpdhIUnqMiwkSV2GhSSpy7CQJHUZFpKkLsNCktRlWEiSugwLSVKXYSFJ6jIsJEldhoUkqcuwkCR1GRaSpC7DQpLUZVhIkroMC0lSl2EhSeoyLCRJXYaFJKnLsJAkdRkWkqQuw0KS1GVYSJK6DAtJUpdhIUnqMiwkSV2GhSSpy7CQJHUZFpKkrqHDIsmSJHck+XRbPybJLUnuT3J1koNa/UVtfaptXzVwjPe3+n1JThqor2u1qSTnLt70JEmLYS5nFu8B7h1Y/zBwUVWtBh4Dzmr1s4DHqupVwEVtHEnWABuA1wLrgI+2AFoCXAKcDKwBzmhjJUljYqiwSLIC+FXg4209wNuAa9uQzcCpbXl9W6dtP6GNXw9cVVVPVdXXgSng+PaaqqoHquoHwFVt7H616tzP7O9fKUk/NYY9s/gb4E+BH7X1w4HvVNXTbX0aWN6WlwMPArTtj7fxz9T32Gdv9WdJsinJZJLJnTt3Dtm6JGmhumGR5NeAR6vq9sHyLEOrs22u9WcXqy6tqomqmli2bNk+upYkLaalQ4x5M/COJKcALwYOYeZM49AkS9vZwwrgoTZ+GlgJTCdZCrwM2DVQ321wn73VJUljoHtmUVXvr6oVVbWKmRvUn6+q3wZuBE5rwzYC17XlLW2dtv3zVVWtvqE9LXUMsBq4FbgNWN2erjqo/Y4tizI7SdKiGObMYm/eB1yV5IPAHcBlrX4Z8IkkU8ycUWwAqKodSa4B7gGeBs6pqh8CJHk3sA1YAlxeVTsW0JckaZHNKSyq6ibgprb8ADNPMu055vvA6XvZ/0LgwlnqW4Gtc+lFkrT/+AluSVKXYYGfsZCkHsNCktRlWEiSugwLSVKXYSFJ6jIsJEldhoUkqcuwkCR1GRaSpC7DQpLUZVhIkroMC0lSl2EhSeoyLCRJXYaFJKnLsJAkdRkWkqQuw0KS1GVYSJK6DAtJUpdhIUnqMiwkSV2GhSSpy7CQJHUZFpKkLsNCktRlWEiSugwLSVKXYSFJ6jIsJEldhoUkqcuwkCR1GRaSpK5uWCR5cZJbk3wlyY4kf9HqxyS5Jcn9Sa5OclCrv6itT7XtqwaO9f5Wvy/JSQP1da02leTcxZ+mJGkhhjmzeAp4W1X9EvAGYF2StcCHgYuqajXwGHBWG38W8FhVvQq4qI0jyRpgA/BaYB3w0SRLkiwBLgFOBtYAZ7SxkqQx0Q2LmvG9tvrC9irgbcC1rb4ZOLUtr2/rtO0nJEmrX1VVT1XV14Ep4Pj2mqqqB6rqB8BVbawkaUwMdc+inQHcCTwKbAe+Bnynqp5uQ6aB5W15OfAgQNv+OHD4YH2PffZWlySNiaHCoqp+WFVvAFYwcybwmtmGtZ/Zy7a51p8lyaYkk0kmd+7c2W9ckrQo5vQ0VFV9B7gJWAscmmRp27QCeKgtTwMrAdr2lwG7But77LO3+my//9KqmqiqiWXLls2ldUnSAgzzNNSyJIe25YOBtwP3AjcCp7VhG4Hr2vKWtk7b/vmqqlbf0J6WOgZYDdwK3Aasbk9XHcTMTfAtizE5SdLiWNofwtHA5vbU0guAa6rq00nuAa5K8kHgDuCyNv4y4BNJppg5o9gAUFU7klwD3AM8DZxTVT8ESPJuYBuwBLi8qnYs2gwlSQvWDYuqugs4dpb6A8zcv9iz/n3g9L0c60LgwlnqW4GtQ/QrSRoBP8EtSeoyLCRJXYaFJKnLsJAkdRkWkqQuw0KS1GVYSJK6DAtJUpdhIUnqMiwkSV2GhSSpy7CQJHUZFpKkLsNCktRlWEiSugwLSVKXYSFJ6jIsJEldhoUkqcuwkCR1GRaSpC7DQpLUZVhIkroMC0lSl2EhSeoyLCRJXYaFJKnLsJAkdRkWkqQuw0KS1GVYSJK6DAtJUpdhIUnqMiwkSV3dsEiyMsmNSe5NsiPJe1r95Um2J7m//Tys1ZPk4iRTSe5K8saBY21s4+9PsnGgflySu9s+FyfJczFZSdL8DHNm8TTw3qp6DbAWOCfJGuBc4IaqWg3c0NYBTgZWt9cm4GMwEy7AecCbgOOB83YHTBuzaWC/dQufmiRpsXTDoqoerqovt+UngHuB5cB6YHMbthk4tS2vB66sGTcDhyY5GjgJ2F5Vu6rqMWA7sK5tO6SqvlRVBVw5cCxJ0hiY0z2LJKuAY4FbgKOq6mGYCRTgyDZsOfDgwG7Trbav+vQsdUnSmBg6LJL8LPCvwB9X1Xf3NXSWWs2jPlsPm5JMJpncuXNnr2VJ0iIZKiySvJCZoPinqvq3Vn6kXUKi/Xy01aeBlQO7rwAe6tRXzFJ/lqq6tKomqmpi2bJlw7QuSVoEwzwNFeAy4N6q+sjApi3A7ieaNgLXDdTPbE9FrQUeb5eptgEnJjms3dg+EdjWtj2RZG37XWcOHEuSNAaWDjHmzcDvAncnubPV/gz4EHBNkrOAbwKnt21bgVOAKeBJ4F0AVbUryQXAbW3c+VW1qy2fDVwBHAxc316SpDHRDYuq+iKz31cAOGGW8QWcs5djXQ5cPkt9EnhdrxdJ0mj4CW5JUpdhIUnqMiwkSV2GhSSpy7CQJHUZFpKkLsNCktRlWEiSugwLSVKXYSFJ6jIsJEldhoUkqcuwkCR1GRaSpC7DQpLUZVhIkroMC0lSl2EhSeoyLCRJXYaFJKnLsJAkdRkWkqQuw0KS1GVYSJK6DAtJUpdhIUnqMiwkSV2GhSSpy7CQJHUZFpKkLsNCktRlWEiSugwLSVKXYSFJ6jIsJEld3bBIcnmSR5N8daD28iTbk9zffh7W6klycZKpJHcleePAPhvb+PuTbByoH5fk7rbPxUmy2JOUJC3MMGcWVwDr9qidC9xQVauBG9o6wMnA6vbaBHwMZsIFOA94E3A8cN7ugGljNg3st+fvkiSNWDcsquoLwK49yuuBzW15M3DqQP3KmnEzcGiSo4GTgO1VtauqHgO2A+vatkOq6ktVVcCVA8eSJI2J+d6zOKqqHgZoP49s9eXAgwPjplttX/XpWeqzSrIpyWSSyZ07d86zdUnSXC32De7Z7jfUPOqzqqpLq2qiqiaWLVs2zxYlSXM137B4pF1Cov18tNWngZUD41YAD3XqK2apS5LGyHzDYguw+4mmjcB1A/Uz21NRa4HH22WqbcCJSQ5rN7ZPBLa1bU8kWduegjpz4FiSpDGxtDcgyT8DbwGOSDLNzFNNHwKuSXIW8E3g9DZ8K3AKMAU8CbwLoKp2JbkAuK2NO7+qdt80P5uZJ64OBq5vL0nSGOmGRVWdsZdNJ8wytoBz9nKcy4HLZ6lPAq/r9SFJGh0/wS1J6jIsJEldhoUkqcuwGLDq3M+MugVJGkuGhSSpy7CQJHUZFpKkLsNCktRlWEiSugwLSVKXYSFJ6jIsJEldhoUkqcuwkCR1GRaSpC7DQpLUZVhIkroMC0lSl2EhSeoyLCRJXYaFJKnLsJAkdRkWkqQuw0KS1GVYSJK6DAtJUpdhIUnqMiwkSV2GhSSpy7CQJHUZFpKkLsNCktRlWEiSugwLSVLX2IRFknVJ7ksyleTcUfcjSfqxsQiLJEuAS4CTgTXAGUnWjLYrSdJuYxEWwPHAVFU9UFU/AK4C1o+4J0lSMy5hsRx4cGB9utUkSWNg6agbaDJLrZ41KNkEbGqr30ty3yL2cATwrXx4EY84WkcA3xp1E4vsQJuT8xl/B9qcfmG+O45LWEwDKwfWVwAP7Tmoqi4FLn0uGkgyWVUTz8WxR+FAmw8ceHNyPuPvQJtTksn57jsul6FuA1YnOSbJQcAGYMuIe5IkNWNxZlFVTyd5N7ANWAJcXlU7RtyWJKkZi7AAqKqtwNYRtvCcXN4aoQNtPnDgzcn5jL8DbU7znk+qnnUfWZKknzAu9ywkSWPseRcWva8VSfKiJFe37bckWbX/uxzeEPP5lSRfTvJ0ktNG0eNcDDGfP0lyT5K7ktyQ5OdG0edcDDGnP0hyd5I7k3xx3L+9YNiv5klyWpJKMtZPEw3x/rwzyc72/tyZ5PdG0edcDPMeJfmt9v/SjiSf7B60qp43L2Zunn8NeCVwEPAVYM0eY/4Q+Pu2vAG4etR9L3A+q4DXA1cCp42650WYz1uBn2nLZ4/z+zOHOR0ysPwO4LOj7nsh82njXgp8AbgZmBh13wt8f94J/N2oe13kOa0G7gAOa+tH9o77fDuzGOZrRdYDm9vytcAJSWb70OA46M6nqr5RVXcBPxpFg3M0zHxurKon2+rNzHwmZ5wNM6fvDqy+hFk+kDpGhv1qnguAvwK+vz+bm4cD8auGhpnT7wOXVNVjAFX1aO+gz7ewGOZrRZ4ZU1VPA48Dh++X7ubuQPualLnO5yzg+ue0o4Ubak5JzknyNWb+gv2j/dTbfHTnk+RYYGVVfXp/NjZPw/6Z+8126fPaJCtn2T5OhpnTq4FXJ/nPJDcnWdc76PMtLIb5WpGhvnpkTPw09TqMoeeT5HeACeCvn9OOFm6oOVXVJVX188D7gD9/zruav33OJ8kLgIuA9+63jhZmmPfn34FVVfV64HP8+MrDuBpmTkuZuRT1FuAM4ONJDt3XQZ9vYTHM14o8MybJUuBlwK790t3cDfU1KT9FhppPkrcDHwDeUVVP7afe5muu79FVwKnPaUcL05vPS4HXATcl+QawFtgyxje5u+9PVX174M/ZPwLH7afe5mvYv+euq6r/q6qvA/cxEx57N+qbMfv5xs9S4AHgGH584+e1e4w5h5+8wX3NqPteyHwGxl7B+N/gHub9OZaZm3erR93vIs5p9cDyrwOTo+57IfPZY/xNjPcN7mHen6MHln8DuHnUfS/CnNYBm9vyEcxctjp8n8cd9cRG8B/yFOC/2184H2i185n5VyrAi4F/AaaAW4FXjrrnBc7nl5n5V8T/At8Gdoy65wXO53PAI8Cd7bVl1D0vwpz+FtjR5nPjvv7yHYdXbz57jB3rsBjy/fnL9v58pb0/vzjqnhdhTgE+AtwD3A1s6B3TT3BLkrqeb/csJEnzYFhIkroMC0lSl2EhSeoyLCRJXYaFJKnLsJAkdRkWkqSu/wcXzoRlMznuhgAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "kinship_sample_selected_noID = kinship_sample_selected.drop(axis = 1, labels = \"NWDID\")\n",
    "kinship_sample_selected_noID_array = kinship_sample_selected_noID.values\n",
    "inds = np.tril_indices(kinship_sample_selected_noID_array.shape[0], -1) \n",
    "vals = kinship_sample_selected_noID_array[inds]\n",
    "_ = plt.hist(vals, bins='auto')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1785105,)\n",
      "1784083\n",
      "1784041\n",
      "1784015\n",
      "1783993\n",
      "1783967\n",
      "1781725\n"
     ]
    }
   ],
   "source": [
    "thres_list = [0.2, 0.1, 0.05, 0.03125, 0.025, 0.005]\n",
    "print(vals.shape)\n",
    "for thres in thres_list:\n",
    "    tmp_vals = vals[vals <= thres]\n",
    "    tmp_sample_num = len(tmp_vals)\n",
    "    print(tmp_sample_num)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Then loading original phenotype file, annotation file, and eGFR file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "raw_dir = os.path.join(\"..\", \"raw_data\", \"phenotype\")\n",
    "pheno_code = \"coh01\"\n",
    "anno_ver = \"freeze8_sample_annot_2019-05-30\"\n",
    "anno_filt_code = \"af01\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "egfr_filename = \"HCHS_TOPMED_Kidney_PHEN_05-23-2019MY_v1_egfr.tsv\"\n",
    "egfr_dir = os.path.join(\"..\", \"raw_data\", \"egfr\")\n",
    "egfr_dir_filename = os.path.join(egfr_dir, egfr_filename)\n",
    "egfr_df_raw = pd.read_csv(egfr_dir_filename, sep = \"\\t\", header = 0, index_col = None)\n",
    "egfr_df_raw.rename(columns = {\"id\":\"SUBJECT_ID\", \"age\":\"age_at_EGFRCKDEPI\", \"EGFR\":\"EGFRCKDEPI\"}, inplace = True)\n",
    "egfr_df = egfr_df_raw[[\"SUBJECT_ID\", \"EGFRCKDEPI\", \"age_at_EGFRCKDEPI\"]]\n",
    "\n",
    "pheno_filename = \"%s_pre.tsv\"%pheno_code\n",
    "pheno_dir_filename = os.path.join(raw_dir, pheno_filename)\n",
    "pheno_df = pd.read_csv(pheno_dir_filename, sep = \"\\t\", header = 0, index_col = None, dtype={0: str})\n",
    "\n",
    "anno_filename = \"%s_%s.tsv\"%(anno_ver, anno_filt_code)\n",
    "anno_dir_filename = os.path.join(raw_dir, anno_filename)\n",
    "anno_df = pd.read_csv(anno_dir_filename, sep = \"\\t\", header = 0, index_col = None)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### SNPs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "def formulate_snp_rawfile(snp):\n",
    "    snp_select = snp.iloc[:, [0, 6]]\n",
    "    snp_size = snp_select.shape\n",
    "    snp_mod = pd.DataFrame(data = np.zeros(snp_size), columns = [\"NWDID\", \"geno\"])\n",
    "    snp_mod = snp_mod_raw.astype({\"NWDID\":str, \"geno\":int})\n",
    "    snp_mod[\"NWDID\"] = snp_select.iloc[:, 0]\n",
    "    snp_mod[\"geno\"] = snp_select.iloc[:, 1]\n",
    "    snp_mod = snp_mod[snp_mod[\"geno\"] != 2]\n",
    "    snp_mod = snp_mod.dropna(axis = 0, how = \"any\")\n",
    "    return snp_mod"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "snp_raw_filename = \"freeze8_rs334_pass_and_fail.raw\"\n",
    "raw_dir = os.path.join(\"..\", \"raw_data\", \"snp\")\n",
    "snp_raw_dir_filename = os.path.join(raw_dir, snp_raw_filename)\n",
    "snp_raw = pd.read_csv(snp_raw_dir_filename, sep = \" \", header = 0, index_col = None)\n",
    "snp = formulate_snp_rawfile(snp_raw)\n",
    "snp_filename = \"freeze8_rs334_hetero.tsv\"\n",
    "save_dir = os.path.join(\"..\", \"prepro_data\", \"snp\")\n",
    "snp_dir_filename = os.path.join(save_dir, snp_filename)\n",
    "snp.to_csv(snp_dir_filename, sep = \"\\t\", header = True, index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "snp_name_list = [\"rs33930165\", \"rs11248850\"]\n",
    "for snp_name in snp_name_list:\n",
    "    snp_raw_filename = \"freeze8_%s.raw\"%snp_name\n",
    "    raw_dir = os.path.join(\"..\", \"raw_data\", \"snp\")\n",
    "    snp_raw_dir_filename = os.path.join(raw_dir, snp_raw_filename)\n",
    "    snp_raw = pd.read_csv(snp_raw_dir_filename, sep = \" \", header = 0, index_col = None)\n",
    "    snp = formulate_snp_rawfile(snp_raw)\n",
    "    snp_filename = \"freeze8_%s_hetero.tsv\"%snp_name\n",
    "    save_dir = os.path.join(\"..\", \"prepro_data\", \"snp\")\n",
    "    snp_dir_filename = os.path.join(save_dir, snp_filename)\n",
    "    snp.to_csv(snp_dir_filename, sep = \"\\t\", header = True, index = False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### PC11"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {},
   "outputs": [],
   "source": [
    "pc_dir = os.path.join(\"..\", \"prepro_data\", \"pc\")\n",
    "pc_filename = \"freeze8_pc11_pcair_noheader.tsv\"\n",
    "pc_dir_filename = os.path.join(pc_dir, pc_filename)\n",
    "pc_raw = pd.read_csv(pc_dir_filename, sep = \"\\t\", header = None, index_col = None)\n",
    "col_num = pc_raw.shape[1]\n",
    "pc_raw.rename(columns = {0:\"NWDID\"}, inplace = True)\n",
    "for col in range(1, col_num):\n",
    "    pc_raw.rename(columns = {col:\"pc%d\"%col}, inplace = True)\n",
    "pc_raw_filename = \"freeze8_pc11_pcair.tsv\"\n",
    "pc_raw_dir_filename = os.path.join(pc_dir, pc_raw_filename)\n",
    "pc_raw.to_csv(pc_raw_dir_filename, sep = \"\\t\", header = True, index = False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Association between Blood Cell Trait and Freeze6a-AA-noJHS"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Preprocess Data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Phenotype"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "(**Deprecated**) Modifying the count of BioMe white blood cell subtypes (neutrophil, lymphocyte, basophil, eosinophil, and monocyte)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "biome_raw_filename = \"BioMe_raw.txt\"\n",
    "pheno_load_dir = os.path.join(\"..\", \"raw_data\", \"phenotype\")\n",
    "biome_raw_dir_filename = os.path.join(pheno_load_dir, biome_raw_filename)\n",
    "biome_raw_df = pd.read_csv(biome_raw_dir_filename, sep=\"\\t\", header=0, index_col = None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "subtype_list = [\"neutrophil\", \"lymphocyte\", \"basophil\", \"eosinophil\", \"monocyte\"]\n",
    "subtype_upper_list = [subtype_i.upper() for subtype_i in subtype_list]\n",
    "subtype_percent_list = [\"%s_PERCENT\"%subtype_upper_i for subtype_upper_i in subtype_upper_list]\n",
    "subtype_count_list = [\"%s_COUNT\"%subtype_upper_i for subtype_upper_i in subtype_upper_list]\n",
    "subtype_num = len(subtype_list)\n",
    "biome_modify_df = biome_raw_df.copy()\n",
    "for subtype_idx in range(subtype_num):\n",
    "    subtype_percent_i = subtype_percent_list[subtype_idx]\n",
    "    subtype_count_i = subtype_count_list[subtype_idx]\n",
    "    biome_modify_df[subtype_count_i] = 0.01 * biome_raw_df[subtype_percent_i] * biome_raw_df[\"WBC_COUNT\"]\n",
    "biome_modify_filename = \"BioMe_wbc_subtype_count_modified.txt\"\n",
    "pheno_save_dir = os.path.join(\"..\", \"raw_data\", \"phenotype\")\n",
    "biome_modify_dir_filename = os.path.join(pheno_save_dir, biome_modify_filename)\n",
    "biome_modify_df.to_csv(biome_modify_dir_filename, sep=\"\\t\", header=True, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Finding Common Sample between CN and Annotation File, and recodeing sex, adding more features."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "cohort = \"freeze6a-AA-noJHS\"\n",
    "save_dir = os.path.join(\"..\", \"cohort\", cohort, \"pre_data\")\n",
    "pheno_filename = \"pheno_%s.tsv\"%cohort\n",
    "common_pheno_anno_cn_egfr = pheno_prep(cohort, egfr_df, anno_df, pheno_df)\n",
    "pheno_wckd_wmicrocytosis_anemia_df = pheno_adding_features_recode_sex(common_pheno_anno_cn_egfr, save_dir, pheno_filename)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### CN"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Finding samples belong to freeze6a-AA but not belongs to JHS."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(7401, 2)"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cn_filename = \"freeze6a-AA_cn_qc20_173344-176029.tsv\"\n",
    "cn_dir = os.path.join(\"..\", \"cohort\", cohort, \"cn\")\n",
    "cn_dir_filename = os.path.join(cn_dir, cn_filename)\n",
    "cn = pd.read_csv(cn_dir_filename, sep=\"\\t\", header=0, index_col=None)\n",
    "\n",
    "jhs_cn_filename = \"JHS_CN.tsv\"\n",
    "jhs_load_dir = os.path.join(\"..\", \"cohort\", \"JHS\", \"cn\")\n",
    "jhs_cn_dir_filename = os.path.join(jhs_load_dir , jhs_cn_filename)\n",
    "jhs_cn = pd.read_csv(jhs_cn_dir_filename, sep=\"\\t\", header=0, index_col=None)\n",
    "\n",
    "common_col = \"NWDID\"\n",
    "cn_nojhs = remove_elements(cn, jhs_cn, common_col)\n",
    "cn_nojhs_filename = \"freeze6a-AA-noJHS_cn.tsv\"\n",
    "save_dir = cn_dir\n",
    "cn_nojhs_dir_filename = os.path.join(save_dir, cn_nojhs_filename)\n",
    "cn_nojhs.to_csv(cn_nojhs_dir_filename, sep=\"\\t\", header=True, index=False)\n",
    "cn_nojhs.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 146,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n"
     ]
    }
   ],
   "source": [
    "print(duplicates_num(cn_nojhs, \"NWDID\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### PC\n",
    "Now we take 1-11 PCs (PCAir)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [],
   "source": [
    "pc_filename = \"freeze6a_pcair_results.txt\"\n",
    "pc_load_dir = os.path.join(\"..\", \"raw_data\", \"pc\")\n",
    "pc_dir_filename = os.path.join(pc_load_dir, pc_filename)\n",
    "pc_raw = pd.read_csv(pc_dir_filename, sep=\"\\t\", header=None, index_col=None)\n",
    "pc = pc_raw.dropna(axis=1)\n",
    "\n",
    "pc_num = 11\n",
    "pc11 = pc[list(range(pc_num + 1))]\n",
    "pc11_header_list = [\"NWDID\"] + [\"pc%d\"%pc_i for pc_i in range(1, pc_num + 1)]\n",
    "pc11.columns = pc11_header_list\n",
    "\n",
    "pc_save_filename = \"freeze6a_pc%d_pcair.tsv\"%pc_num\n",
    "pc_save_dir = os.path.join(\"..\", \"prepro_data\", \"pc\")\n",
    "pc_save_dir_filename = os.path.join(pc_save_dir, pc_save_filename)\n",
    "pc11.to_csv(pc_save_dir_filename, sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n"
     ]
    }
   ],
   "source": [
    "print(duplicates_num(pc11, \"NWDID\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Kinship"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "rawdata_dir = os.path.join(\"..\", \"prepro_data\", \"kinship\")\n",
    "kinship_sample_filename = \"freeze6a_kinship_sample.tsv\"\n",
    "kinship_sample_dir_filename = os.path.join(rawdata_dir, kinship_sample_filename)\n",
    "kinship_sample_df = pd.read_csv(kinship_sample_dir_filename, sep=\"\\t\", header=0, index_col=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n"
     ]
    }
   ],
   "source": [
    "print(duplicates_num(kinship_sample_df, \"NWDID\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Common Samples Among PC, CN and SNPs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cohort = \"freeze6a-AA-noJHS\"\n",
    "freeze_ver = \"freeze6a\"\n",
    "pc_num = 11\n",
    "common_col = \"NWDID\"\n",
    "snp_list = [\"rs334\", \"rs33930165\", \"rs11248850\"]\n",
    "kinship_df_ID = freeze8_kinship_df_ID\n",
    "common_snp_pc_cn_pheno_kinship(cohort, freeze_ver, pc_num, common_col, snp_list, kinship_df_ID)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Common PC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(5989, 1)\n"
     ]
    }
   ],
   "source": [
    "cohort = \"freeze6a-AA-noJHS\"\n",
    "pc_num = 11\n",
    "filename_list = [\"freeze6a_pc%d_pcair.tsv\"%pc_num, \"snp_rs334_hetero.tsv\", \"snp_rs33930165_hetero.tsv\",\n",
    "                 \"snp_rs11248850_hetero.tsv\", \"%s_cn.tsv\"%cohort, \"pheno_%s.tsv\"%cohort, \"freeze6a_kinship_sample.tsv\"]\n",
    "predata_dir_list = [os.path.join(\"..\", \"prepro_data\", \"pc\"), os.path.join(\"..\", \"prepro_data\", \"snp\"),\n",
    "                    os.path.join(\"..\", \"prepro_data\", \"snp\"), os.path.join(\"..\", \"prepro_data\", \"snp\"),\n",
    "                    os.path.join(\"..\", \"cohort\", cohort, \"cn\"), os.path.join(\"..\", \"cohort\", cohort, \"pre_data\"),\n",
    "                    os.path.join(\"..\", \"prepro_data\", \"kinship\")]\n",
    "var_num = len(predata_dir_list)\n",
    "pc_snps_cn_dir_filename_list = [os.path.join(predata_dir_list[var_idx], filename_list[var_idx]) for var_idx in range(var_num)]\n",
    "pc_snps_cn_df = read2df_list(pc_snps_cn_dir_filename_list)\n",
    "common_col = \"NWDID\"\n",
    "common_pc = merge_df_list(pc_snps_cn_df, common_col, merge_method='first', how = 'inner')\n",
    "common_pc_filename = \"common_pc.tsv\"\n",
    "save_dir = os.path.join(\"..\", \"cohort\", cohort, \"ready_data\")\n",
    "common_pc_dir_filename = os.path.join(save_dir, common_pc_filename)\n",
    "common_pc.to_csv(common_pc_dir_filename, sep=\"\\t\", index=False)\n",
    "sample_df = common_pc[[\"NWDID\"]]\n",
    "sample_filename = \"common_sample.tsv\"\n",
    "sample_dir_filename = os.path.join(save_dir, sample_filename)\n",
    "sample_df.to_csv(sample_dir_filename, sep=\"\\t\", index=False)\n",
    "print(sample_df.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Common rs334"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(5989, 2)\n"
     ]
    }
   ],
   "source": [
    "pc_rs334_df_list = [pc_snps_cn_df[1], common_pc]\n",
    "common_col = \"NWDID\"\n",
    "common_rs334 = merge_df_list(pc_rs334_df_list, common_col, merge_method='first')\n",
    "common_rs334_filename = \"common_rs334.tsv\"\n",
    "common_rs334_dir_filename = os.path.join(save_dir, common_rs334_filename)\n",
    "common_rs334.to_csv(common_rs334_dir_filename, sep=\"\\t\", index=False)\n",
    "print(common_rs334.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Common rs33930165"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(5989, 2)\n"
     ]
    }
   ],
   "source": [
    "pc_rs33930165_df_list = [pc_snps_cn_df[2], common_pc]\n",
    "common_col = \"NWDID\"\n",
    "common_rs33930165 = merge_df_list(pc_rs33930165_df_list, common_col, merge_method='first')\n",
    "common_rs33930165_filename = \"common_rs33930165.tsv\"\n",
    "common_rs33930165_dir_filename = os.path.join(save_dir, common_rs33930165_filename)\n",
    "common_rs33930165.to_csv(common_rs33930165_dir_filename, sep=\"\\t\", index=False)\n",
    "print(common_rs33930165.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Common rs11248850"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(5989, 2)\n"
     ]
    }
   ],
   "source": [
    "pc_rs11248850_df_list = [pc_snps_cn_df[3], common_pc]\n",
    "common_col = \"NWDID\"\n",
    "common_rs11248850 = merge_df_list(pc_rs11248850_df_list, common_col, merge_method='first')\n",
    "common_rs11248850_filename = \"common_rs11248850.tsv\"\n",
    "common_rs11248850_dir_filename = os.path.join(save_dir, common_rs11248850_filename)\n",
    "common_rs11248850.to_csv(common_rs11248850_dir_filename, sep=\"\\t\", index=False)\n",
    "print(common_rs11248850.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Common CN"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(5989, 2)\n"
     ]
    }
   ],
   "source": [
    "pc_cn_df_list = [pc_snps_cn_df[4], common_pc]\n",
    "common_col = \"NWDID\"\n",
    "common_cn = merge_df_list(pc_cn_df_list, common_col, merge_method='first')\n",
    "common_cn_filename = \"common_cn.tsv\"\n",
    "common_cn_dir_filename = os.path.join(save_dir, common_cn_filename)\n",
    "common_cn.to_csv(common_cn_dir_filename, sep=\"\\t\", index=False)\n",
    "print(common_cn.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Common Phenotype"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(5989, 43)\n"
     ]
    }
   ],
   "source": [
    "pc_cn_df_list = [pheno_wckd_wmicrocytosis_anemia_df, common_pc]\n",
    "common_col = \"NWDID\"\n",
    "common_pheno = merge_df_list(pc_cn_df_list, common_col, merge_method='first')\n",
    "common_pheno_filename = \"common_pheno.tsv\"\n",
    "common_pheno_dir_filename = os.path.join(save_dir, common_pheno_filename)\n",
    "common_pheno.to_csv(common_pheno_dir_filename, sep=\"\\t\", index=False)\n",
    "print(common_pheno.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Apply Common Samples to Kinship Matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(5989, 1)\n",
      "(5989, 5990)\n"
     ]
    }
   ],
   "source": [
    "save_dir = os.path.join(\"..\", \"cohort\", cohort, \"ready_data\")\n",
    "sample_filename = \"common_sample.tsv\"\n",
    "sample_dir_filename = os.path.join(save_dir, sample_filename)\n",
    "sample_df = pd.read_csv(sample_dir_filename, sep=\"\\t\")\n",
    "print(sample_df.shape)\n",
    "kinship_sample_selected = kinship_select_sample(freeze6a_kinship_df_ID, sample_df)\n",
    "print(kinship_sample_selected.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [],
   "source": [
    "common_kinship_filename = \"common_kinship.tsv\"\n",
    "common_kinship_dir_filename = os.path.join(save_dir, common_kinship_filename)\n",
    "kinship_sample_selected.to_csv(common_kinship_dir_filename, sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Preparing Data for Each Table"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Table 1\n",
    "cn_binary_cn0, cn_binary_cn1, cn_binary_cn2, cn_binary_cn3, cn_binary_cn4, rs334, rs33930165"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### CN Binary: Recoding CN Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "cohort = \"freeze6a-AA-noJHS\"\n",
    "cn_orig_filename = \"common_cn.tsv\"\n",
    "cn_orig_dir = os.path.join(\"..\", \"cohort\", cohort, \"ready_data\")\n",
    "cn_orig_dir_filename = os.path.join(cn_orig_dir, cn_orig_filename)\n",
    "cn_orig_df = pd.read_csv(cn_orig_dir_filename, sep=\"\\t\")\n",
    "cn_binary_df_full = cn2binary_df(cn_orig_df)\n",
    "cn_binary_df = cn_binary_df_full[[\"NWDID\", \"cn0\", \"cn1\", \"cn3\", \"cn4\"]]\n",
    "cn_type_list = [\"cn0\", \"cn1\", \"cn3\"]\n",
    "for cn_type in cn_type_list:\n",
    "    cn_binary_df_move = move_cn(cn_binary_df, cn_type)\n",
    "    cn_binary_df_move_filename = \"common_cn-binary_%s.tsv\"%cn_type\n",
    "    cn_binary_df_move_dir_filename = os.path.join(cn_orig_dir, cn_binary_df_move_filename)\n",
    "    cn_binary_df_move.to_csv(cn_binary_df_move_dir_filename, sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Table 2\n",
    "rs334_del0, rs334_del1, rs334_del2, del_rs334"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### CN to DEL"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [],
   "source": [
    "cohort = \"freeze6a-AA-noJHS\"\n",
    "load_dir = os.path.join(\"..\", \"cohort\", cohort, \"ready_data\")\n",
    "cn_filename = \"common_cn.tsv\"\n",
    "cn_dir_filename = os.path.join(load_dir, cn_filename)\n",
    "cn = pd.read_csv(cn_dir_filename, sep=\"\\t\")\n",
    "del_cn = cn2del(cn)\n",
    "del_filename = \"common_del.tsv\"\n",
    "del_dir_filename = os.path.join(load_dir, del_filename)\n",
    "del_cn.to_csv(del_dir_filename, sep=\"\\t\", header=True, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### DEL Ã— rs334"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [],
   "source": [
    "load_dir = os.path.join(\"..\", \"cohort\", cohort, \"ready_data\")\n",
    "del_filename = \"common_del.tsv\"\n",
    "del_dir_filename = os.path.join(load_dir, del_filename)\n",
    "del_cn = pd.read_csv(del_dir_filename, sep=\"\\t\")\n",
    "\n",
    "rs334_filename = \"common_rs334.tsv\"\n",
    "rs334_dir_filename = os.path.join(load_dir, rs334_filename)\n",
    "rs334 = pd.read_csv(rs334_dir_filename, sep=\"\\t\")\n",
    "\n",
    "del_rs334 = pd.concat(objs=[del_cn, rs334[[\"geno\"]]], axis=1)\n",
    "del_rs334_filename = \"common_del-rs334.tsv\"\n",
    "save_dir = load_dir\n",
    "del_rs334_dir_filename = os.path.join(save_dir, del_rs334_filename)\n",
    "del_rs334.to_csv(del_rs334_dir_filename, sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Classifying rs334 by Number of Copies of Deletion"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [],
   "source": [
    "load_dir = os.path.join(\"..\", \"cohort\", cohort, \"ready_data\")\n",
    "del_rs334_filename = \"common_del-rs334.tsv\"\n",
    "del_rs334_dir_filename = os.path.join(load_dir, del_rs334_filename)\n",
    "del_rs334 = pd.read_csv(del_rs334_dir_filename, sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### No Copies of deletion, a.k.a. CN = 2, 3, 4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [],
   "source": [
    "save_dir = os.path.join(\"..\", \"cohort\", cohort, \"ready_data\")\n",
    "character_value_dict = {\"NWDID\":[\"no_thres\"], \"DEL\":[0, \"eq\"], \"geno\":[\"no_thres\"]}\n",
    "snp_classified_del_del0 = filter_characters(del_rs334, character_value_dict)\n",
    "snp_classified_del0 = snp_classified_del_del0[[\"NWDID\", \"geno\"]]\n",
    "snp_classified_del0_filename = \"common_rs334_del0.tsv\"\n",
    "snp_classified_del0_dir_filename = os.path.join(save_dir, snp_classified_del0_filename)\n",
    "snp_classified_del0.to_csv(snp_classified_del0_dir_filename, sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 1 Copy of deletion, a.k.a. CN = 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [],
   "source": [
    "save_dir = os.path.join(\"..\", \"cohort\", cohort, \"ready_data\")\n",
    "character_value_dict = {\"NWDID\":[\"no_thres\"], \"DEL\":[1, \"eq\"], \"geno\":[\"no_thres\"]}\n",
    "snp_classified_del_del1 = filter_characters(del_rs334, character_value_dict)\n",
    "snp_classified_del1 = snp_classified_del_del1[[\"NWDID\", \"geno\"]]\n",
    "snp_classified_del1_filename = \"common_rs334_del1.tsv\"\n",
    "snp_classified_del1_dir_filename = os.path.join(save_dir, snp_classified_del1_filename)\n",
    "snp_classified_del1.to_csv(snp_classified_del1_dir_filename, sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 2 Copies of deletion, a.k.a. CN = 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [],
   "source": [
    "save_dir = os.path.join(\"..\", \"cohort\", cohort, \"ready_data\")\n",
    "character_value_dict = {\"NWDID\":[\"no_thres\"], \"DEL\":[2, \"eq\"], \"geno\":[\"no_thres\"]}\n",
    "snp_classified_del_del2 = filter_characters(del_rs334, character_value_dict)\n",
    "snp_classified_del2 = snp_classified_del_del2[[\"NWDID\", \"geno\"]]\n",
    "snp_classified_del2_filename = \"common_rs334_del2.tsv\"\n",
    "snp_classified_del2_dir_filename = os.path.join(save_dir, snp_classified_del2_filename)\n",
    "snp_classified_del2.to_csv(snp_classified_del2_dir_filename, sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Table 3\n",
    "rs11248850, rs11248850_del0, rs11248850_del12, del_rs11248850"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### DEL Ã— rs11248850"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [],
   "source": [
    "cohort = \"freeze6a-AA-noJHS\"\n",
    "load_dir = os.path.join(\"..\", \"cohort\", cohort, \"ready_data\")\n",
    "del_filename = \"common_del.tsv\"\n",
    "del_dir_filename = os.path.join(load_dir, del_filename)\n",
    "del_cn = pd.read_csv(del_dir_filename, sep=\"\\t\")\n",
    "\n",
    "rs11248850_filename = \"common_rs11248850.tsv\"\n",
    "rs11248850_dir_filename = os.path.join(load_dir, rs11248850_filename)\n",
    "rs11248850 = pd.read_csv(rs11248850_dir_filename, sep=\"\\t\")\n",
    "\n",
    "del_rs11248850 = pd.concat(objs=[del_cn, rs11248850[[\"geno\"]]], axis=1)\n",
    "del_rs11248850_filename = \"common_del-rs11248850.tsv\"\n",
    "save_dir = load_dir\n",
    "del_rs11248850_dir_filename = os.path.join(save_dir, del_rs11248850_filename)\n",
    "del_rs11248850.to_csv(del_rs11248850_dir_filename, sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Classifying rs334 by # Copies of Deletion"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [],
   "source": [
    "load_dir = os.path.join(\"..\", \"cohort\", cohort, \"ready_data\")\n",
    "del_rs11248850_filename = \"common_del-rs11248850.tsv\"\n",
    "del_rs11248850_dir_filename = os.path.join(load_dir, del_rs11248850_filename)\n",
    "del_rs11248850 = pd.read_csv(del_rs11248850_dir_filename, sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### DEL = 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [],
   "source": [
    "save_dir = os.path.join(\"..\", \"cohort\", cohort, \"ready_data\")\n",
    "character_value_dict = {\"NWDID\":[\"no_thres\"], \"DEL\":[0, \"eq\"], \"geno\":[\"no_thres\"]}\n",
    "snp_classified_del_del0 = filter_characters(del_rs11248850, character_value_dict)\n",
    "snp_classified_del0 = snp_classified_del_del0[[\"NWDID\", \"geno\"]]\n",
    "snp_classified_del0_filename = \"common_rs11248850_del0.tsv\"\n",
    "snp_classified_del0_dir_filename = os.path.join(save_dir, snp_classified_del0_filename)\n",
    "snp_classified_del0.to_csv(snp_classified_del0_dir_filename, sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### DEL = 1, 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [],
   "source": [
    "save_dir = os.path.join(\"..\", \"cohort\", cohort, \"ready_data\")\n",
    "character_value_dict = {\"NWDID\":[\"no_thres\"], \"DEL\":[1, \"eq\"], \"geno\":[\"no_thres\"]}\n",
    "snp_classified_del_del1 = filter_characters(del_rs11248850, character_value_dict)\n",
    "snp_classified_del1 = snp_classified_del_del1[[\"NWDID\", \"geno\"]]\n",
    "\n",
    "character_value_dict = {\"NWDID\":[\"no_thres\"], \"DEL\":[2, \"eq\"], \"geno\":[\"no_thres\"]}\n",
    "snp_classified_del_del2 = filter_characters(del_rs11248850, character_value_dict)\n",
    "snp_classified_del2 = snp_classified_del_del2[[\"NWDID\", \"geno\"]]\n",
    "\n",
    "snp_classified_del12 = pd.concat(objs=[snp_classified_del1, snp_classified_del2], axis=0)\n",
    "\n",
    "snp_classified_del12_filename = \"common_rs11248850_del12.tsv\"\n",
    "snp_classified_del12_dir_filename = os.path.join(save_dir, snp_classified_del12_filename)\n",
    "snp_classified_del12.to_csv(snp_classified_del12_dir_filename, sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "table3 completed.\n"
     ]
    }
   ],
   "source": [
    "phenotype_list = [\"hemoglobin_mcnc_bld_1\", \"hematocrit_vfr_bld_1\", \"rbc_ncnc_bld_1\",\n",
    "                  \"mcv_entvol_rbc_1\", \"mch_entmass_rbc_1\", \"mchc_mcnc_rbc_1\", \"rdw_ratio_rbc_1\",\n",
    "                  \"neutrophil_ncnc_bld_1\", \"lymphocyte_ncnc_bld_1\", \"basophil_ncnc_bld_1\",\n",
    "                  \"eosinophil_ncnc_bld_1\", \"monocyte_ncnc_bld_1\", \"wbc_ncnc_bld_1\",\n",
    "                  \"pmv_entvol_bld_1\", \"platelet_ncnc_bld_1\", \"EGFRCKDEPI\", \"CKD\", \"microcytosis\", \"anemia\"]\n",
    "table_dict = {\"table1\":[\"cn-binary_cn0\", \"cn-binary_cn1\", \"cn-binary_cn3\", \"rs334\", \"rs33930165\"],\n",
    "              \"table2\":[\"rs334_del0\", \"rs334_del1\", \"rs334_del2\", \"del-rs334\"],\n",
    "              \"table3\":[\"rs11248850\", \"rs11248850_del0\", \"rs11248850_del12\", \"del-rs11248850\"]}\n",
    "load_dir = os.path.join(\"..\", \"cohort\", cohort, \"ready_data\")\n",
    "wrap_prepare_matrix_pheno_agesex_in_pheno(phenotype_list, table_dict, load_dir)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Checking RBC Count Array of \"Table 2 two copies deletion (CN=0)\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 249,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXcAAAD9CAYAAABHnDf0AAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDMuMC4yLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvOIA7rQAADk9JREFUeJzt3X2MpWV9h/HrK2sFtwu2oqMFwrTVmI0rtGZ8C8TOSmIqGGlSGrFVK9l2g0XkjzZ2yabW0GxKTZv61kq2bhp825IQqOjajRb3hK4WcJY3jUNTomvYLI1QWuquiF389Y99SIdhXs6ZmTNn9ub6JJM98zz3uZ97yXDNs/ecmUlVIUlqy3NGvQBJ0soz7pLUIOMuSQ0y7pLUIOMuSQ0y7pLUIOMuSQ0y7pLUIOMuSQ1aN6oLn3766TU+Pj6qy0sLOnr0KOvXrx/1MqRnOHDgwCNV9aLFxo0s7uPj40xNTY3q8tKCer0ek5OTo16G9AxJvt/POLdlJKlBxl2SGmTcJalBxl2SGmTcJalBxl2SGmTcJalBxl2SGjSyb2KSVkuSVbmOv49Ya4l37mpeVQ38dvYff2ng50hriXGXpAYZd0lqkHGXpAYZd0lqkHGXpAYZd0lqkHGXpAYZd0lqkHGXpAYZd0lqkHGXpAYZd0lqkHGXpAYZd0lqkHGXpAYZd0lqkHGXpAYZd0lqkHGXpAYZd0lqUF9xT/KCJDcmuT/JdJI3zDqfJB9L8kCS+5K8ejjLlST1Y12f4z4K7K2qS5L8DPD8WeffAry8e3sd8MnuT0nSCCx6557kVOCNwC6AqvpJVf33rGEXA5+u424HXpDkpSu+WklSX/q5c/8l4GHg75OcCxwArqqqozPGnAE8OOP9Q92xh2ZOlGQrsBVgbGyMXq+39JVLQ+bHp05k/cR9HfBq4MqquiPJR4FtwJ/MGJM5nlfPOFC1E9gJMDExUZOTkwMvWFoVe/fgx6dOZP18QfUQcKiq7ujev5HjsZ895qwZ758JHF7+8iRJS7Fo3KvqP4AHk7yiO3QB8J1Zw24B3t29aub1wGNV9RCSpJHo99UyVwKf614p813gsiSXA1TVdcCXgQuBB4AfAZcNYa2SpD71FfequgeYmHX4uhnnC7hiBdclSVoGv0NVkhpk3CWpQcZdkhpk3CWpQcZdkhpk3CWpQcZdkhpk3CWpQcZdkhpk3CWpQcZdkhpk3CWpQcZdkhpk3CWpQcZdkhpk3CWpQcZdkhpk3CWpQcZdkhpk3CWpQcZdkhpk3CWpQcZdkhpk3CWpQcZdkhpk3CWpQcZdkhpk3CWpQev6GZTkIPBD4EngWFVNzDo/CXwB+F536KaqumbllilJGkRfce9srqpHFjj/L1X11uUuSJK0fG7LSFKD+o17AV9JciDJ1nnGvCHJvUn+KckrV2h9kqQl6Hdb5ryqOpzkxcBXk9xfVbfNOH8XcHZVHUlyIfCPwMtnT9J9YtgKMDY2Rq/XW97q9axzxa1HOfq/q3Ot8W17hjr/+ufC31ywfqjX0LNXqmqwJyQfAo5U1V8uMOYgMLHQHv3ExERNTU0NdG1pfNseDl570dCv0+v1mJycHOo1VuvvorYkOTD7RS1zWXRbJsn6JBueegy8Gfj2rDEvSZLu8Wu7ef9zKQuXJC1fP9syY8DNXbvXAZ+vqr1JLgeoquuAS4D3JjkGPA5cWoP+k0CStGIWjXtVfRc4d47j1814/AngEyu7NEnSUvlSSElqkHGXpAYZd0lqkHGXpAYZd0lqkHGXpAYZd0lqkHGXpAYZd0lqkHGXpAYZd0lqkHGXpAYZd0lqkHGXpAYZd0lqkHGXpAYZd0lqkHGXpAb18ztUpTVjw8ZtvOr6batzseuHO/2GjQAXDfcietYy7jqh/HD6Wg5eO/wg9no9Jicnh3qN8W17hjq/nt3clpGkBhl3SWqQcZekBhl3SWqQcZekBhl3SWqQcZekBhl3SWqQcZekBvUV9yQHk3wryT1JpuY4nyQfS/JAkvuSvHrllypJ6tcgP35gc1U9Ms+5twAv795eB3yy+1OSNAIrtS1zMfDpOu524AVJXrpCc0uSBtRv3Av4SpIDSbbOcf4M4MEZ7x/qjkmSRqDfbZnzqupwkhcDX01yf1XdNuN85nhOzT7QfWLYCjA2Nkav1xt0vdKqfNwcOXJkVa7j/wMalr7iXlWHuz9/kORm4LXAzLgfAs6a8f6ZwOE55tkJ7ASYmJioYf9IVTVo756h/yheWJ0f+btafxc9Oy26LZNkfZINTz0G3gx8e9awW4B3d6+aeT3wWFU9tOKrlST1pZ879zHg5iRPjf98Ve1NcjlAVV0HfBm4EHgA+BFw2XCWK0nqx6Jxr6rvAufOcfy6GY8LuGJllyZJWiq/Q1WSGmTcJalBxl2SGmTcJalBxl2SGmTcJalBxl2SGmTcJalBxl2SGjTIL+uQ1oTxbXtW50J7h3ud00557lDn17ObcdcJ5eC1F63Kdca37Vm1a0nD4LaMJDXIuEtSg4y7JDXIuEtSg4y7JDXIuEtSg4y7JDXIuEtSg4y7JDXIuEtSg4y7JDXIuEtSg4y7JDXIuEtSg4y7JDXIuEtSg4y7JDXIuEtSg/qOe5KTktyd5EtznHtPkoeT3NO9/d7KLlOSNIhBfofqVcA0cOo852+oqvctf0mSpOXq6849yZnARcCnhrscSdJK6Hdb5iPAB4CfLjDmN5Pcl+TGJGctf2mSpKVadFsmyVuBH1TVgSST8wz7IrC7qp5IcjlwPfCmOebaCmwFGBsbo9frLXXd0tD58akTWapq4QHJnwPvAo4BJ3N8z/2mqnrnPONPAh6tqtMWmndiYqKmpqaWtGhp2Ma37eHgtReNehnSMyQ5UFUTi41bdFumqq6uqjOrahy4FPja7LAneemMd9/G8S+8SpJGZJBXyzxNkmuAqaq6BXh/krdx/O7+UeA9K7M8SdJSDBT3quoBve7xB2ccvxq4eiUXJklaOr9DVZIaZNwlqUHGXZIaZNwlqUHGXZIaZNwlqUHGXZIaZNwlqUHGXZIaZNwlqUHGXZIaZNwlqUHGXZIaZNwlqUHGXZIaZNwlqUHGXZIaZNwlqUHGXZIaZNwlqUHGXZIaZNwlqUHGXZIaZNwlqUHGXZIaZNwlqUHGXZIaZNwlqUHrRr0AadiSLO15fzHY+Kpa0nWkYej7zj3JSUnuTvKlOc49L8kNSR5IckeS8ZVcpLQcVTXw2759+wZ+jrSWDLItcxUwPc+5LcB/VdXLgL8GBrznkSStpL7inuRM4CLgU/MMuRi4vnt8I3BBlvpvYUnSsvV75/4R4APAT+c5fwbwIEBVHQMeA1647NVJkpZk0S+oJnkr8IOqOpBkcr5hcxx7xiZkkq3AVoCxsTF6vV7/K5VW0ZEjR/z41Amtn1fLnAe8LcmFwMnAqUk+W1XvnDHmEHAWcCjJOuA04NHZE1XVTmAnwMTERE1OTi5z+dJw9Ho9/PjUiWzRbZmqurqqzqyqceBS4Guzwg5wC/C73eNLujG+fECSRmTJr3NPcg0wVVW3ALuAzyR5gON37Jeu0PokSUswUNyrqgf0uscfnHH8x8BvreTCJElL548fkKQGGXdJapBxl6QGGXdJapBxl6QGGXdJapBxl6QGGXdJapBxl6QGGXdJapBxl6QGGXdJapBxl6QGGXdpht27d7Np0yYuuOACNm3axO7du0e9JGlJlvzz3KXW7N69m+3bt7Nr1y6efPJJTjrpJLZs2QLAO97xjhGvThqMd+5SZ8eOHezatYvNmzezbt06Nm/ezK5du9ixY8eolyYNzLhLnenpac4///ynHTv//POZnp4e0YqkpTPuUmfjxo3s37//acf279/Pxo0bR7QiaemMu9TZvn07W7ZsYd++fRw7dox9+/axZcsWtm/fPuqlSQPzC6pS56kvml555ZVMT0+zceNGduzY4RdTdUJKVY3kwhMTEzU1NTWSa0uL6fV6TE5OjnoZ0jMkOVBVE4uNc1tGkhpk3CWpQcZdkhpk3CWpQcZdkho0slfLJHkY+P5ILi4t7nTgkVEvQprD2VX1osUGjSzu0lqWZKqfl5tJa5XbMpLUIOMuSQ0y7tLcdo56AdJyuOcuSQ3yzl2SGmTcJalBxl1rSpIjo17DoJJ8KMkfzXF8PMm353lOL8m8L7VMsiPJgyfifw+tDcZda0aS4MfkU74IvHbUi9CJy/+RNFLd3e10kr8F7gJOSfJXSe5KcmuSF3XjXpbkn5Pc25375Xnmm+zuim9Mcn+Sz3WfNEjymiTf6Oa4M8mGJO9JclOSvUn+PcmHZ8z169217k1y6yJ/lXOTfK2b4/fnWNcpSf4hyX1JbgBOWWiyqrq9qh5a5JrSvPxNTFoLXgFcVlV/kKSAu6rqD5N8EPhT4H3A54Brq+rmJCez8I3JrwKvBA4DXwfOS3IncAPw9qr6ZpJTgce78b/SPecJ4N+SfBz4MfB3wBur6ntJfn6Rv8M5wOuB9cDdSfbMOv9e4EdVdU6Sczj+iUwaGuOuteD7VXV79/inHI8wwGeBm5JsAM6oqpsBqurHi8x3Z1UdAkhyDzAOPAY8VFXf7Ob4n+48wK1V9Vj3/neAs4GfA26rqu914x9d5JpfqKrHgceT7OP4lso9M86/EfhYN9d9Se5bZD5pWdyW0VpwdIFzBWTA+Z6Y8fhJjt/EpJtrJcbPZfbYuZ7rN5Vo1Rh3rTXPAS7pHv82sL+7yz6U5DcAkjwvyfMHnPd+4BeSvKabY0OShf7l+q/AryX5xW78YtsyFyc5OckLgUngm7PO3wb8TjfXJo5v40hDY9y11hwFXpnkAPAm4Jru+LuA93fbGd8AXjLIpFX1E+DtwMeT3At8FTh5gfEPA1s5vi10L/+/VTSfO4E9wO3An1XV4VnnPwn8bLf+D3Tj55Xkw0kOAc9PcijJhxa5vvQ0/vgBSWqQd+6S1CBfLaMTUpJXAZ+ZdfiJqnrdEK95GXDVrMNfr6orljHnHcDzZh1+V1V9a6lzSuC2jCQ1yW0ZSWqQcZekBhl3SWqQcZekBhl3SWrQ/wGSPe9rt/zeFgAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "pheno_filename = \"common_pheno_table2_rs334_del2_rbc_ncnc_bld_1.tsv\"\n",
    "pheno_dir = \"/proj/yunligrp/users/minzhi/asso/cohort/freeze6a-AA-noJHS/table2/rs334_del2\"\n",
    "pheno_dir_filename = os.path.join(pheno_dir, pheno_filename)\n",
    "pheno_df = pd.read_csv(pheno_dir_filename, sep=\"\\t\", header=0, index_col = 0)\n",
    "rbc_boxplot = pheno_df.boxplot()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Association between Blood Cell Trait and SOL"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Preprocess Data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Phenotype"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "cohort = \"SOL\"\n",
    "save_dir = os.path.join(\"..\", \"cohort\", cohort, \"pre_data\")\n",
    "pheno_filename = \"pheno_%s.tsv\"%cohort\n",
    "common_pheno_anno_cn_egfr = pheno_prep(cohort, egfr_df, anno_df, pheno_df)\n",
    "pheno_wckd_wmicrocytosis_anemia_df = pheno_adding_features_recode_sex(common_pheno_anno_cn_egfr, save_dir, pheno_filename)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### D-Dimer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "metadata": {},
   "outputs": [],
   "source": [
    "ddmier_filename = \"TOPMED_HarmonizedPhenotypes_DDIMER_21MAY2019.csv\"\n",
    "ddmier_dir = os.path.join(\"..\", \"raw_data\", \"phenotype\")\n",
    "ddmier_dir_filename = os.path.join(ddmier_dir, ddmier_filename)\n",
    "ddmier = pd.read_csv(ddmier_dir_filename, sep = \",\", header = 0, index_col = None)\n",
    "ddmier_selected = ddmier[[\"sample.id\", \"sample_remove_DDIMER\", \"DDIMER\", \"AGE_DDIMER\"]] #, \"study_race\", \"sex\"\n",
    "ddmier_selected.rename(columns={\"sample.id\":\"NWDID\", \"AGE_DDIMER\":\"age_at_DDIMER\"}, inplace=True)\n",
    "ddmier_selected_filename = \"SOL_ddimer.tsv\"\n",
    "ddmier_selected_dir = os.path.join(\"..\", \"prepro_data\", \"phenotype\")\n",
    "ddmier_selected_dir_filename = os.path.join(ddmier_selected_dir, ddmier_selected_filename)\n",
    "ddmier_selected.to_csv(ddmier_selected_dir_filename, sep = \"\\t\", header = True, index = False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Gengrp6\n",
    "After testing, to each sample, if s/he has non-NaN SUBJECT_ID and gengrp6 value, then it must have non-NaN CONSENT_text and INTERNAL_USE_ONLY value."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "metadata": {},
   "outputs": [],
   "source": [
    "gengrp6_filename = \"page-harmonized-phenotypes-pca-freeze2-candidate2-2016-12-14.GWASid_fid_22May2018internalPCs.SOLv2consent.txt\"\n",
    "gengrp6_dir = os.path.join(\"..\", \"raw_data\", \"adjustment\", \"gengrp6\")\n",
    "gengrp6_dir_filename = os.path.join(gengrp6_dir, gengrp6_filename)\n",
    "gengrp6 = pd.read_csv(gengrp6_dir_filename, sep = \"\\t\", header = 0, index_col = None)\n",
    "gengrp6.replace(\".\", np.nan, inplace=True)\n",
    "gengrp6_select = gengrp6[[\"z_sol_id\", \"analysis_id\", \"CONSENT_text\", \"INTERNAL_USE_ONLY\",\n",
    "                          \"gengrp6\"]].dropna(axis=0, subset=[\"analysis_id\",\"gengrp6\"],how=\"any\")\n",
    "gengrp6_select.rename(columns = {\"analysis_id\":\"SUBJECT_ID\"}, inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Weight & Center"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 173,
   "metadata": {},
   "outputs": [],
   "source": [
    "weight_center_dir = os.path.join(\"..\", \"raw_data\", \"adjustment\")\n",
    "weight_center_filename = \"bloodcell_output.csv\"\n",
    "weight_center_dir_filename = os.path.join(weight_center_dir, weight_center_filename)\n",
    "weight_center = pd.read_csv(weight_center_dir_filename, sep = \",\", header = 0, index_col = None)\n",
    "weight_center_select = weight_center[[\"ID\", \"WEIGHT_FINAL_NORM_OVERALL\", \"CENTER\"]].dropna(axis = 0, how = \"any\")\n",
    "weight_center_select.rename(columns = {\"ID\":\"z_sol_id\"}, inplace = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Merging Weight & Center Table with Gengrp6 Table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "metadata": {},
   "outputs": [],
   "source": [
    "common_col = \"z_sol_id\"\n",
    "gengrp6_weight_center_list = [gengrp6_select, weight_center_select]\n",
    "gengrp6_weight_center = merge_df_list(gengrp6_weight_center_list, common_col, merge_method='merge', how = 'inner')\n",
    "gengrp6_weight_center.drop(axis = 1, columns = [\"z_sol_id\"], inplace = True)\n",
    "gengrp6_weight_center_filename = \"SOL_gengrp6_weight_center.tsv\"\n",
    "gengrp6_weight_center_dir = os.path.join(\"..\", \"prepro_data\", \"phenotype\")\n",
    "gengrp6_weight_center_dir_filename = os.path.join(gengrp6_weight_center_dir, gengrp6_weight_center_filename)\n",
    "gengrp6_weight_center.to_csv(gengrp6_weight_center_dir_filename, sep = \"\\t\", header = True, index = False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Merging Weight, Center, Gengrp6 and D-Dimer into Phenotype Table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "gengrp6_weight_center_filename = \"SOL_gengrp6_weight_center.tsv\"\n",
    "gengrp6_weight_center_dir = os.path.join(\"..\", \"prepro_data\", \"phenotype\")\n",
    "gengrp6_weight_center_dir_filename = os.path.join(gengrp6_weight_center_dir, gengrp6_weight_center_filename)\n",
    "gengrp6_weight_center = pd.read_csv(gengrp6_weight_center_dir_filename, sep = \"\\t\", header = 0, index_col = None)\n",
    "#display(gengrp6_weight_center)\n",
    "\n",
    "ddimer_filename = \"SOL_ddimer.tsv\"\n",
    "ddimer_dir = os.path.join(\"..\", \"prepro_data\", \"phenotype\")\n",
    "ddimer_dir_filename = os.path.join(ddimer_dir, ddimer_filename)\n",
    "ddimer = pd.read_csv(ddimer_dir_filename, sep = \"\\t\", header = 0, index_col = None)\n",
    "#display(ddimer)\n",
    "\n",
    "common_col = \"SUBJECT_ID\"\n",
    "pheno_gengrp6_weight_center_list = [pheno_wckd_wmicrocytosis_anemia_df, gengrp6_weight_center]\n",
    "pheno_gengrp6_weight_center = merge_df_list(pheno_gengrp6_weight_center_list, common_col, merge_method='merge', how = 'outer')\n",
    "\n",
    "common_col = \"NWDID\"\n",
    "pheno_gengrp6_weight_center_ddimer_list = [pheno_gengrp6_weight_center, ddimer]\n",
    "pheno_gengrp6_weight_center_ddimer = merge_df_list(pheno_gengrp6_weight_center_ddimer_list, common_col, merge_method='merge', how = 'outer')\n",
    "#display(pheno_gengrp6_weight_center_ddimer)\n",
    "\n",
    "pheno_gengrp6_weight_center_ddimer.dropna(axis = 0, subset = [\"NWDID\"], inplace = True)\n",
    "pheno_gengrp6_weight_center_ddimer = pheno_gengrp6_weight_center_ddimer[pheno_gengrp6_weight_center_ddimer[\"CONSENT_text\"] != \"DROP\"]\n",
    "pheno_gengrp6_weight_center_ddimer = pheno_gengrp6_weight_center_ddimer[pheno_gengrp6_weight_center_ddimer[\"sample_remove_DDIMER\"] != 1]\n",
    "\n",
    "add_adj_dict = {\"quant\":[\"WEIGHT_FINAL_NORM_OVERALL\"],\"cati\":[\"gengrp6\", \"CENTER\"]}\n",
    "pheno_gengrp6_weight_center_ddimer_dummy_adj, add_adj_list = cati2dummy_df(pheno_gengrp6_weight_center_ddimer, add_adj_dict)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "save_dir = os.path.join(\"..\", \"cohort\", cohort, \"pre_data\")\n",
    "pheno_filename = \"pheno_%s_20190904.tsv\"%cohort\n",
    "pheno_dir_filename = os.path.join(save_dir, pheno_filename)\n",
    "pheno_gengrp6_weight_center_ddimer.to_csv(pheno_dir_filename, sep = \"\\t\", header = True, index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "cohort = \"SOL\"\n",
    "cn_filename = \"%s_cn.tsv\"%cohort\n",
    "cn_dir = os.path.join(\"..\", \"cohort\", cohort, \"cn\")\n",
    "cn_dir_filename = os.path.join(cn_dir, cn_filename)\n",
    "cn_df = pd.read_csv(cn_dir_filename, sep = \"\\t\", header = 0, index_col = None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [],
   "source": [
    "ddmier_filename = \"TOPMED_HarmonizedPhenotypes_DDIMER_21MAY2019.csv\"\n",
    "ddmier_dir = os.path.join(\"..\", \"raw_data\", \"phenotype\")\n",
    "ddmier_dir_filename = os.path.join(ddmier_dir, ddmier_filename)\n",
    "ddmier = pd.read_csv(ddmier_dir_filename, sep = \",\", header = 0, index_col = None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "common_col = \"NWDID\"\n",
    "cn_df_ddimer_list = [cn_df, ddimer]\n",
    "cn_df_ddimer_inner = merge_df_list(cn_df_ddimer_list, common_col, merge_method='merge', how = 'inner')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>NWDID</th>\n",
       "      <th>CN</th>\n",
       "      <th>sample.id</th>\n",
       "      <th>unique_subject_key</th>\n",
       "      <th>unique_subject_key_f6a</th>\n",
       "      <th>flag_diff_USK_F8_f6a</th>\n",
       "      <th>subject_id</th>\n",
       "      <th>topmed_project</th>\n",
       "      <th>topmed_phase</th>\n",
       "      <th>funding</th>\n",
       "      <th>...</th>\n",
       "      <th>PC_24</th>\n",
       "      <th>PC_25</th>\n",
       "      <th>PC_26</th>\n",
       "      <th>PC_27</th>\n",
       "      <th>PC_28</th>\n",
       "      <th>PC_29</th>\n",
       "      <th>PC_30</th>\n",
       "      <th>PC_31</th>\n",
       "      <th>PC_32</th>\n",
       "      <th>study</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>0 rows Ã— 112 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [NWDID, CN, sample.id, unique_subject_key, unique_subject_key_f6a, flag_diff_USK_F8_f6a, subject_id, topmed_project, topmed_phase, funding, seq_center, sex_dcc_f6a, Xchr.anom_f6a, Ychr.anom_f6a, exclude_f6a, id, STUDY, ETHNICITY, ETHNICITY_LET, study_race, sex, FIB, FVII, FVII_METHOD, FVII_ACTIVITY, FVII_ANTIGEN, FVIII, VWF, DDIMER, TPA, PAI1, AGE_FIB, AGE_FVII, AGE_FVII_ACTIVITY, AGE_FVII_ANTIGEN, AGE_FVIII, AGE_VWF, AGE_DDIMER, AGE_TPA, AGE_PAI1, ANTICOAG_FIB, ANTICOAG_FVII, ANTICOAG_FVII_ACTIVITY, ANTICOAG_FVII_ANTIGEN, ANTICOAG_FVIII, ANTICOAG_VWF, ANTICOAG_DDIMER, ANTICOAG_TPA, ANTICOAG_PAI1, AMISH_SSC_FIB, ARIC_SSC, CARDIA_SSC, CARDIA_SSC_FIB, CARDIA_SSC_FVII, CARDIA_SSC_FVIII, CARDIA_SSC_VWF, CFS_SSC, CHS_SSC, FHS_SSC_FIB_PAI1, GeneSTAR_SSC, GeneSTAR_SSC_FIB, GENOA_SSC, MESA_SSC, WHI_SSC, WHI_SSC_FIB, WHI_SSC_DDIMER, WHI_SSC_TPA, WHI_SSC_PAI1, MZtwinID, TechnicalRepID, BetweenStudyDupID, sample_remove_default, sample_remove_FVII, sample_remove_FVII_antigen, sample_remove_FVIII, sample_remove_VWF, sample_remove_DDIMER, sample_remove_TPA, sample_remove_PAI1, PC_1, PC_2, PC_3, PC_4, PC_5, PC_6, PC_7, PC_8, PC_9, PC_10, PC_11, PC_12, PC_13, PC_14, PC_15, PC_16, PC_17, PC_18, PC_19, PC_20, PC_21, ...]\n",
       "Index: []\n",
       "\n",
       "[0 rows x 112 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "a = cn_df.merge(ddmier, left_on=\"NWDID\", right_on=\"sample.id\")\n",
    "display(a)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n"
     ]
    }
   ],
   "source": [
    "print(duplicates_num(pheno_gengrp6_weight_center_ddimer, \"NWDID\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### CN"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Finding samples belong to freeze6a-AA but not belongs to JHS."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(4001, 2)"
      ]
     },
     "execution_count": 141,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cohort = \"SOL\"\n",
    "cn_filename = \"SOL_cn.tsv\"\n",
    "cn_dir = os.path.join(\"..\", \"cohort\", cohort, \"cn\")\n",
    "cn_dir_filename = os.path.join(cn_dir, cn_filename)\n",
    "cn = pd.read_csv(cn_dir_filename, sep=\"\\t\", header=0, index_col=None)\n",
    "cn.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Preparing Data for Each Table"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Table 1\n",
    "cn_binary_cn0, cn_binary_cn1, cn_binary_cn2, cn_binary_cn3, cn_binary_cn4, rs334, rs33930165"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### CN Binary: Recoding CN Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "metadata": {},
   "outputs": [],
   "source": [
    "cohort = \"SOL\"\n",
    "cn_orig_filename = \"common_cn.tsv\"\n",
    "cn_orig_dir = os.path.join(\"..\", \"cohort\", cohort, \"ready_data\")\n",
    "cn_orig_dir_filename = os.path.join(cn_orig_dir, cn_orig_filename)\n",
    "cn_orig_df = pd.read_csv(cn_orig_dir_filename, sep=\"\\t\")\n",
    "cn_binary_df_full = cn2binary_df(cn_orig_df)\n",
    "cn_binary_df = cn_binary_df_full[[\"NWDID\", \"cn0\", \"cn1\", \"cn3\", \"cn4\"]]\n",
    "cn_type_list = [\"cn0\", \"cn1\", \"cn3\"]\n",
    "for cn_type in cn_type_list:\n",
    "    cn_binary_df_move = move_cn(cn_binary_df, cn_type)\n",
    "    cn_binary_df_move_filename = \"common_cn-binary_%s.tsv\"%cn_type\n",
    "    cn_binary_df_move_dir_filename = os.path.join(cn_orig_dir, cn_binary_df_move_filename)\n",
    "    cn_binary_df_move.to_csv(cn_binary_df_move_dir_filename, sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Table 2\n",
    "rs334_del0, rs334_del1, rs334_del2, del_rs334"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### CN to DEL"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 138,
   "metadata": {},
   "outputs": [],
   "source": [
    "cohort = \"SOL\"\n",
    "load_dir = os.path.join(\"..\", \"cohort\", cohort, \"ready_data\")\n",
    "cn_filename = \"common_cn.tsv\"\n",
    "cn_dir_filename = os.path.join(load_dir, cn_filename)\n",
    "cn = pd.read_csv(cn_dir_filename, sep=\"\\t\")\n",
    "del_cn = cn2del(cn)\n",
    "del_filename = \"common_del.tsv\"\n",
    "del_dir_filename = os.path.join(load_dir, del_filename)\n",
    "del_cn.to_csv(del_dir_filename, sep=\"\\t\", header=True, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### DEL Ã— rs334"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "metadata": {},
   "outputs": [],
   "source": [
    "load_dir = os.path.join(\"..\", \"cohort\", cohort, \"ready_data\")\n",
    "del_filename = \"common_del.tsv\"\n",
    "del_dir_filename = os.path.join(load_dir, del_filename)\n",
    "del_cn = pd.read_csv(del_dir_filename, sep=\"\\t\")\n",
    "\n",
    "rs334_filename = \"common_rs334.tsv\"\n",
    "rs334_dir_filename = os.path.join(load_dir, rs334_filename)\n",
    "rs334 = pd.read_csv(rs334_dir_filename, sep=\"\\t\")\n",
    "\n",
    "del_rs334 = pd.concat(objs=[del_cn, rs334[[\"geno\"]]], axis=1)\n",
    "del_rs334_filename = \"common_del-rs334.tsv\"\n",
    "save_dir = load_dir\n",
    "del_rs334_dir_filename = os.path.join(save_dir, del_rs334_filename)\n",
    "del_rs334.to_csv(del_rs334_dir_filename, sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Classifying rs334 by Number of Copies of Deletion"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "metadata": {},
   "outputs": [],
   "source": [
    "load_dir = os.path.join(\"..\", \"cohort\", cohort, \"ready_data\")\n",
    "del_rs334_filename = \"common_del-rs334.tsv\"\n",
    "del_rs334_dir_filename = os.path.join(load_dir, del_rs334_filename)\n",
    "del_rs334 = pd.read_csv(del_rs334_dir_filename, sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### No Copies of deletion, a.k.a. CN = 2, 3, 4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "metadata": {},
   "outputs": [],
   "source": [
    "save_dir = os.path.join(\"..\", \"cohort\", cohort, \"ready_data\")\n",
    "character_value_dict = {\"NWDID\":[\"no_thres\"], \"DEL\":[0, \"eq\"], \"geno\":[\"no_thres\"]}\n",
    "snp_classified_del_del0 = filter_characters(del_rs334, character_value_dict)\n",
    "snp_classified_del0 = snp_classified_del_del0[[\"NWDID\", \"geno\"]]\n",
    "snp_classified_del0_filename = \"common_rs334_del0.tsv\"\n",
    "snp_classified_del0_dir_filename = os.path.join(save_dir, snp_classified_del0_filename)\n",
    "snp_classified_del0.to_csv(snp_classified_del0_dir_filename, sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 1 Copy of deletion, a.k.a. CN = 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "metadata": {},
   "outputs": [],
   "source": [
    "save_dir = os.path.join(\"..\", \"cohort\", cohort, \"ready_data\")\n",
    "character_value_dict = {\"NWDID\":[\"no_thres\"], \"DEL\":[1, \"eq\"], \"geno\":[\"no_thres\"]}\n",
    "snp_classified_del_del1 = filter_characters(del_rs334, character_value_dict)\n",
    "snp_classified_del1 = snp_classified_del_del1[[\"NWDID\", \"geno\"]]\n",
    "snp_classified_del1_filename = \"common_rs334_del1.tsv\"\n",
    "snp_classified_del1_dir_filename = os.path.join(save_dir, snp_classified_del1_filename)\n",
    "snp_classified_del1.to_csv(snp_classified_del1_dir_filename, sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 2 Copies of deletion, a.k.a. CN = 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "metadata": {},
   "outputs": [],
   "source": [
    "save_dir = os.path.join(\"..\", \"cohort\", cohort, \"ready_data\")\n",
    "character_value_dict = {\"NWDID\":[\"no_thres\"], \"DEL\":[2, \"eq\"], \"geno\":[\"no_thres\"]}\n",
    "snp_classified_del_del2 = filter_characters(del_rs334, character_value_dict)\n",
    "snp_classified_del2 = snp_classified_del_del2[[\"NWDID\", \"geno\"]]\n",
    "snp_classified_del2_filename = \"common_rs334_del2.tsv\"\n",
    "snp_classified_del2_dir_filename = os.path.join(save_dir, snp_classified_del2_filename)\n",
    "snp_classified_del2.to_csv(snp_classified_del2_dir_filename, sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Table 3\n",
    "rs11248850, rs11248850_del0, rs11248850_del12, del_rs11248850"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### DEL Ã— rs11248850"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "metadata": {},
   "outputs": [],
   "source": [
    "cohort = \"SOL\"\n",
    "load_dir = os.path.join(\"..\", \"cohort\", cohort, \"ready_data\")\n",
    "del_filename = \"common_del.tsv\"\n",
    "del_dir_filename = os.path.join(load_dir, del_filename)\n",
    "del_cn = pd.read_csv(del_dir_filename, sep=\"\\t\")\n",
    "\n",
    "rs11248850_filename = \"common_rs11248850.tsv\"\n",
    "rs11248850_dir_filename = os.path.join(load_dir, rs11248850_filename)\n",
    "rs11248850 = pd.read_csv(rs11248850_dir_filename, sep=\"\\t\")\n",
    "\n",
    "del_rs11248850 = pd.concat(objs=[del_cn, rs11248850[[\"geno\"]]], axis=1)\n",
    "del_rs11248850_filename = \"common_del-rs11248850.tsv\"\n",
    "save_dir = load_dir\n",
    "del_rs11248850_dir_filename = os.path.join(save_dir, del_rs11248850_filename)\n",
    "del_rs11248850.to_csv(del_rs11248850_dir_filename, sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Classifying rs334 by # Copies of Deletion"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "metadata": {},
   "outputs": [],
   "source": [
    "load_dir = os.path.join(\"..\", \"cohort\", cohort, \"ready_data\")\n",
    "del_rs11248850_filename = \"common_del-rs11248850.tsv\"\n",
    "del_rs11248850_dir_filename = os.path.join(load_dir, del_rs11248850_filename)\n",
    "del_rs11248850 = pd.read_csv(del_rs11248850_dir_filename, sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### DEL = 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 146,
   "metadata": {},
   "outputs": [],
   "source": [
    "save_dir = os.path.join(\"..\", \"cohort\", cohort, \"ready_data\")\n",
    "character_value_dict = {\"NWDID\":[\"no_thres\"], \"DEL\":[0, \"eq\"], \"geno\":[\"no_thres\"]}\n",
    "snp_classified_del_del0 = filter_characters(del_rs11248850, character_value_dict)\n",
    "snp_classified_del0 = snp_classified_del_del0[[\"NWDID\", \"geno\"]]\n",
    "snp_classified_del0_filename = \"common_rs11248850_del0.tsv\"\n",
    "snp_classified_del0_dir_filename = os.path.join(save_dir, snp_classified_del0_filename)\n",
    "snp_classified_del0.to_csv(snp_classified_del0_dir_filename, sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### DEL = 1, 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "metadata": {},
   "outputs": [],
   "source": [
    "save_dir = os.path.join(\"..\", \"cohort\", cohort, \"ready_data\")\n",
    "character_value_dict = {\"NWDID\":[\"no_thres\"], \"DEL\":[1, \"eq\"], \"geno\":[\"no_thres\"]}\n",
    "snp_classified_del_del1 = filter_characters(del_rs11248850, character_value_dict)\n",
    "snp_classified_del1 = snp_classified_del_del1[[\"NWDID\", \"geno\"]]\n",
    "\n",
    "character_value_dict = {\"NWDID\":[\"no_thres\"], \"DEL\":[2, \"eq\"], \"geno\":[\"no_thres\"]}\n",
    "snp_classified_del_del2 = filter_characters(del_rs11248850, character_value_dict)\n",
    "snp_classified_del2 = snp_classified_del_del2[[\"NWDID\", \"geno\"]]\n",
    "\n",
    "snp_classified_del12 = pd.concat(objs=[snp_classified_del1, snp_classified_del2], axis=0)\n",
    "\n",
    "snp_classified_del12_filename = \"common_rs11248850_del12.tsv\"\n",
    "snp_classified_del12_dir_filename = os.path.join(save_dir, snp_classified_del12_filename)\n",
    "snp_classified_del12.to_csv(snp_classified_del12_dir_filename, sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 164,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "table1 completed.\n",
      "table2 completed.\n",
      "table3 completed.\n"
     ]
    }
   ],
   "source": [
    "phenotype_list = [\"hemoglobin_mcnc_bld_1\", \"hematocrit_vfr_bld_1\", \"rbc_ncnc_bld_1\",\n",
    "                  \"mcv_entvol_rbc_1\", \"mch_entmass_rbc_1\", \"mchc_mcnc_rbc_1\", \"rdw_ratio_rbc_1\",\n",
    "                  \"neutrophil_ncnc_bld_1\", \"lymphocyte_ncnc_bld_1\", \"basophil_ncnc_bld_1\",\n",
    "                  \"eosinophil_ncnc_bld_1\", \"monocyte_ncnc_bld_1\", \"wbc_ncnc_bld_1\",\n",
    "                  \"pmv_entvol_bld_1\", \"platelet_ncnc_bld_1\", \"EGFRCKDEPI\", \"CKD\", \"microcytosis\", \"anemia\"]\n",
    "table_dict = {\"table1\":[\"cn-binary_cn0\", \"cn-binary_cn1\", \"cn-binary_cn3\", \"rs334\", \"rs33930165\"],\n",
    "              \"table2\":[\"rs334_del0\", \"rs334_del1\", \"rs334_del2\", \"del-rs334\"],\n",
    "              \"table3\":[\"rs11248850\", \"rs11248850_del0\", \"rs11248850_del12\", \"del-rs11248850\"]}\n",
    "adj_list = [\"gengrp6\", \"CENTER\", \"WEIGHT_FINAL_NORM_OVERALL\"]\n",
    "load_dir = os.path.join(\"..\", \"cohort\", cohort, \"ready_data\")\n",
    "wrap_prepare_matrix_pheno_adj_in_pheno(phenotype_list, table_dict, load_dir, adj_list)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Table 1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Recoding CN Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "metadata": {},
   "outputs": [],
   "source": [
    "cohort = \"SOL\"\n",
    "cn_orig_filename = \"common_cn.tsv\"\n",
    "cn_orig_dir = os.path.join(\"..\", \"cohort\", cohort, \"ready_data\")\n",
    "cn_orig_dir_filename = os.path.join(cn_orig_dir, cn_orig_filename)\n",
    "cn_orig_df = pd.read_csv(cn_orig_dir_filename, sep=\"\\t\")\n",
    "cn_binary_df_full = cn2binary_df(cn_orig_df)\n",
    "cn_binary_df = cn_binary_df_full[[\"NWDID\", \"cn0\", \"cn1\", \"cn3\", \"cn4\"]]\n",
    "cn_binary_df_filename = \"common_cn-binary.tsv\"\n",
    "cn_binary_df_dir_filename = os.path.join(cn_orig_dir, cn_binary_df_filename)\n",
    "cn_binary_df.to_csv(cn_binary_df_dir_filename, sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### CN Binary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 206,
   "metadata": {},
   "outputs": [],
   "source": [
    "phenotype_list = [\"hemoglobin_mcnc_bld_1\", \"hematocrit_vfr_bld_1\", \"rbc_ncnc_bld_1\",\n",
    "                  \"mcv_entvol_rbc_1\", \"mch_entmass_rbc_1\", \"mchc_mcnc_rbc_1\", \"rdw_ratio_rbc_1\",\n",
    "                  \"neutrophil_ncnc_bld_1\", \"lymphocyte_ncnc_bld_1\", \"basophil_ncnc_bld_1\",\n",
    "                  \"eosinophil_ncnc_bld_1\", \"monocyte_ncnc_bld_1\", \"wbc_ncnc_bld_1\",\n",
    "                  \"pmv_entvol_bld_1\", \"platelet_ncnc_bld_1\", \"EGFRCKDEPI\", \"CKD\", \"microcytosis\", \"anemia\"]\n",
    "load_dir = os.path.join(\"..\", \"cohort\", cohort, \"ready_data\")\n",
    "\n",
    "pheno_full_filename = \"common_pheno.tsv\"\n",
    "pheno_full_dir_filename = os.path.join(load_dir, pheno_full_filename)\n",
    "pheno_full_df = pd.read_csv(pheno_full_dir_filename, sep=\"\\t\")\n",
    "\n",
    "var_name = \"cn-binary\"\n",
    "\n",
    "var_filename = \"common_cn-binary.tsv\"\n",
    "var_dir_filename = os.path.join(load_dir, var_filename)\n",
    "var_df = pd.read_csv(var_dir_filename, sep=\"\\t\")\n",
    "\n",
    "table_name = \"table1\"\n",
    "\n",
    "kinship_filename = \"common_kinship.tsv\"\n",
    "kinship_dir_filename = os.path.join(load_dir, kinship_filename)\n",
    "kinship_df_ID = pd.read_csv(kinship_dir_filename, sep=\"\\t\")\n",
    "\n",
    "pc_filename = \"common_pc.tsv\"\n",
    "pc_dir_filename = os.path.join(load_dir, pc_filename)\n",
    "pc_df = pd.read_csv(pc_dir_filename, sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 207,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "hemoglobin_mcnc_bld_1 (1/19) completed.\n",
      "hematocrit_vfr_bld_1 (2/19) completed.\n",
      "rbc_ncnc_bld_1 (3/19) completed.\n",
      "mcv_entvol_rbc_1 (4/19) completed.\n",
      "mch_entmass_rbc_1 (5/19) completed.\n",
      "mchc_mcnc_rbc_1 (6/19) completed.\n",
      "rdw_ratio_rbc_1 (7/19) completed.\n",
      "neutrophil_ncnc_bld_1 (8/19) completed.\n",
      "lymphocyte_ncnc_bld_1 (9/19) completed.\n",
      "basophil_ncnc_bld_1 (10/19) completed.\n",
      "eosinophil_ncnc_bld_1 (11/19) completed.\n",
      "monocyte_ncnc_bld_1 (12/19) completed.\n",
      "wbc_ncnc_bld_1 (13/19) completed.\n",
      "pmv_entvol_bld_1 (14/19) completed.\n",
      "platelet_ncnc_bld_1 (15/19) completed.\n",
      "EGFRCKDEPI (16/19) completed.\n",
      "CKD (17/19) completed.\n",
      "microcytosis (18/19) completed.\n",
      "anemia (19/19) completed.\n"
     ]
    }
   ],
   "source": [
    "for phenotype, i in zip(phenotype_list, range(len(phenotype_list))):\n",
    "    prepare_matrix_pheno_agesex_in_pheno(cohort, phenotype, pheno_full_df, var_name, var_df, table_name, kinship_df_ID, pc_df)\n",
    "    print(\"%s (%d/%d) completed.\"%(phenotype, i + 1, len(phenotype_list)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### rs334"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 208,
   "metadata": {},
   "outputs": [],
   "source": [
    "phenotype_list = [\"hemoglobin_mcnc_bld_1\", \"hematocrit_vfr_bld_1\", \"rbc_ncnc_bld_1\",\n",
    "                  \"mcv_entvol_rbc_1\", \"mch_entmass_rbc_1\", \"mchc_mcnc_rbc_1\", \"rdw_ratio_rbc_1\",\n",
    "                  \"neutrophil_ncnc_bld_1\", \"lymphocyte_ncnc_bld_1\", \"basophil_ncnc_bld_1\",\n",
    "                  \"eosinophil_ncnc_bld_1\", \"monocyte_ncnc_bld_1\", \"wbc_ncnc_bld_1\",\n",
    "                  \"pmv_entvol_bld_1\", \"platelet_ncnc_bld_1\", \"EGFRCKDEPI\", \"CKD\", \"microcytosis\", \"anemia\"]\n",
    "load_dir = os.path.join(\"..\", \"cohort\", cohort, \"ready_data\")\n",
    "\n",
    "pheno_full_filename = \"common_pheno.tsv\"\n",
    "pheno_full_dir_filename = os.path.join(load_dir, pheno_full_filename)\n",
    "pheno_full_df = pd.read_csv(pheno_full_dir_filename, sep=\"\\t\")\n",
    "\n",
    "var_name = \"rs334\"\n",
    "\n",
    "var_filename = \"common_rs334.tsv\"\n",
    "var_dir_filename = os.path.join(load_dir, var_filename)\n",
    "var_df = pd.read_csv(var_dir_filename, sep=\"\\t\")\n",
    "\n",
    "table_name = \"table1\"\n",
    "\n",
    "kinship_filename = \"common_kinship.tsv\"\n",
    "kinship_dir_filename = os.path.join(load_dir, kinship_filename)\n",
    "kinship_df_ID = pd.read_csv(kinship_dir_filename, sep=\"\\t\")\n",
    "\n",
    "pc_filename = \"common_pc.tsv\"\n",
    "pc_dir_filename = os.path.join(load_dir, pc_filename)\n",
    "pc_df = pd.read_csv(pc_dir_filename, sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 209,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "hemoglobin_mcnc_bld_1 (1/19) completed.\n",
      "hematocrit_vfr_bld_1 (2/19) completed.\n",
      "rbc_ncnc_bld_1 (3/19) completed.\n",
      "mcv_entvol_rbc_1 (4/19) completed.\n",
      "mch_entmass_rbc_1 (5/19) completed.\n",
      "mchc_mcnc_rbc_1 (6/19) completed.\n",
      "rdw_ratio_rbc_1 (7/19) completed.\n",
      "neutrophil_ncnc_bld_1 (8/19) completed.\n",
      "lymphocyte_ncnc_bld_1 (9/19) completed.\n",
      "basophil_ncnc_bld_1 (10/19) completed.\n",
      "eosinophil_ncnc_bld_1 (11/19) completed.\n",
      "monocyte_ncnc_bld_1 (12/19) completed.\n",
      "wbc_ncnc_bld_1 (13/19) completed.\n",
      "pmv_entvol_bld_1 (14/19) completed.\n",
      "platelet_ncnc_bld_1 (15/19) completed.\n",
      "EGFRCKDEPI (16/19) completed.\n",
      "CKD (17/19) completed.\n",
      "microcytosis (18/19) completed.\n",
      "anemia (19/19) completed.\n"
     ]
    }
   ],
   "source": [
    "for phenotype, i in zip(phenotype_list, range(len(phenotype_list))):\n",
    "    prepare_matrix_pheno_agesex_in_pheno(cohort, phenotype, pheno_full_df, var_name, var_df, table_name, kinship_df_ID, pc_df)\n",
    "    print(\"%s (%d/%d) completed.\"%(phenotype, i + 1, len(phenotype_list)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### rs33930165"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 210,
   "metadata": {},
   "outputs": [],
   "source": [
    "phenotype_list = [\"hemoglobin_mcnc_bld_1\", \"hematocrit_vfr_bld_1\", \"rbc_ncnc_bld_1\",\n",
    "                  \"mcv_entvol_rbc_1\", \"mch_entmass_rbc_1\", \"mchc_mcnc_rbc_1\", \"rdw_ratio_rbc_1\",\n",
    "                  \"neutrophil_ncnc_bld_1\", \"lymphocyte_ncnc_bld_1\", \"basophil_ncnc_bld_1\",\n",
    "                  \"eosinophil_ncnc_bld_1\", \"monocyte_ncnc_bld_1\", \"wbc_ncnc_bld_1\",\n",
    "                  \"pmv_entvol_bld_1\", \"platelet_ncnc_bld_1\", \"EGFRCKDEPI\", \"CKD\", \"microcytosis\", \"anemia\"]\n",
    "load_dir = os.path.join(\"..\", \"cohort\", cohort, \"ready_data\")\n",
    "\n",
    "pheno_full_filename = \"common_pheno.tsv\"\n",
    "pheno_full_dir_filename = os.path.join(load_dir, pheno_full_filename)\n",
    "pheno_full_df = pd.read_csv(pheno_full_dir_filename, sep=\"\\t\")\n",
    "\n",
    "var_name = \"rs33930165\"\n",
    "\n",
    "var_filename = \"common_rs33930165.tsv\"\n",
    "var_dir_filename = os.path.join(load_dir, var_filename)\n",
    "var_df = pd.read_csv(var_dir_filename, sep=\"\\t\")\n",
    "\n",
    "table_name = \"table1\"\n",
    "\n",
    "kinship_filename = \"common_kinship.tsv\"\n",
    "kinship_dir_filename = os.path.join(load_dir, kinship_filename)\n",
    "kinship_df_ID = pd.read_csv(kinship_dir_filename, sep=\"\\t\")\n",
    "\n",
    "pc_filename = \"common_pc.tsv\"\n",
    "pc_dir_filename = os.path.join(load_dir, pc_filename)\n",
    "pc_df = pd.read_csv(pc_dir_filename, sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 211,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "hemoglobin_mcnc_bld_1 (1/19) completed.\n",
      "hematocrit_vfr_bld_1 (2/19) completed.\n",
      "rbc_ncnc_bld_1 (3/19) completed.\n",
      "mcv_entvol_rbc_1 (4/19) completed.\n",
      "mch_entmass_rbc_1 (5/19) completed.\n",
      "mchc_mcnc_rbc_1 (6/19) completed.\n",
      "rdw_ratio_rbc_1 (7/19) completed.\n",
      "neutrophil_ncnc_bld_1 (8/19) completed.\n",
      "lymphocyte_ncnc_bld_1 (9/19) completed.\n",
      "basophil_ncnc_bld_1 (10/19) completed.\n",
      "eosinophil_ncnc_bld_1 (11/19) completed.\n",
      "monocyte_ncnc_bld_1 (12/19) completed.\n",
      "wbc_ncnc_bld_1 (13/19) completed.\n",
      "pmv_entvol_bld_1 (14/19) completed.\n",
      "platelet_ncnc_bld_1 (15/19) completed.\n",
      "EGFRCKDEPI (16/19) completed.\n",
      "CKD (17/19) completed.\n",
      "microcytosis (18/19) completed.\n",
      "anemia (19/19) completed.\n"
     ]
    }
   ],
   "source": [
    "for phenotype, i in zip(phenotype_list, range(len(phenotype_list))):\n",
    "    prepare_matrix_pheno_agesex_in_pheno(cohort, phenotype, pheno_full_df, var_name, var_df, table_name, kinship_df_ID, pc_df)\n",
    "    print(\"%s (%d/%d) completed.\"%(phenotype, i + 1, len(phenotype_list)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Table 2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### CN to DEL"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 212,
   "metadata": {},
   "outputs": [],
   "source": [
    "cohort = \"SOL\"\n",
    "load_dir = os.path.join(\"..\", \"cohort\", cohort, \"ready_data\")\n",
    "cn_filename = \"common_cn.tsv\"\n",
    "cn_dir_filename = os.path.join(load_dir, cn_filename)\n",
    "cn = pd.read_csv(cn_dir_filename, sep=\"\\t\")\n",
    "del_cn = cn2del(cn)\n",
    "del_filename = \"common_del.tsv\"\n",
    "del_dir_filename = os.path.join(load_dir, del_filename)\n",
    "del_cn.to_csv(del_dir_filename, sep=\"\\t\", header=True, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### DEL Ã— rs334"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 213,
   "metadata": {},
   "outputs": [],
   "source": [
    "load_dir = os.path.join(\"..\", \"cohort\", cohort, \"ready_data\")\n",
    "del_filename = \"common_del.tsv\"\n",
    "del_dir_filename = os.path.join(load_dir, del_filename)\n",
    "del_cn = pd.read_csv(del_dir_filename, sep=\"\\t\")\n",
    "\n",
    "rs334_filename = \"common_rs334.tsv\"\n",
    "rs334_dir_filename = os.path.join(load_dir, rs334_filename)\n",
    "rs334 = pd.read_csv(rs334_dir_filename, sep=\"\\t\")\n",
    "\n",
    "del_rs334 = pd.concat(objs=[del_cn, rs334[[\"geno\"]]], axis=1)\n",
    "del_rs334_filename = \"common_del-rs334.tsv\"\n",
    "save_dir = load_dir\n",
    "del_rs334_dir_filename = os.path.join(save_dir, del_rs334_filename)\n",
    "del_rs334.to_csv(del_rs334_dir_filename, sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Classifying rs334 by Number of Copies of Deletion"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 214,
   "metadata": {},
   "outputs": [],
   "source": [
    "load_dir = os.path.join(\"..\", \"cohort\", cohort, \"ready_data\")\n",
    "del_rs334_filename = \"common_del-rs334.tsv\"\n",
    "del_rs334_dir_filename = os.path.join(load_dir, del_rs334_filename)\n",
    "del_rs334 = pd.read_csv(del_rs334_dir_filename, sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### No Copies of deletion, a.k.a. CN = 2, 3, 4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 215,
   "metadata": {},
   "outputs": [],
   "source": [
    "save_dir = os.path.join(\"..\", \"cohort\", cohort, \"ready_data\")\n",
    "character_value_dict = {\"NWDID\":[\"no_thres\"], \"DEL\":[0, \"eq\"], \"geno\":[\"no_thres\"]}\n",
    "snp_classified_del_del0 = filter_characters(del_rs334, character_value_dict)\n",
    "snp_classified_del0 = snp_classified_del_del0[[\"NWDID\", \"geno\"]]\n",
    "snp_classified_del0_filename = \"common_rs334_del0.tsv\"\n",
    "snp_classified_del0_dir_filename = os.path.join(save_dir, snp_classified_del0_filename)\n",
    "snp_classified_del0.to_csv(snp_classified_del0_dir_filename, sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 1 Copy of deletion, a.k.a. CN = 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 216,
   "metadata": {},
   "outputs": [],
   "source": [
    "save_dir = os.path.join(\"..\", \"cohort\", cohort, \"ready_data\")\n",
    "character_value_dict = {\"NWDID\":[\"no_thres\"], \"DEL\":[1, \"eq\"], \"geno\":[\"no_thres\"]}\n",
    "snp_classified_del_del1 = filter_characters(del_rs334, character_value_dict)\n",
    "snp_classified_del1 = snp_classified_del_del1[[\"NWDID\", \"geno\"]]\n",
    "snp_classified_del1_filename = \"common_rs334_del1.tsv\"\n",
    "snp_classified_del1_dir_filename = os.path.join(save_dir, snp_classified_del1_filename)\n",
    "snp_classified_del1.to_csv(snp_classified_del1_dir_filename, sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 2 Copies of deletion, a.k.a. CN = 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 217,
   "metadata": {},
   "outputs": [],
   "source": [
    "save_dir = os.path.join(\"..\", \"cohort\", cohort, \"ready_data\")\n",
    "character_value_dict = {\"NWDID\":[\"no_thres\"], \"DEL\":[2, \"eq\"], \"geno\":[\"no_thres\"]}\n",
    "snp_classified_del_del2 = filter_characters(del_rs334, character_value_dict)\n",
    "snp_classified_del2 = snp_classified_del_del2[[\"NWDID\", \"geno\"]]\n",
    "snp_classified_del2_filename = \"common_rs334_del2.tsv\"\n",
    "snp_classified_del2_dir_filename = os.path.join(save_dir, snp_classified_del2_filename)\n",
    "snp_classified_del2.to_csv(snp_classified_del2_dir_filename, sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### rs334 when DEL = 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 218,
   "metadata": {},
   "outputs": [],
   "source": [
    "phenotype_list = [\"hemoglobin_mcnc_bld_1\", \"hematocrit_vfr_bld_1\", \"rbc_ncnc_bld_1\",\n",
    "                  \"mcv_entvol_rbc_1\", \"mch_entmass_rbc_1\", \"mchc_mcnc_rbc_1\", \"rdw_ratio_rbc_1\",\n",
    "                  \"neutrophil_ncnc_bld_1\", \"lymphocyte_ncnc_bld_1\", \"basophil_ncnc_bld_1\",\n",
    "                  \"eosinophil_ncnc_bld_1\", \"monocyte_ncnc_bld_1\", \"wbc_ncnc_bld_1\",\n",
    "                  \"pmv_entvol_bld_1\", \"platelet_ncnc_bld_1\", \"EGFRCKDEPI\", \"CKD\", \"microcytosis\", \"anemia\"]\n",
    "load_dir = os.path.join(\"..\", \"cohort\", cohort, \"ready_data\")\n",
    "\n",
    "pheno_full_filename = \"common_pheno.tsv\"\n",
    "pheno_full_dir_filename = os.path.join(load_dir, pheno_full_filename)\n",
    "pheno_full_df = pd.read_csv(pheno_full_dir_filename, sep=\"\\t\")\n",
    "\n",
    "var_name = \"rs334_del0\"\n",
    "\n",
    "var_filename = \"common_rs334_del0.tsv\"\n",
    "var_dir_filename = os.path.join(load_dir, var_filename)\n",
    "var_df = pd.read_csv(var_dir_filename, sep=\"\\t\")\n",
    "\n",
    "table_name = \"table2\"\n",
    "\n",
    "kinship_filename = \"common_kinship.tsv\"\n",
    "kinship_dir_filename = os.path.join(load_dir, kinship_filename)\n",
    "kinship_df_ID = pd.read_csv(kinship_dir_filename, sep=\"\\t\")\n",
    "\n",
    "pc_filename = \"common_pc.tsv\"\n",
    "pc_dir_filename = os.path.join(load_dir, pc_filename)\n",
    "pc_df = pd.read_csv(pc_dir_filename, sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 219,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "hemoglobin_mcnc_bld_1 (1/19) completed.\n",
      "hematocrit_vfr_bld_1 (2/19) completed.\n",
      "rbc_ncnc_bld_1 (3/19) completed.\n",
      "mcv_entvol_rbc_1 (4/19) completed.\n",
      "mch_entmass_rbc_1 (5/19) completed.\n",
      "mchc_mcnc_rbc_1 (6/19) completed.\n",
      "rdw_ratio_rbc_1 (7/19) completed.\n",
      "neutrophil_ncnc_bld_1 (8/19) completed.\n",
      "lymphocyte_ncnc_bld_1 (9/19) completed.\n",
      "basophil_ncnc_bld_1 (10/19) completed.\n",
      "eosinophil_ncnc_bld_1 (11/19) completed.\n",
      "monocyte_ncnc_bld_1 (12/19) completed.\n",
      "wbc_ncnc_bld_1 (13/19) completed.\n",
      "pmv_entvol_bld_1 (14/19) completed.\n",
      "platelet_ncnc_bld_1 (15/19) completed.\n",
      "EGFRCKDEPI (16/19) completed.\n",
      "CKD (17/19) completed.\n",
      "microcytosis (18/19) completed.\n",
      "anemia (19/19) completed.\n"
     ]
    }
   ],
   "source": [
    "for phenotype, i in zip(phenotype_list, range(len(phenotype_list))):\n",
    "    prepare_matrix_pheno_agesex_in_pheno(cohort, phenotype, pheno_full_df, var_name, var_df, table_name, kinship_df_ID, pc_df)\n",
    "    print(\"%s (%d/%d) completed.\"%(phenotype, i + 1, len(phenotype_list)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### rs334 when DEL = 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 220,
   "metadata": {},
   "outputs": [],
   "source": [
    "phenotype_list = [\"hemoglobin_mcnc_bld_1\", \"hematocrit_vfr_bld_1\", \"rbc_ncnc_bld_1\",\n",
    "                  \"mcv_entvol_rbc_1\", \"mch_entmass_rbc_1\", \"mchc_mcnc_rbc_1\", \"rdw_ratio_rbc_1\",\n",
    "                  \"neutrophil_ncnc_bld_1\", \"lymphocyte_ncnc_bld_1\", \"basophil_ncnc_bld_1\",\n",
    "                  \"eosinophil_ncnc_bld_1\", \"monocyte_ncnc_bld_1\", \"wbc_ncnc_bld_1\",\n",
    "                  \"pmv_entvol_bld_1\", \"platelet_ncnc_bld_1\", \"EGFRCKDEPI\", \"CKD\", \"microcytosis\", \"anemia\"]\n",
    "load_dir = os.path.join(\"..\", \"cohort\", cohort, \"ready_data\")\n",
    "\n",
    "pheno_full_filename = \"common_pheno.tsv\"\n",
    "pheno_full_dir_filename = os.path.join(load_dir, pheno_full_filename)\n",
    "pheno_full_df = pd.read_csv(pheno_full_dir_filename, sep=\"\\t\")\n",
    "\n",
    "var_name = \"rs334_del1\"\n",
    "\n",
    "var_filename = \"common_rs334_del1.tsv\"\n",
    "var_dir_filename = os.path.join(load_dir, var_filename)\n",
    "var_df = pd.read_csv(var_dir_filename, sep=\"\\t\")\n",
    "\n",
    "table_name = \"table2\"\n",
    "\n",
    "kinship_filename = \"common_kinship.tsv\"\n",
    "kinship_dir_filename = os.path.join(load_dir, kinship_filename)\n",
    "kinship_df_ID = pd.read_csv(kinship_dir_filename, sep=\"\\t\")\n",
    "\n",
    "pc_filename = \"common_pc.tsv\"\n",
    "pc_dir_filename = os.path.join(load_dir, pc_filename)\n",
    "pc_df = pd.read_csv(pc_dir_filename, sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 221,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "hemoglobin_mcnc_bld_1 (1/19) completed.\n",
      "hematocrit_vfr_bld_1 (2/19) completed.\n",
      "rbc_ncnc_bld_1 (3/19) completed.\n",
      "mcv_entvol_rbc_1 (4/19) completed.\n",
      "mch_entmass_rbc_1 (5/19) completed.\n",
      "mchc_mcnc_rbc_1 (6/19) completed.\n",
      "rdw_ratio_rbc_1 (7/19) completed.\n",
      "neutrophil_ncnc_bld_1 (8/19) completed.\n",
      "lymphocyte_ncnc_bld_1 (9/19) completed.\n",
      "basophil_ncnc_bld_1 (10/19) completed.\n",
      "eosinophil_ncnc_bld_1 (11/19) completed.\n",
      "monocyte_ncnc_bld_1 (12/19) completed.\n",
      "wbc_ncnc_bld_1 (13/19) completed.\n",
      "pmv_entvol_bld_1 (14/19) completed.\n",
      "platelet_ncnc_bld_1 (15/19) completed.\n",
      "EGFRCKDEPI (16/19) completed.\n",
      "CKD (17/19) completed.\n",
      "microcytosis (18/19) completed.\n",
      "anemia (19/19) completed.\n"
     ]
    }
   ],
   "source": [
    "for phenotype, i in zip(phenotype_list, range(len(phenotype_list))):\n",
    "    prepare_matrix_pheno_agesex_in_pheno(cohort, phenotype, pheno_full_df, var_name, var_df, table_name, kinship_df_ID, pc_df)\n",
    "    print(\"%s (%d/%d) completed.\"%(phenotype, i + 1, len(phenotype_list)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### rs334 when DEL = 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 222,
   "metadata": {},
   "outputs": [],
   "source": [
    "phenotype_list = [\"hemoglobin_mcnc_bld_1\", \"hematocrit_vfr_bld_1\", \"rbc_ncnc_bld_1\",\n",
    "                  \"mcv_entvol_rbc_1\", \"mch_entmass_rbc_1\", \"mchc_mcnc_rbc_1\", \"rdw_ratio_rbc_1\",\n",
    "                  \"neutrophil_ncnc_bld_1\", \"lymphocyte_ncnc_bld_1\", \"basophil_ncnc_bld_1\",\n",
    "                  \"eosinophil_ncnc_bld_1\", \"monocyte_ncnc_bld_1\", \"wbc_ncnc_bld_1\",\n",
    "                  \"pmv_entvol_bld_1\", \"platelet_ncnc_bld_1\", \"EGFRCKDEPI\", \"CKD\", \"microcytosis\", \"anemia\"]\n",
    "load_dir = os.path.join(\"..\", \"cohort\", cohort, \"ready_data\")\n",
    "\n",
    "pheno_full_filename = \"common_pheno.tsv\"\n",
    "pheno_full_dir_filename = os.path.join(load_dir, pheno_full_filename)\n",
    "pheno_full_df = pd.read_csv(pheno_full_dir_filename, sep=\"\\t\")\n",
    "\n",
    "var_name = \"rs334_del2\"\n",
    "\n",
    "var_filename = \"common_rs334_del2.tsv\"\n",
    "var_dir_filename = os.path.join(load_dir, var_filename)\n",
    "var_df = pd.read_csv(var_dir_filename, sep=\"\\t\")\n",
    "\n",
    "table_name = \"table2\"\n",
    "\n",
    "kinship_filename = \"common_kinship.tsv\"\n",
    "kinship_dir_filename = os.path.join(load_dir, kinship_filename)\n",
    "kinship_df_ID = pd.read_csv(kinship_dir_filename, sep=\"\\t\")\n",
    "\n",
    "pc_filename = \"common_pc.tsv\"\n",
    "pc_dir_filename = os.path.join(load_dir, pc_filename)\n",
    "pc_df = pd.read_csv(pc_dir_filename, sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 223,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "hemoglobin_mcnc_bld_1 (1/19) completed.\n",
      "hematocrit_vfr_bld_1 (2/19) completed.\n",
      "rbc_ncnc_bld_1 (3/19) completed.\n",
      "mcv_entvol_rbc_1 (4/19) completed.\n",
      "mch_entmass_rbc_1 (5/19) completed.\n",
      "mchc_mcnc_rbc_1 (6/19) completed.\n",
      "rdw_ratio_rbc_1 (7/19) completed.\n",
      "neutrophil_ncnc_bld_1 (8/19) completed.\n",
      "lymphocyte_ncnc_bld_1 (9/19) completed.\n",
      "basophil_ncnc_bld_1 (10/19) completed.\n",
      "eosinophil_ncnc_bld_1 (11/19) completed.\n",
      "monocyte_ncnc_bld_1 (12/19) completed.\n",
      "wbc_ncnc_bld_1 (13/19) completed.\n",
      "pmv_entvol_bld_1 (14/19) completed.\n",
      "platelet_ncnc_bld_1 (15/19) completed.\n",
      "EGFRCKDEPI (16/19) completed.\n",
      "CKD (17/19) completed.\n",
      "microcytosis (18/19) completed.\n",
      "anemia (19/19) completed.\n"
     ]
    }
   ],
   "source": [
    "for phenotype, i in zip(phenotype_list, range(len(phenotype_list))):\n",
    "    prepare_matrix_pheno_agesex_in_pheno(cohort, phenotype, pheno_full_df, var_name, var_df, table_name, kinship_df_ID, pc_df)\n",
    "    print(\"%s (%d/%d) completed.\"%(phenotype, i + 1, len(phenotype_list)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### GENxGEN Interaction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 224,
   "metadata": {},
   "outputs": [],
   "source": [
    "phenotype_list = [\"hemoglobin_mcnc_bld_1\", \"hematocrit_vfr_bld_1\", \"rbc_ncnc_bld_1\",\n",
    "                  \"mcv_entvol_rbc_1\", \"mch_entmass_rbc_1\", \"mchc_mcnc_rbc_1\", \"rdw_ratio_rbc_1\",\n",
    "                  \"neutrophil_ncnc_bld_1\", \"lymphocyte_ncnc_bld_1\", \"basophil_ncnc_bld_1\",\n",
    "                  \"eosinophil_ncnc_bld_1\", \"monocyte_ncnc_bld_1\", \"wbc_ncnc_bld_1\",\n",
    "                  \"pmv_entvol_bld_1\", \"platelet_ncnc_bld_1\", \"EGFRCKDEPI\", \"CKD\", \"microcytosis\", \"anemia\"]\n",
    "load_dir = os.path.join(\"..\", \"cohort\", cohort, \"ready_data\")\n",
    "\n",
    "pheno_full_filename = \"common_pheno.tsv\"\n",
    "pheno_full_dir_filename = os.path.join(load_dir, pheno_full_filename)\n",
    "pheno_full_df = pd.read_csv(pheno_full_dir_filename, sep=\"\\t\")\n",
    "\n",
    "var_name = \"del_rs334\"\n",
    "\n",
    "var_filename = \"common_del-rs334.tsv\"\n",
    "var_dir_filename = os.path.join(load_dir, var_filename)\n",
    "var_df = pd.read_csv(var_dir_filename, sep=\"\\t\")\n",
    "\n",
    "table_name = \"table2\"\n",
    "\n",
    "kinship_filename = \"common_kinship.tsv\"\n",
    "kinship_dir_filename = os.path.join(load_dir, kinship_filename)\n",
    "kinship_df_ID = pd.read_csv(kinship_dir_filename, sep=\"\\t\")\n",
    "\n",
    "pc_filename = \"common_pc.tsv\"\n",
    "pc_dir_filename = os.path.join(load_dir, pc_filename)\n",
    "pc_df = pd.read_csv(pc_dir_filename, sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 225,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "hemoglobin_mcnc_bld_1 (1/19) completed.\n",
      "hematocrit_vfr_bld_1 (2/19) completed.\n",
      "rbc_ncnc_bld_1 (3/19) completed.\n",
      "mcv_entvol_rbc_1 (4/19) completed.\n",
      "mch_entmass_rbc_1 (5/19) completed.\n",
      "mchc_mcnc_rbc_1 (6/19) completed.\n",
      "rdw_ratio_rbc_1 (7/19) completed.\n",
      "neutrophil_ncnc_bld_1 (8/19) completed.\n",
      "lymphocyte_ncnc_bld_1 (9/19) completed.\n",
      "basophil_ncnc_bld_1 (10/19) completed.\n",
      "eosinophil_ncnc_bld_1 (11/19) completed.\n",
      "monocyte_ncnc_bld_1 (12/19) completed.\n",
      "wbc_ncnc_bld_1 (13/19) completed.\n",
      "pmv_entvol_bld_1 (14/19) completed.\n",
      "platelet_ncnc_bld_1 (15/19) completed.\n",
      "EGFRCKDEPI (16/19) completed.\n",
      "CKD (17/19) completed.\n",
      "microcytosis (18/19) completed.\n",
      "anemia (19/19) completed.\n"
     ]
    }
   ],
   "source": [
    "for phenotype, i in zip(phenotype_list, range(len(phenotype_list))):\n",
    "    prepare_matrix_pheno_agesex_in_pheno(cohort, phenotype, pheno_full_df, var_name, var_df, table_name, kinship_df_ID, pc_df)\n",
    "    print(\"%s (%d/%d) completed.\"%(phenotype, i + 1, len(phenotype_list)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Table 3"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### CN Ã— rs11248850"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 226,
   "metadata": {},
   "outputs": [],
   "source": [
    "cohort = \"SOL\"\n",
    "load_dir = os.path.join(\"..\", \"cohort\", cohort, \"ready_data\")\n",
    "cn_filename = \"common_cn.tsv\"\n",
    "cn_dir_filename = os.path.join(load_dir, cn_filename)\n",
    "cn = pd.read_csv(cn_dir_filename, sep=\"\\t\")\n",
    "\n",
    "rs11248850_filename = \"common_rs11248850.tsv\"\n",
    "rs11248850_dir_filename = os.path.join(load_dir, rs11248850_filename)\n",
    "rs11248850 = pd.read_csv(rs11248850_dir_filename, sep=\"\\t\")\n",
    "\n",
    "cn_rs11248850 = pd.concat(objs=[cn, rs11248850[[\"geno\"]]], axis=1)\n",
    "cn_rs11248850_filename = \"common_cn-rs11248850.tsv\"\n",
    "save_dir = load_dir\n",
    "cn_rs11248850_dir_filename = os.path.join(save_dir, cn_rs11248850_filename)\n",
    "cn_rs11248850.to_csv(cn_rs11248850_dir_filename, sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Classifying rs11248850 by # Copy Numbers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 227,
   "metadata": {},
   "outputs": [],
   "source": [
    "load_dir = os.path.join(\"..\", \"cohort\", cohort, \"ready_data\")\n",
    "cn_rs11248850_filename = \"common_cn-rs11248850.tsv\"\n",
    "cn_rs11248850_dir_filename = os.path.join(load_dir, cn_rs11248850_filename)\n",
    "cn_rs11248850 = pd.read_csv(cn_rs11248850_dir_filename, sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### CN = 3, 4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 228,
   "metadata": {},
   "outputs": [],
   "source": [
    "save_dir = os.path.join(\"..\", \"cohort\", cohort, \"ready_data\")\n",
    "character_value_dict = {\"NWDID\":[\"no_thres\"], \"CN\":[3, \"eq\"], \"geno\":[\"no_thres\"]}\n",
    "snp_classified_cn_cn3 = filter_characters(cn_rs11248850, character_value_dict)\n",
    "snp_classified_cn3 = snp_classified_cn_cn3[[\"NWDID\", \"geno\"]]\n",
    "\n",
    "character_value_dict = {\"NWDID\":[\"no_thres\"], \"CN\":[4, \"eq\"], \"geno\":[\"no_thres\"]}\n",
    "snp_classified_cn_cn4 = filter_characters(cn_rs11248850, character_value_dict)\n",
    "snp_classified_cn4 = snp_classified_cn_cn4[[\"NWDID\", \"geno\"]]\n",
    "\n",
    "snp_classified_cn34 = pd.concat(objs=[snp_classified_cn3, snp_classified_cn4], axis=0)\n",
    "\n",
    "snp_classified_cn34_filename = \"common_rs11248850_cn34.tsv\"\n",
    "snp_classified_cn34_dir_filename = os.path.join(save_dir, snp_classified_cn34_filename)\n",
    "snp_classified_cn34.to_csv(snp_classified_cn34_dir_filename, sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### CN = 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 229,
   "metadata": {},
   "outputs": [],
   "source": [
    "save_dir = os.path.join(\"..\", \"cohort\", cohort, \"ready_data\")\n",
    "character_value_dict = {\"NWDID\":[\"no_thres\"], \"CN\":[2, \"eq\"], \"geno\":[\"no_thres\"]}\n",
    "snp_classified_cn_cn2 = filter_characters(cn_rs11248850, character_value_dict)\n",
    "snp_classified_cn2 = snp_classified_cn_cn2[[\"NWDID\", \"geno\"]]\n",
    "snp_classified_cn2_filename = \"common_rs11248850_cn2.tsv\"\n",
    "snp_classified_cn2_dir_filename = os.path.join(save_dir, snp_classified_cn2_filename)\n",
    "snp_classified_cn2.to_csv(snp_classified_cn2_dir_filename, sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### DEL Ã— rs11248850"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 230,
   "metadata": {},
   "outputs": [],
   "source": [
    "cohort = \"SOL\"\n",
    "load_dir = os.path.join(\"..\", \"cohort\", cohort, \"ready_data\")\n",
    "del_filename = \"common_del.tsv\"\n",
    "del_dir_filename = os.path.join(load_dir, del_filename)\n",
    "del_cn = pd.read_csv(del_dir_filename, sep=\"\\t\")\n",
    "\n",
    "rs11248850_filename = \"common_rs11248850.tsv\"\n",
    "rs11248850_dir_filename = os.path.join(load_dir, rs11248850_filename)\n",
    "rs11248850 = pd.read_csv(rs11248850_dir_filename, sep=\"\\t\")\n",
    "\n",
    "del_rs11248850 = pd.concat(objs=[del_cn, rs11248850[[\"geno\"]]], axis=1)\n",
    "del_rs11248850_filename = \"common_del-rs11248850.tsv\"\n",
    "save_dir = load_dir\n",
    "del_rs11248850_dir_filename = os.path.join(save_dir, del_rs11248850_filename)\n",
    "del_rs11248850.to_csv(del_rs11248850_dir_filename, sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Classifying rs11248850 by # Copies of Deletion"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 231,
   "metadata": {},
   "outputs": [],
   "source": [
    "load_dir = os.path.join(\"..\", \"cohort\", cohort, \"ready_data\")\n",
    "del_rs11248850_filename = \"common_del-rs11248850.tsv\"\n",
    "del_rs11248850_dir_filename = os.path.join(load_dir, del_rs11248850_filename)\n",
    "del_rs11248850 = pd.read_csv(del_rs11248850_dir_filename, sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### DEL = 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 232,
   "metadata": {},
   "outputs": [],
   "source": [
    "save_dir = os.path.join(\"..\", \"cohort\", cohort, \"ready_data\")\n",
    "character_value_dict = {\"NWDID\":[\"no_thres\"], \"DEL\":[0, \"eq\"], \"geno\":[\"no_thres\"]}\n",
    "snp_classified_del_del0 = filter_characters(del_rs11248850, character_value_dict)\n",
    "snp_classified_del0 = snp_classified_del_del0[[\"NWDID\", \"geno\"]]\n",
    "snp_classified_del0_filename = \"common_rs11248850_del0.tsv\"\n",
    "snp_classified_del0_dir_filename = os.path.join(save_dir, snp_classified_del0_filename)\n",
    "snp_classified_del0.to_csv(snp_classified_del0_dir_filename, sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### DEL = 1, 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 233,
   "metadata": {},
   "outputs": [],
   "source": [
    "save_dir = os.path.join(\"..\", \"cohort\", cohort, \"ready_data\")\n",
    "character_value_dict = {\"NWDID\":[\"no_thres\"], \"DEL\":[1, \"eq\"], \"geno\":[\"no_thres\"]}\n",
    "snp_classified_del_del1 = filter_characters(del_rs11248850, character_value_dict)\n",
    "snp_classified_del1 = snp_classified_del_del1[[\"NWDID\", \"geno\"]]\n",
    "\n",
    "character_value_dict = {\"NWDID\":[\"no_thres\"], \"DEL\":[2, \"eq\"], \"geno\":[\"no_thres\"]}\n",
    "snp_classified_del_del2 = filter_characters(del_rs11248850, character_value_dict)\n",
    "snp_classified_del2 = snp_classified_del_del2[[\"NWDID\", \"geno\"]]\n",
    "\n",
    "snp_classified_del12 = pd.concat(objs=[snp_classified_del1, snp_classified_del2], axis=0)\n",
    "\n",
    "snp_classified_del12_filename = \"common_rs11248850_del12.tsv\"\n",
    "snp_classified_del12_dir_filename = os.path.join(save_dir, snp_classified_del12_filename)\n",
    "snp_classified_del12.to_csv(snp_classified_del12_dir_filename, sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### rs11248850 when CN = 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 234,
   "metadata": {},
   "outputs": [],
   "source": [
    "phenotype_list = [\"hemoglobin_mcnc_bld_1\", \"hematocrit_vfr_bld_1\", \"rbc_ncnc_bld_1\",\n",
    "                  \"mcv_entvol_rbc_1\", \"mch_entmass_rbc_1\", \"mchc_mcnc_rbc_1\", \"rdw_ratio_rbc_1\",\n",
    "                  \"neutrophil_ncnc_bld_1\", \"lymphocyte_ncnc_bld_1\", \"basophil_ncnc_bld_1\",\n",
    "                  \"eosinophil_ncnc_bld_1\", \"monocyte_ncnc_bld_1\", \"wbc_ncnc_bld_1\",\n",
    "                  \"pmv_entvol_bld_1\", \"platelet_ncnc_bld_1\", \"EGFRCKDEPI\", \"CKD\", \"microcytosis\", \"anemia\"]\n",
    "load_dir = os.path.join(\"..\", \"cohort\", cohort, \"ready_data\")\n",
    "\n",
    "pheno_full_filename = \"common_pheno.tsv\"\n",
    "pheno_full_dir_filename = os.path.join(load_dir, pheno_full_filename)\n",
    "pheno_full_df = pd.read_csv(pheno_full_dir_filename, sep=\"\\t\")\n",
    "\n",
    "var_name = \"rs11248850_cn2\"\n",
    "\n",
    "var_filename = \"common_rs11248850_cn2.tsv\"\n",
    "var_dir_filename = os.path.join(load_dir, var_filename)\n",
    "var_df = pd.read_csv(var_dir_filename, sep=\"\\t\")\n",
    "\n",
    "table_name = \"table3\"\n",
    "\n",
    "kinship_filename = \"common_kinship.tsv\"\n",
    "kinship_dir_filename = os.path.join(load_dir, kinship_filename)\n",
    "kinship_df_ID = pd.read_csv(kinship_dir_filename, sep=\"\\t\")\n",
    "\n",
    "pc_filename = \"common_pc.tsv\"\n",
    "pc_dir_filename = os.path.join(load_dir, pc_filename)\n",
    "pc_df = pd.read_csv(pc_dir_filename, sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 235,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "hemoglobin_mcnc_bld_1 (1/19) completed.\n",
      "hematocrit_vfr_bld_1 (2/19) completed.\n",
      "rbc_ncnc_bld_1 (3/19) completed.\n",
      "mcv_entvol_rbc_1 (4/19) completed.\n",
      "mch_entmass_rbc_1 (5/19) completed.\n",
      "mchc_mcnc_rbc_1 (6/19) completed.\n",
      "rdw_ratio_rbc_1 (7/19) completed.\n",
      "neutrophil_ncnc_bld_1 (8/19) completed.\n",
      "lymphocyte_ncnc_bld_1 (9/19) completed.\n",
      "basophil_ncnc_bld_1 (10/19) completed.\n",
      "eosinophil_ncnc_bld_1 (11/19) completed.\n",
      "monocyte_ncnc_bld_1 (12/19) completed.\n",
      "wbc_ncnc_bld_1 (13/19) completed.\n",
      "pmv_entvol_bld_1 (14/19) completed.\n",
      "platelet_ncnc_bld_1 (15/19) completed.\n",
      "EGFRCKDEPI (16/19) completed.\n",
      "CKD (17/19) completed.\n",
      "microcytosis (18/19) completed.\n",
      "anemia (19/19) completed.\n"
     ]
    }
   ],
   "source": [
    "for phenotype, i in zip(phenotype_list, range(len(phenotype_list))):\n",
    "    prepare_matrix_pheno_agesex_in_pheno(cohort, phenotype, pheno_full_df, var_name, var_df, table_name, kinship_df_ID, pc_df)\n",
    "    print(\"%s (%d/%d) completed.\"%(phenotype, i + 1, len(phenotype_list)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### rs11248850 when CN = 3, 4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 236,
   "metadata": {},
   "outputs": [],
   "source": [
    "phenotype_list = [\"hemoglobin_mcnc_bld_1\", \"hematocrit_vfr_bld_1\", \"rbc_ncnc_bld_1\",\n",
    "                  \"mcv_entvol_rbc_1\", \"mch_entmass_rbc_1\", \"mchc_mcnc_rbc_1\", \"rdw_ratio_rbc_1\",\n",
    "                  \"neutrophil_ncnc_bld_1\", \"lymphocyte_ncnc_bld_1\", \"basophil_ncnc_bld_1\",\n",
    "                  \"eosinophil_ncnc_bld_1\", \"monocyte_ncnc_bld_1\", \"wbc_ncnc_bld_1\",\n",
    "                  \"pmv_entvol_bld_1\", \"platelet_ncnc_bld_1\", \"EGFRCKDEPI\", \"CKD\", \"microcytosis\", \"anemia\"]\n",
    "load_dir = os.path.join(\"..\", \"cohort\", cohort, \"ready_data\")\n",
    "\n",
    "pheno_full_filename = \"common_pheno.tsv\"\n",
    "pheno_full_dir_filename = os.path.join(load_dir, pheno_full_filename)\n",
    "pheno_full_df = pd.read_csv(pheno_full_dir_filename, sep=\"\\t\")\n",
    "\n",
    "var_name = \"rs11248850_cn34\"\n",
    "\n",
    "var_filename = \"common_rs11248850_cn34.tsv\"\n",
    "var_dir_filename = os.path.join(load_dir, var_filename)\n",
    "var_df = pd.read_csv(var_dir_filename, sep=\"\\t\")\n",
    "\n",
    "table_name = \"table3\"\n",
    "\n",
    "kinship_filename = \"common_kinship.tsv\"\n",
    "kinship_dir_filename = os.path.join(load_dir, kinship_filename)\n",
    "kinship_df_ID = pd.read_csv(kinship_dir_filename, sep=\"\\t\")\n",
    "\n",
    "pc_filename = \"common_pc.tsv\"\n",
    "pc_dir_filename = os.path.join(load_dir, pc_filename)\n",
    "pc_df = pd.read_csv(pc_dir_filename, sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 237,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "hemoglobin_mcnc_bld_1 (1/19) completed.\n",
      "hematocrit_vfr_bld_1 (2/19) completed.\n",
      "rbc_ncnc_bld_1 (3/19) completed.\n",
      "mcv_entvol_rbc_1 (4/19) completed.\n",
      "mch_entmass_rbc_1 (5/19) completed.\n",
      "mchc_mcnc_rbc_1 (6/19) completed.\n",
      "rdw_ratio_rbc_1 (7/19) completed.\n",
      "neutrophil_ncnc_bld_1 (8/19) completed.\n",
      "lymphocyte_ncnc_bld_1 (9/19) completed.\n",
      "basophil_ncnc_bld_1 (10/19) completed.\n",
      "eosinophil_ncnc_bld_1 (11/19) completed.\n",
      "monocyte_ncnc_bld_1 (12/19) completed.\n",
      "wbc_ncnc_bld_1 (13/19) completed.\n",
      "pmv_entvol_bld_1 (14/19) completed.\n",
      "platelet_ncnc_bld_1 (15/19) completed.\n",
      "EGFRCKDEPI (16/19) completed.\n",
      "CKD (17/19) completed.\n",
      "microcytosis (18/19) completed.\n",
      "anemia (19/19) completed.\n"
     ]
    }
   ],
   "source": [
    "for phenotype, i in zip(phenotype_list, range(len(phenotype_list))):\n",
    "    prepare_matrix_pheno_agesex_in_pheno(cohort, phenotype, pheno_full_df, var_name, var_df, table_name, kinship_df_ID, pc_df)\n",
    "    print(\"%s (%d/%d) completed.\"%(phenotype, i + 1, len(phenotype_list)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### rs11248850 when DEL = 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 238,
   "metadata": {},
   "outputs": [],
   "source": [
    "phenotype_list = [\"hemoglobin_mcnc_bld_1\", \"hematocrit_vfr_bld_1\", \"rbc_ncnc_bld_1\",\n",
    "                  \"mcv_entvol_rbc_1\", \"mch_entmass_rbc_1\", \"mchc_mcnc_rbc_1\", \"rdw_ratio_rbc_1\",\n",
    "                  \"neutrophil_ncnc_bld_1\", \"lymphocyte_ncnc_bld_1\", \"basophil_ncnc_bld_1\",\n",
    "                  \"eosinophil_ncnc_bld_1\", \"monocyte_ncnc_bld_1\", \"wbc_ncnc_bld_1\",\n",
    "                  \"pmv_entvol_bld_1\", \"platelet_ncnc_bld_1\", \"EGFRCKDEPI\", \"CKD\", \"microcytosis\", \"anemia\"]\n",
    "load_dir = os.path.join(\"..\", \"cohort\", cohort, \"ready_data\")\n",
    "\n",
    "pheno_full_filename = \"common_pheno.tsv\"\n",
    "pheno_full_dir_filename = os.path.join(load_dir, pheno_full_filename)\n",
    "pheno_full_df = pd.read_csv(pheno_full_dir_filename, sep=\"\\t\")\n",
    "\n",
    "var_name = \"rs11248850_del0\"\n",
    "\n",
    "var_filename = \"common_rs11248850_del0.tsv\"\n",
    "var_dir_filename = os.path.join(load_dir, var_filename)\n",
    "var_df = pd.read_csv(var_dir_filename, sep=\"\\t\")\n",
    "\n",
    "table_name = \"table3\"\n",
    "\n",
    "kinship_filename = \"common_kinship.tsv\"\n",
    "kinship_dir_filename = os.path.join(load_dir, kinship_filename)\n",
    "kinship_df_ID = pd.read_csv(kinship_dir_filename, sep=\"\\t\")\n",
    "\n",
    "pc_filename = \"common_pc.tsv\"\n",
    "pc_dir_filename = os.path.join(load_dir, pc_filename)\n",
    "pc_df = pd.read_csv(pc_dir_filename, sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 239,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "hemoglobin_mcnc_bld_1 (1/19) completed.\n",
      "hematocrit_vfr_bld_1 (2/19) completed.\n",
      "rbc_ncnc_bld_1 (3/19) completed.\n",
      "mcv_entvol_rbc_1 (4/19) completed.\n",
      "mch_entmass_rbc_1 (5/19) completed.\n",
      "mchc_mcnc_rbc_1 (6/19) completed.\n",
      "rdw_ratio_rbc_1 (7/19) completed.\n",
      "neutrophil_ncnc_bld_1 (8/19) completed.\n",
      "lymphocyte_ncnc_bld_1 (9/19) completed.\n",
      "basophil_ncnc_bld_1 (10/19) completed.\n",
      "eosinophil_ncnc_bld_1 (11/19) completed.\n",
      "monocyte_ncnc_bld_1 (12/19) completed.\n",
      "wbc_ncnc_bld_1 (13/19) completed.\n",
      "pmv_entvol_bld_1 (14/19) completed.\n",
      "platelet_ncnc_bld_1 (15/19) completed.\n",
      "EGFRCKDEPI (16/19) completed.\n",
      "CKD (17/19) completed.\n",
      "microcytosis (18/19) completed.\n",
      "anemia (19/19) completed.\n"
     ]
    }
   ],
   "source": [
    "for phenotype, i in zip(phenotype_list, range(len(phenotype_list))):\n",
    "    prepare_matrix_pheno_agesex_in_pheno(cohort, phenotype, pheno_full_df, var_name, var_df, table_name, kinship_df_ID, pc_df)\n",
    "    print(\"%s (%d/%d) completed.\"%(phenotype, i + 1, len(phenotype_list)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### rs11248850 when DEL = 1, 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 240,
   "metadata": {},
   "outputs": [],
   "source": [
    "phenotype_list = [\"hemoglobin_mcnc_bld_1\", \"hematocrit_vfr_bld_1\", \"rbc_ncnc_bld_1\",\n",
    "                  \"mcv_entvol_rbc_1\", \"mch_entmass_rbc_1\", \"mchc_mcnc_rbc_1\", \"rdw_ratio_rbc_1\",\n",
    "                  \"neutrophil_ncnc_bld_1\", \"lymphocyte_ncnc_bld_1\", \"basophil_ncnc_bld_1\",\n",
    "                  \"eosinophil_ncnc_bld_1\", \"monocyte_ncnc_bld_1\", \"wbc_ncnc_bld_1\",\n",
    "                  \"pmv_entvol_bld_1\", \"platelet_ncnc_bld_1\", \"EGFRCKDEPI\", \"CKD\", \"microcytosis\", \"anemia\"]\n",
    "load_dir = os.path.join(\"..\", \"cohort\", cohort, \"ready_data\")\n",
    "\n",
    "pheno_full_filename = \"common_pheno.tsv\"\n",
    "pheno_full_dir_filename = os.path.join(load_dir, pheno_full_filename)\n",
    "pheno_full_df = pd.read_csv(pheno_full_dir_filename, sep=\"\\t\")\n",
    "\n",
    "var_name = \"rs11248850_del12\"\n",
    "\n",
    "var_filename = \"common_rs11248850_del12.tsv\"\n",
    "var_dir_filename = os.path.join(load_dir, var_filename)\n",
    "var_df = pd.read_csv(var_dir_filename, sep=\"\\t\")\n",
    "\n",
    "table_name = \"table3\"\n",
    "\n",
    "kinship_filename = \"common_kinship.tsv\"\n",
    "kinship_dir_filename = os.path.join(load_dir, kinship_filename)\n",
    "kinship_df_ID = pd.read_csv(kinship_dir_filename, sep=\"\\t\")\n",
    "\n",
    "pc_filename = \"common_pc.tsv\"\n",
    "pc_dir_filename = os.path.join(load_dir, pc_filename)\n",
    "pc_df = pd.read_csv(pc_dir_filename, sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 241,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "hemoglobin_mcnc_bld_1 (1/19) completed.\n",
      "hematocrit_vfr_bld_1 (2/19) completed.\n",
      "rbc_ncnc_bld_1 (3/19) completed.\n",
      "mcv_entvol_rbc_1 (4/19) completed.\n",
      "mch_entmass_rbc_1 (5/19) completed.\n",
      "mchc_mcnc_rbc_1 (6/19) completed.\n",
      "rdw_ratio_rbc_1 (7/19) completed.\n",
      "neutrophil_ncnc_bld_1 (8/19) completed.\n",
      "lymphocyte_ncnc_bld_1 (9/19) completed.\n",
      "basophil_ncnc_bld_1 (10/19) completed.\n",
      "eosinophil_ncnc_bld_1 (11/19) completed.\n",
      "monocyte_ncnc_bld_1 (12/19) completed.\n",
      "wbc_ncnc_bld_1 (13/19) completed.\n",
      "pmv_entvol_bld_1 (14/19) completed.\n",
      "platelet_ncnc_bld_1 (15/19) completed.\n",
      "EGFRCKDEPI (16/19) completed.\n",
      "CKD (17/19) completed.\n",
      "microcytosis (18/19) completed.\n",
      "anemia (19/19) completed.\n"
     ]
    }
   ],
   "source": [
    "for phenotype, i in zip(phenotype_list, range(len(phenotype_list))):\n",
    "    prepare_matrix_pheno_agesex_in_pheno(cohort, phenotype, pheno_full_df, var_name, var_df, table_name, kinship_df_ID, pc_df)\n",
    "    print(\"%s (%d/%d) completed.\"%(phenotype, i + 1, len(phenotype_list)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### rs11248850"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 242,
   "metadata": {},
   "outputs": [],
   "source": [
    "phenotype_list = [\"hemoglobin_mcnc_bld_1\", \"hematocrit_vfr_bld_1\", \"rbc_ncnc_bld_1\",\n",
    "                  \"mcv_entvol_rbc_1\", \"mch_entmass_rbc_1\", \"mchc_mcnc_rbc_1\", \"rdw_ratio_rbc_1\",\n",
    "                  \"neutrophil_ncnc_bld_1\", \"lymphocyte_ncnc_bld_1\", \"basophil_ncnc_bld_1\",\n",
    "                  \"eosinophil_ncnc_bld_1\", \"monocyte_ncnc_bld_1\", \"wbc_ncnc_bld_1\",\n",
    "                  \"pmv_entvol_bld_1\", \"platelet_ncnc_bld_1\", \"EGFRCKDEPI\", \"CKD\", \"microcytosis\", \"anemia\"]\n",
    "load_dir = os.path.join(\"..\", \"cohort\", cohort, \"ready_data\")\n",
    "\n",
    "pheno_full_filename = \"common_pheno.tsv\"\n",
    "pheno_full_dir_filename = os.path.join(load_dir, pheno_full_filename)\n",
    "pheno_full_df = pd.read_csv(pheno_full_dir_filename, sep=\"\\t\")\n",
    "\n",
    "var_name = \"rs11248850\"\n",
    "\n",
    "var_filename = \"common_rs11248850.tsv\"\n",
    "var_dir_filename = os.path.join(load_dir, var_filename)\n",
    "var_df = pd.read_csv(var_dir_filename, sep=\"\\t\")\n",
    "\n",
    "table_name = \"table3\"\n",
    "\n",
    "kinship_filename = \"common_kinship.tsv\"\n",
    "kinship_dir_filename = os.path.join(load_dir, kinship_filename)\n",
    "kinship_df_ID = pd.read_csv(kinship_dir_filename, sep=\"\\t\")\n",
    "\n",
    "pc_filename = \"common_pc.tsv\"\n",
    "pc_dir_filename = os.path.join(load_dir, pc_filename)\n",
    "pc_df = pd.read_csv(pc_dir_filename, sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 243,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "hemoglobin_mcnc_bld_1 (1/19) completed.\n",
      "hematocrit_vfr_bld_1 (2/19) completed.\n",
      "rbc_ncnc_bld_1 (3/19) completed.\n",
      "mcv_entvol_rbc_1 (4/19) completed.\n",
      "mch_entmass_rbc_1 (5/19) completed.\n",
      "mchc_mcnc_rbc_1 (6/19) completed.\n",
      "rdw_ratio_rbc_1 (7/19) completed.\n",
      "neutrophil_ncnc_bld_1 (8/19) completed.\n",
      "lymphocyte_ncnc_bld_1 (9/19) completed.\n",
      "basophil_ncnc_bld_1 (10/19) completed.\n",
      "eosinophil_ncnc_bld_1 (11/19) completed.\n",
      "monocyte_ncnc_bld_1 (12/19) completed.\n",
      "wbc_ncnc_bld_1 (13/19) completed.\n",
      "pmv_entvol_bld_1 (14/19) completed.\n",
      "platelet_ncnc_bld_1 (15/19) completed.\n",
      "EGFRCKDEPI (16/19) completed.\n",
      "CKD (17/19) completed.\n",
      "microcytosis (18/19) completed.\n",
      "anemia (19/19) completed.\n"
     ]
    }
   ],
   "source": [
    "for phenotype, i in zip(phenotype_list, range(len(phenotype_list))):\n",
    "    prepare_matrix_pheno_agesex_in_pheno(cohort, phenotype, pheno_full_df, var_name, var_df, table_name, kinship_df_ID, pc_df)\n",
    "    print(\"%s (%d/%d) completed.\"%(phenotype, i + 1, len(phenotype_list)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### GENxGEN Interaction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 244,
   "metadata": {},
   "outputs": [],
   "source": [
    "phenotype_list = [\"hemoglobin_mcnc_bld_1\", \"hematocrit_vfr_bld_1\", \"rbc_ncnc_bld_1\",\n",
    "                  \"mcv_entvol_rbc_1\", \"mch_entmass_rbc_1\", \"mchc_mcnc_rbc_1\", \"rdw_ratio_rbc_1\",\n",
    "                  \"neutrophil_ncnc_bld_1\", \"lymphocyte_ncnc_bld_1\", \"basophil_ncnc_bld_1\",\n",
    "                  \"eosinophil_ncnc_bld_1\", \"monocyte_ncnc_bld_1\", \"wbc_ncnc_bld_1\",\n",
    "                  \"pmv_entvol_bld_1\", \"platelet_ncnc_bld_1\", \"EGFRCKDEPI\", \"CKD\", \"microcytosis\", \"anemia\"]\n",
    "load_dir = os.path.join(\"..\", \"cohort\", cohort, \"ready_data\")\n",
    "\n",
    "pheno_full_filename = \"common_pheno.tsv\"\n",
    "pheno_full_dir_filename = os.path.join(load_dir, pheno_full_filename)\n",
    "pheno_full_df = pd.read_csv(pheno_full_dir_filename, sep=\"\\t\")\n",
    "\n",
    "var_name = \"del_rs11248850\"\n",
    "\n",
    "var_filename = \"common_del-rs11248850.tsv\"\n",
    "var_dir_filename = os.path.join(load_dir, var_filename)\n",
    "var_df = pd.read_csv(var_dir_filename, sep=\"\\t\")\n",
    "\n",
    "table_name = \"table3\"\n",
    "\n",
    "kinship_filename = \"common_kinship.tsv\"\n",
    "kinship_dir_filename = os.path.join(load_dir, kinship_filename)\n",
    "kinship_df_ID = pd.read_csv(kinship_dir_filename, sep=\"\\t\")\n",
    "\n",
    "pc_filename = \"common_pc.tsv\"\n",
    "pc_dir_filename = os.path.join(load_dir, pc_filename)\n",
    "pc_df = pd.read_csv(pc_dir_filename, sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 245,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "hemoglobin_mcnc_bld_1 (1/19) completed.\n",
      "hematocrit_vfr_bld_1 (2/19) completed.\n",
      "rbc_ncnc_bld_1 (3/19) completed.\n",
      "mcv_entvol_rbc_1 (4/19) completed.\n",
      "mch_entmass_rbc_1 (5/19) completed.\n",
      "mchc_mcnc_rbc_1 (6/19) completed.\n",
      "rdw_ratio_rbc_1 (7/19) completed.\n",
      "neutrophil_ncnc_bld_1 (8/19) completed.\n",
      "lymphocyte_ncnc_bld_1 (9/19) completed.\n",
      "basophil_ncnc_bld_1 (10/19) completed.\n",
      "eosinophil_ncnc_bld_1 (11/19) completed.\n",
      "monocyte_ncnc_bld_1 (12/19) completed.\n",
      "wbc_ncnc_bld_1 (13/19) completed.\n",
      "pmv_entvol_bld_1 (14/19) completed.\n",
      "platelet_ncnc_bld_1 (15/19) completed.\n",
      "EGFRCKDEPI (16/19) completed.\n",
      "CKD (17/19) completed.\n",
      "microcytosis (18/19) completed.\n",
      "anemia (19/19) completed.\n"
     ]
    }
   ],
   "source": [
    "for phenotype, i in zip(phenotype_list, range(len(phenotype_list))):\n",
    "    prepare_matrix_pheno_agesex_in_pheno(cohort, phenotype, pheno_full_df, var_name, var_df, table_name, kinship_df_ID, pc_df)\n",
    "    print(\"%s (%d/%d) completed.\"%(phenotype, i + 1, len(phenotype_list)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Scan All Cohorts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "def save_common_df(common_0, common_1, common_col, save_dir, file_type):\n",
    "    common_df_list = [common_0, common_1]\n",
    "    common_df = merge_df_list(common_df_list, common_col, merge_method='first')\n",
    "    common_df_filename = \"common_%s.tsv\"%file_type\n",
    "    common_df_dir_filename = os.path.join(save_dir, common_df_filename)\n",
    "    common_df.to_csv(common_df_dir_filename, sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "def common_snp_pc_cn_pheno_kinship(cohort, freeze_ver, pc_num, common_col, snp_list, kinship_df_ID, cn_var = \"cn\", pheno_filename = None):\n",
    "    if pheno_filename == None:\n",
    "        pheno_filename = \"pheno_%s.tsv\"%cohort\n",
    "    save_dir = os.path.join(\"..\", \"cohort\", cohort, \"ready_data\")\n",
    "    if not os.path.isdir(save_dir):\n",
    "        os.makedirs(save_dir, exist_ok = True) \n",
    "    snp_dir_filename_list = [os.path.join(\"..\", \"prepro_data\", \"snp\", \"%s_%s_hetero.tsv\"%(freeze_ver, snp_i)) for snp_i in snp_list]\n",
    "    pc_dir_filename = os.path.join(\"..\", \"prepro_data\", \"pc\", \"%s_pc%d_pcair.tsv\"%(freeze_ver, pc_num))\n",
    "    cn_dir_filename = os.path.join(\"..\", \"cohort\", cohort, cn_var, \"%s_%s.tsv\"%(cohort, cn_var))\n",
    "    pheno_dir_filename = os.path.join(\"..\", \"cohort\", cohort, \"pre_data\", pheno_filename)\n",
    "    kinship_dir_filename = os.path.join(\"..\", \"prepro_data\", \"kinship\", \"%s_kinship_sample.tsv\"%freeze_ver)\n",
    "    snp_pc_cn_dir_filename_list = snp_dir_filename_list + [pc_dir_filename, cn_dir_filename, pheno_dir_filename, kinship_dir_filename]\n",
    "    \n",
    "    snp_pc_cn_df = read2df_list(snp_pc_cn_dir_filename_list)\n",
    "    common_snp_0 = merge_df_list(snp_pc_cn_df, common_col, merge_method='first', how = 'inner')\n",
    "    \n",
    "    common_snp_0_filename = \"common_%s.tsv\"%snp_list[0]\n",
    "    common_snp_0_dir_filename = os.path.join(save_dir, common_snp_0_filename)\n",
    "    common_snp_0.to_csv(common_snp_0_dir_filename, sep=\"\\t\", index=False)\n",
    "    \n",
    "    sample_df = common_snp_0[[\"NWDID\"]]\n",
    "    sample_filename = \"common_sample.tsv\"\n",
    "    sample_dir_filename = os.path.join(save_dir, sample_filename)\n",
    "    sample_df.to_csv(sample_dir_filename, sep=\"\\t\", index=False)\n",
    "    \n",
    "    snp_num = len(snp_list)\n",
    "    for snp_idx in range(1, snp_num):\n",
    "        common_0 = snp_pc_cn_df[snp_idx]\n",
    "        common_1 = common_snp_0\n",
    "        file_type = snp_list[snp_idx]\n",
    "        save_common_df(common_0, common_1, common_col, save_dir, file_type)\n",
    "    \n",
    "    save_common_df(snp_pc_cn_df[snp_num], common_snp_0, common_col, save_dir, \"pc\")\n",
    "    save_common_df(snp_pc_cn_df[snp_num + 1], common_snp_0, common_col, save_dir, cn_var)\n",
    "    save_common_df(snp_pc_cn_df[snp_num + 2], common_snp_0, common_col, save_dir, \"pheno\")\n",
    "    if cn_var == \"cn\":\n",
    "        common_del = cn2del(snp_pc_cn_df[snp_num + 1])\n",
    "        save_common_df(common_del, common_snp_0, common_col, save_dir, \"del\")\n",
    "    \n",
    "    kinship_sample_selected = kinship_select_sample(kinship_df_ID, sample_df)\n",
    "    print(kinship_sample_selected.shape)\n",
    "    common_kinship_filename_prefix = \"common_kinship\"\n",
    "    save_kinship_simple(save_dir, common_kinship_filename_prefix, kinship_sample_selected, sample_col = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "def categorize_df(df, cat_col):\n",
    "    df_raw = df.copy()\n",
    "    df_raw[cat_col].fillna(\"Missing Values\", inplace = True) \n",
    "    category = list(set(df_raw[cat_col].values.tolist()))\n",
    "    cat_num_list = []\n",
    "    for cat_i in category:\n",
    "        tmp_cat_df = df.loc[df_raw[cat_col] == cat_i, :]\n",
    "        tmp_cat_num = tmp_cat_df.shape[0]\n",
    "        cat_num_list.append(tmp_cat_num)\n",
    "    category_summary_tuple = list(zip(category, cat_num_list))\n",
    "    category_summary_df = pd.DataFrame(category_summary_tuple, columns = [cat_col, \"sample_size\"])\n",
    "    category_summary_df.sort_values(axis = 0, by = cat_col, inplace = True)\n",
    "    return category_summary_df, category"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "def genxgen(gen_1, gen_2, gen_1_name, gen_2_name, common_col, save_dir):\n",
    "#     gen_gen = pd.concat(objs=[gen_1, gen_2[[gen_2_name]]], axis=1)\n",
    "    gen_gen = gen_1.merge(gen_2, on = common_col, how = \"inner\")\n",
    "    gen_gen[\"%s-%s\"%(gen_1_name, gen_2_name)] = gen_gen[gen_1_name] * gen_gen[gen_2_name]\n",
    "    rearrange_list = [list(gen_gen)[0]] + [\"%s-%s\"%(gen_1_name, gen_2_name), gen_1_name, gen_2_name]\n",
    "    gen_gen_rearrange = gen_gen.copy()\n",
    "    gen_gen_rearrange = gen_gen[rearrange_list]\n",
    "    gen_gen_rearrange_filename = \"common_%s-%s.tsv\"%(gen_1_name, gen_2_name)\n",
    "    gen_gen_rearrange_dir_filename = os.path.join(save_dir, gen_gen_rearrange_filename)\n",
    "    gen_gen_rearrange.to_csv(gen_gen_rearrange_dir_filename, sep = \"\\t\", header = True, index = False)\n",
    "    return gen_gen_rearrange"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "def table1(cohort, cn_var):\n",
    "    cn_orig_filename = \"common_%s.tsv\"%cn_var\n",
    "    cn_orig_dir = os.path.join(\"..\", \"cohort\", cohort, \"ready_data\")\n",
    "    cn_orig_dir_filename = os.path.join(cn_orig_dir, cn_orig_filename)\n",
    "    cn_orig_df = pd.read_csv(cn_orig_dir_filename, sep=\"\\t\")\n",
    "    if cn_var == \"cn\":\n",
    "        cn_binary_df_full = cn2binary_df(cn_orig_df)\n",
    "        cn_binary_df = cn_binary_df_full[[\"NWDID\", \"cn0\", \"cn1\", \"cn3\", \"cn4\"]]\n",
    "        cn_type_list = [\"cn0\", \"cn1\", \"cn3\"]\n",
    "        for cn_type in cn_type_list:\n",
    "            cn_binary_df_move = move_cn(cn_binary_df, cn_type)\n",
    "            cn_binary_df_move_filename = \"common_cn-binary_%s.tsv\"%cn_type\n",
    "            cn_binary_df_move_dir_filename = os.path.join(cn_orig_dir, cn_binary_df_move_filename)\n",
    "            cn_binary_df_move.to_csv(cn_binary_df_move_dir_filename, sep=\"\\t\", index=False)\n",
    "    # to do: add the case of cn_var == \"status\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "def table2(cohort, gen_name):\n",
    "#     Load del\n",
    "    load_dir = os.path.join(\"..\", \"cohort\", cohort, \"ready_data\")\n",
    "    save_dir = load_dir\n",
    "    del_cn_filename = \"common_del.tsv\"\n",
    "    del_cn_dir_filename = os.path.join(load_dir, del_cn_filename)\n",
    "    del_cn = pd.read_csv(del_cn_dir_filename, sep=\"\\t\", header = 0, index_col = None)\n",
    "    \n",
    "#     del x gen\n",
    "    gen_filename = \"common_%s.tsv\"%gen_name\n",
    "    gen_dir_filename = os.path.join(load_dir, gen_filename)\n",
    "    gen = pd.read_csv(gen_dir_filename, sep=\"\\t\")\n",
    "    common_col = \"NWDID\"\n",
    "    del_gen = genxgen(del_cn, gen, \"del\", gen_name, common_col, save_dir)\n",
    "    \n",
    "#     Stratify gen by # del\n",
    "    del_val_tuple = ([0], [1], [2])\n",
    "    var_name = \"del\"\n",
    "    for del_val_list in del_val_tuple:\n",
    "        strati_gen(del_gen, del_val_list, var_name, gen_name, save_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "def strati_gen(var_gen, var_val_list, var_name, gen_name, save_dir):\n",
    "    gen_var_val = var_gen.loc[var_gen.loc[:, var_name].isin(var_val_list), [\"NWDID\", gen_name]]\n",
    "    var_val_name = \"\".join(map(str, var_val_list))\n",
    "    gen_var_val_filename = \"common_%s_%s%s.tsv\"%(gen_name, var_name, var_val_name)\n",
    "    gen_var_val_dir_filename = os.path.join(save_dir, gen_var_val_filename)\n",
    "    gen_var_val.to_csv(gen_var_val_dir_filename, sep = \"\\t\", header = True, index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "def table3(cohort, gen_name, table3_del_val_tuple, table3_cn_val_tuple):\n",
    "#     Load del and cn\n",
    "    load_dir = os.path.join(\"..\", \"cohort\", cohort, \"ready_data\")\n",
    "    save_dir = load_dir\n",
    "    del_cn_filename = \"common_del.tsv\"\n",
    "    del_cn_dir_filename = os.path.join(load_dir, del_cn_filename)\n",
    "    del_cn = pd.read_csv(del_cn_dir_filename, sep=\"\\t\", header = 0, index_col = None)\n",
    "    \n",
    "    cn_filename = \"common_cn.tsv\"\n",
    "    cn_dir_filename = os.path.join(load_dir, cn_filename)\n",
    "    cn = pd.read_csv(cn_dir_filename, sep=\"\\t\", header = 0, index_col = None)\n",
    "    \n",
    "#     cn x gen, del x gen\n",
    "    gen_filename = \"common_%s.tsv\"%gen_name\n",
    "    gen_dir_filename = os.path.join(load_dir, gen_filename)\n",
    "    gen = pd.read_csv(gen_dir_filename, sep=\"\\t\")\n",
    "    common_col = \"NWDID\"\n",
    "    cn_gen = genxgen(cn, gen, \"cn\", gen_name, common_col, save_dir)\n",
    "    del_gen = genxgen(del_cn, gen, \"del\", gen_name, common_col, save_dir)\n",
    "    \n",
    "#     Stratify gen by # del\n",
    "    var_name = \"del\"\n",
    "    for del_val_list in table3_del_val_tuple:\n",
    "        strati_gen(del_gen, del_val_list, var_name, gen_name, save_dir)\n",
    "        \n",
    "#     Stratify gen by # cn\n",
    "    var_name = \"cn\"\n",
    "    for cn_val_list in table3_cn_val_tuple:\n",
    "        strati_gen(cn_gen, cn_val_list, var_name, gen_name, save_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [],
   "source": [
    "full_pheno_dir = os.path.join(\"..\", \"prepro_data\", \"phenotype\")\n",
    "full_pheno_filename = \"freeze8_anno04_af02_btc01-coh03_ddimer_egfr03_adad01.tsv\"\n",
    "full_pheno_dir_filename = os.path.join(full_pheno_dir, full_pheno_filename)\n",
    "full_pheno = pd.read_csv(full_pheno_dir_filename, sep = \"\\t\", header = 0, index_col = None)\n",
    "_, cohort_list = categorize_df(full_pheno, \"study\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['FHS', 'COPDGene', 'BioMe', 'ARIC', 'GeneSTAR', 'GenSalt', 'CARDIA', 'HyperGEN', 'DHS', 'HCHS_SOL', 'CHS', 'WHI', 'MESA', 'GENOA', 'JHS']\n"
     ]
    }
   ],
   "source": [
    "print(cohort_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(2594, 2595)\n",
      "FHS common samples found.\n",
      "(4699, 4700)\n",
      "COPDGene common samples found.\n",
      "(8029, 8030)\n",
      "BioMe common samples found.\n",
      "(3015, 3016)\n",
      "ARIC common samples found.\n",
      "(1266, 1267)\n",
      "GeneSTAR common samples found.\n",
      "(1717, 1718)\n",
      "GenSalt common samples found.\n",
      "(2596, 2597)\n",
      "CARDIA common samples found.\n",
      "(1700, 1701)\n",
      "HyperGEN common samples found.\n",
      "(344, 345)\n",
      "DHS common samples found.\n",
      "(3382, 3383)\n",
      "HCHS_SOL common samples found.\n",
      "(2768, 2769)\n",
      "CHS common samples found.\n",
      "(8646, 8647)\n",
      "WHI common samples found.\n",
      "(3972, 3973)\n",
      "MESA common samples found.\n",
      "(1013, 1014)\n",
      "GENOA common samples found.\n",
      "(2933, 2934)\n",
      "JHS common samples found.\n"
     ]
    }
   ],
   "source": [
    "freeze_ver = \"freeze8\"\n",
    "common_col = \"NWDID\"\n",
    "cn_var = \"cn\"\n",
    "pc_num = 11\n",
    "snp_list = [\"rs334\", \"rs33930165\", \"rs11248850\"]\n",
    "pheno_prefix = \"freeze8_anno04_af02_btc03-coh03_egfr03-ckd_adad01\"\n",
    "for cohort in cohort_list:\n",
    "    pheno_filename = \"%s_%s.tsv\"%(pheno_prefix, cohort)\n",
    "    common_snp_pc_cn_pheno_kinship(cohort, freeze_ver, pc_num, common_col, snp_list, freeze8_kinship_df_ID, cn_var, pheno_filename)\n",
    "    print(\"%s common samples found.\"%cohort)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Preparing Data for Each Table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "def load_kinship_simple(load_dir, filename_prefix):\n",
    "    load_dir_filename_prefix = os.path.join(load_dir, filename_prefix)\n",
    "    load_dir_filename_tsv = \"%s.tsv\"%load_dir_filename_prefix\n",
    "    load_dir_filename_feather = \"%s.feather\"%load_dir_filename_prefix\n",
    "    if os.path.exists(load_dir_filename_tsv):\n",
    "        kinship = pd.read_csv(load_dir_filename_tsv, sep = \"\\t\", header = 0, index_col = None)\n",
    "    elif os.path.exists(load_dir_filename_feather):\n",
    "        kinship = pd.read_feather(load_dir_filename_feather, use_threads = True)\n",
    "    return kinship\n",
    "\n",
    "def save_kinship_simple(save_dir, filename_prefix, kinship, sample_col = True):\n",
    "    save_dir_filename_prefix = os.path.join(save_dir, filename_prefix)\n",
    "    save_dir_filename_tsv = \"%s.tsv\"%save_dir_filename_prefix\n",
    "    save_dir_filename_feather = \"%s.feather\"%save_dir_filename_prefix\n",
    "    if kinship.shape[0] > 5000:\n",
    "        if sample_col == False:\n",
    "            kinship.drop(axis = 1, labels = [\"NWDID\"], inplace = True)\n",
    "        feather.write_dataframe(kinship, save_dir_filename_feather)\n",
    "    elif kinship.shape[0] <= 5000:\n",
    "        kinship.to_csv(save_dir_filename_tsv, sep = \"\\t\", header = True, index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "def prepare_matrix_pheno_adad_in_pheno(cohort, phenotype, pheno_full_df, var_name, var_df, table_name, kinship_df_ID, pc10_df, adad_header):\n",
    "    if not os.path.exists(os.path.join(\"..\", \"cohort\", cohort, table_name)):\n",
    "        os.mkdir(os.path.join(\"..\", \"cohort\", cohort, table_name))\n",
    "    if not os.path.exists(os.path.join(\"..\", \"cohort\", cohort, table_name, var_name)):\n",
    "        os.mkdir(os.path.join(\"..\", \"cohort\", cohort, table_name, var_name))\n",
    "        \n",
    "    save_dir = os.path.join(\"..\", \"cohort\", cohort, table_name, var_name)\n",
    "    \n",
    "    pheno_df_raw = pheno_full_df[[\"NWDID\", \"sex\", phenotype, \"age_at_%s\"%phenotype] + adad_header].copy()\n",
    "    pheno_df_raw.dropna(axis = 1, how = \"all\", inplace = True)\n",
    "    adad_header_used = intersection_list(list(pheno_df_raw), adad_header)\n",
    "    if phenotype not in list(pheno_df_raw):\n",
    "        return None\n",
    "    \n",
    "    pheno_agesex_df = pheno_df_raw.dropna(axis=0)\n",
    "    common_col = \"NWDID\"\n",
    "    pheno_agesex_var_df_list = [pheno_agesex_df, var_df]\n",
    "    common_pheno_agesex = merge_df_list(pheno_agesex_var_df_list, common_col, merge_method='first')\n",
    "    common_var = merge_df_list(pheno_agesex_var_df_list, common_col, merge_method='second')\n",
    "    sample_df = common_pheno_agesex[[\"NWDID\"]].copy()\n",
    "    \n",
    "    common_pheno = common_pheno_agesex[[\"NWDID\", phenotype]].copy()\n",
    "    common_pheno_filename = \"common_pheno_%s_%s_%s.tsv\"%(table_name, var_name, phenotype)\n",
    "    common_pheno_dir_filename = os.path.join(save_dir, common_pheno_filename)\n",
    "    common_pheno.to_csv(common_pheno_dir_filename, sep=\"\\t\", index=False)\n",
    "    \n",
    "    common_var_filename = \"common_var_%s_%s_%s.tsv\"%(table_name, var_name, phenotype)\n",
    "    common_var_dir_filename = os.path.join(save_dir, common_var_filename)\n",
    "    common_var.to_csv(common_var_dir_filename, sep=\"\\t\", index=False)\n",
    "\n",
    "    common_kinship = kinship_select_sample(kinship_df_ID, sample_df)\n",
    "    common_kinship_filename_prefix = \"common_kinship_%s_%s_%s\"%(table_name, var_name, phenotype)\n",
    "    save_kinship_simple(save_dir, common_kinship_filename_prefix, common_kinship, sample_col = False)\n",
    "\n",
    "    common_age_sex = common_pheno_agesex[[\"NWDID\", \"age_at_%s\"%phenotype, \"sex\"] + adad_header_used].copy()\n",
    "    common_age_sex_filename = \"common_age-sex_%s_%s_%s.tsv\"%(table_name, var_name, phenotype)\n",
    "    common_age_sex_dir_filename = os.path.join(save_dir, common_age_sex_filename)\n",
    "    common_age_sex.to_csv(common_age_sex_dir_filename, sep=\"\\t\", index=False)\n",
    "    \n",
    "    common_pc10 = merge_df_list([pc10_df, sample_df], common_col, merge_method='first')\n",
    "    common_pc10_filename = \"common_pc10_%s_%s_%s.tsv\"%(table_name, var_name, phenotype)\n",
    "    common_pc10_dir_filename = os.path.join(save_dir, common_pc10_filename)\n",
    "    common_pc10.to_csv(common_pc10_dir_filename, sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "def wrap_prepare_matrix_pheno_adad_in_pheno(phenotype_list, table_dict, load_dir, adad_dict):\n",
    "    for table_name in table_dict:\n",
    "        pheno_full_filename = \"common_pheno.tsv\"\n",
    "        pheno_full_dir_filename = os.path.join(load_dir, pheno_full_filename)\n",
    "        pheno_full_df = pd.read_csv(pheno_full_dir_filename, sep=\"\\t\")\n",
    "        if bool(adad_dict):\n",
    "            pheno_full_df_raw = pheno_full_df.copy()\n",
    "            del pheno_full_df\n",
    "            pheno_full_df, adad_header = adad_header_conversion(pheno_full_df_raw, adad_dict)\n",
    "            pheno_dummy_adad_filename = \"common_pheno_adad_dummy.tsv\"\n",
    "            pheno_dummy_adad_dir_filename = os.path.join(load_dir, pheno_dummy_adad_filename)\n",
    "            pheno_full_df.to_csv(pheno_dummy_adad_dir_filename, sep = \"\\t\", header = True, index = False)\n",
    "        else:\n",
    "            adad_header = []\n",
    "        \n",
    "        kinship_filename_prefix = \"common_kinship\"\n",
    "        kinship_df_ID = load_kinship_simple(load_dir, kinship_filename_prefix)\n",
    "\n",
    "        pc_filename = \"common_pc.tsv\"\n",
    "        pc_dir_filename = os.path.join(load_dir, pc_filename)\n",
    "        pc_df = pd.read_csv(pc_dir_filename, sep=\"\\t\")\n",
    "        \n",
    "        var_name_list = table_dict[table_name]\n",
    "        for var_name in var_name_list:\n",
    "            var_filename = \"common_%s.tsv\"%var_name\n",
    "            var_dir_filename = os.path.join(load_dir, var_filename)\n",
    "            var_df = pd.read_csv(var_dir_filename, sep=\"\\t\")\n",
    "        \n",
    "            for phenotype, i in zip(phenotype_list, range(len(phenotype_list))):\n",
    "                prepare_matrix_pheno_adad_in_pheno(cohort, phenotype, pheno_full_df, var_name, var_df, table_name, kinship_df_ID, pc_df, adad_header)\n",
    "#                print(\"%s (%d/%d) completed.\"%(phenotype, i + 1, len(phenotype_list)))\n",
    "        print(\"%s completed.\"%table_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "def adad_header_conversion(df_original, adad_dict):\n",
    "#     adad_quan_list = adad_dict[\"quan\"]\n",
    "#     adad_cati_list = adad_dict[\"cati\"]\n",
    "#     df, adad_cati_dummy_list = cati2dummy_df(df_original, adad_cati_list)\n",
    "#     adad_header = adad_quan_list + adad_cati_dummy_list\n",
    "    adad_cati_list = adad_dict[\"cati\"]\n",
    "    df, adad_header = cati2dummy_df(df_original, adad_cati_list)\n",
    "    return df, adad_header"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "table1 completed.\n",
      "table2 completed.\n",
      "table3 completed.\n",
      "FHS completed.\n",
      "table1 completed.\n",
      "table2 completed.\n",
      "table3 completed.\n",
      "COPDGene completed.\n",
      "table1 completed.\n",
      "table2 completed.\n",
      "table3 completed.\n",
      "BioMe completed.\n",
      "table1 completed.\n",
      "table2 completed.\n",
      "table3 completed.\n",
      "ARIC completed.\n",
      "table1 completed.\n",
      "table2 completed.\n",
      "table3 completed.\n",
      "GeneSTAR completed.\n",
      "table1 completed.\n",
      "table2 completed.\n",
      "table3 completed.\n",
      "GenSalt completed.\n",
      "table1 completed.\n",
      "table2 completed.\n",
      "table3 completed.\n",
      "CARDIA completed.\n",
      "table1 completed.\n",
      "table2 completed.\n",
      "table3 completed.\n",
      "HyperGEN completed.\n",
      "table1 completed.\n",
      "table2 completed.\n",
      "table3 completed.\n",
      "DHS completed.\n",
      "table1 completed.\n",
      "table2 completed.\n",
      "table3 completed.\n",
      "HCHS_SOL completed.\n",
      "table1 completed.\n",
      "table2 completed.\n",
      "table3 completed.\n",
      "CHS completed.\n",
      "table1 completed.\n",
      "table2 completed.\n",
      "table3 completed.\n",
      "WHI completed.\n",
      "table1 completed.\n",
      "table2 completed.\n",
      "table3 completed.\n",
      "MESA completed.\n",
      "table1 completed.\n",
      "table2 completed.\n",
      "table3 completed.\n",
      "GENOA completed.\n",
      "table1 completed.\n",
      "table2 completed.\n",
      "table3 completed.\n",
      "JHS completed.\n"
     ]
    }
   ],
   "source": [
    "cohort_list = ['FHS', 'COPDGene', 'BioMe', 'ARIC', 'GeneSTAR', 'GenSalt', 'CARDIA', 'HyperGEN', 'DHS',\n",
    "               'HCHS_SOL', 'CHS', 'WHI', 'MESA', 'GENOA', 'JHS']\n",
    "# cohort_list = ['HCHS_SOL']\n",
    "gen_name_1 = \"rs334\"\n",
    "gen_name_2 = \"rs33930165\"\n",
    "gen_name_3 = \"rs11248850\"\n",
    "table3_del_val_tuple = ([0], [1, 2])\n",
    "table3_cn_val_tuple = ([2], [3, 4])\n",
    "cn_var = \"cn\"\n",
    "\n",
    "phenotype_list = [\"hemoglobin_mcnc_bld_1\", \"hematocrit_vfr_bld_1\", \"rbc_ncnc_bld_1\", \"DDIMER\",\n",
    "                  \"mcv_entvol_rbc_1\", \"mch_entmass_rbc_1\", \"mchc_mcnc_rbc_1\", \"rdw_ratio_rbc_1\",\n",
    "                  \"neutrophil_ncnc_bld_1\", \"lymphocyte_ncnc_bld_1\", \"basophil_ncnc_bld_1\",\n",
    "                  \"eosinophil_ncnc_bld_1\", \"monocyte_ncnc_bld_1\", \"wbc_ncnc_bld_1\", \"lnHBA1C\", \n",
    "                  \"pmv_entvol_bld_1\", \"platelet_ncnc_bld_1\", \"EGFRCKDEPI\", \"CKD\", \"microcytosis\", \"anemia\"]\n",
    "table_dict = {\"table1\":[\"cn-binary_cn0\", \"cn-binary_cn1\", \"cn-binary_cn3\", \"rs334\", \"rs33930165\"],\n",
    "              \"table2\":[\"rs334_del0\", \"rs334_del1\", \"rs334_del2\", \"del-rs334\"],\n",
    "              \"table3\":[\"rs11248850\", \"rs11248850_del0\", \"rs11248850_del12\", \"rs11248850_cn2\", \"rs11248850_cn34\", \"del-rs11248850\"]}\n",
    "adad_dict = {\"quan\":[\"WEIGHT_FINAL_NORM_OVERALL\"], \"cati\":[\"gengrp6\", \"CENTER\"]}\n",
    "\n",
    "for cohort in cohort_list:\n",
    "    table1(cohort, cn_var)\n",
    "    table2(cohort, gen_name_1)\n",
    "    table3(cohort, gen_name_3, table3_del_val_tuple, table3_cn_val_tuple)\n",
    "    load_dir = os.path.join(\"..\", \"cohort\", cohort, \"ready_data\")\n",
    "    wrap_prepare_matrix_pheno_adad_in_pheno(phenotype_list, table_dict, load_dir, adad_dict)\n",
    "    print(\"%s completed.\"%cohort)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.10"
  },
  "toc-autonumbering": false
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
